// @mc-made/web-components/dist/components/badge/src/badge.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './badge.styles.js';
class MadeBadge extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotContent = new HasSlotController(this, '[default]');
        /** Text or number displayed inside the badge. */
        this.content = '';
        /** The badge's sentiment defines it's color. */
        this.sentiment = 'neutral';
        /** Type of content or status dot.  */
        this.variant = 'status';
        /** Size of the badge. */
        this.size = 'medium';
        /** Alter the contrast of the sentiment attribute. */
        this.contrast = 'high';
        /** Control the visibility of the Badge. */
        this.invisible = false;
        /** When the badge component wraps an associated component, control which corner it's positioned at. */
        this.anchorOrigin = { vertical: 'top', horizontal: 'right' };
        /** Set to true when the badge component wraps an associated component thats circular for better positioning.  */
        this.circularOverlap = false;
    }
    render() {
        return html `
      <span
        part="base"
        class=${classMap({
            'made-c-badge': true,
            'made-c-badge--invisible': this.invisible,
            'made-c-badge--neutral': this.sentiment === 'neutral',
            'made-c-badge--success': this.sentiment === 'success',
            'made-c-badge--information': this.sentiment === 'information',
            'made-c-badge--warning': this.sentiment === 'warning',
            'made-c-badge--danger': this.sentiment === 'danger',
            'made-c-badge--brand': this.sentiment === 'brand',
            'made-c-badge--number': this.variant === 'number',
            'made-c-badge--text': this.variant === 'text',
            'made-c-badge--status': this.variant === 'status',
            'made-c-badge--large': this.size === 'large' && this.variant !== 'status',
            'made-c-badge--has-slot-content': this.hasSlotContent.test('[default]'),
            'made-c-badge--low-contrast': this.contrast === 'low',
            'made-c-badge--top': this.anchorOrigin.vertical === 'top',
            'made-c-badge--bottom': this.anchorOrigin.vertical === 'bottom',
            'made-c-badge--left': this.anchorOrigin.horizontal === 'left',
            'made-c-badge--right': this.anchorOrigin.horizontal === 'right',
            'made-c-badge--circular-overlap': this.circularOverlap
        })}
        role="status"
        >${this.variant === 'status' ? '' : this.content}</span
      >
      <slot></slot>
    `;
    }
}
MadeBadge.styles = [componentStyles, styles];
export default MadeBadge;
__decorate([
    property()
], MadeBadge.prototype, "content", void 0);
__decorate([
    property({ reflect: true })
], MadeBadge.prototype, "sentiment", void 0);
__decorate([
    property({ reflect: true })
], MadeBadge.prototype, "variant", void 0);
__decorate([
    property()
], MadeBadge.prototype, "size", void 0);
__decorate([
    property()
], MadeBadge.prototype, "contrast", void 0);
__decorate([
    property({ type: Boolean })
], MadeBadge.prototype, "invisible", void 0);
__decorate([
    property({ type: Object, attribute: 'anchor-origin' })
], MadeBadge.prototype, "anchorOrigin", void 0);
__decorate([
    property({ type: Boolean, attribute: 'circular-overlap' })
], MadeBadge.prototype, "circularOverlap", void 0);
