// @mc-made/web-components/dist/components/badge/src/badge.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  :host {
    display: inline-flex;
    position: relative;
  }

  .made-c-badge {
    display: inline-flex;
    justify-content: center;
    align-items: center;
    flex-shrink: 0;
    color: var(--made-color-text-on-color);
    text-align: center;
    font-family: var(--made-font-family-primary);
    font-size: var(--made-component-badge-medium-font-size);
    font-style: normal;
    font-weight: var(--made-component-badge-medium-font-weight);
    line-height: var(--made-component-badge-medium-line-height);
    letter-spacing: var(--made-component-badge-medium-letter-spacing);
    border-radius: var(--made-component-badge-border-radius);
    transition: transform var(--made-motion-duration-default)
      var(--made-motion-easing-standard-ease);
    box-shadow: var(--made-component-badge-box-shadow);
    z-index: var(--made-z-index-1);
  }

  .made-c-badge:is(.made-c-badge--text, .made-c-badge--number) {
    height: var(--made-size-space-4-x);
    min-width: var(--made-size-space-4-x);
  }

  .made-c-badge--large:is(.made-c-badge--text, .made-c-badge--number) {
    height: var(--made-size-space-5-x);
    min-width: var(--made-size-space-5-x);
    font-size: var(--made-component-badge-large-font-size);
    font-weight: var(--made-component-badge-large-font-weight);
    line-height: var(--made-component-badge-large-line-height);
    letter-spacing: var(--made-component-badge-large-letter-spacing);
  }

  .made-c-badge--text {
    padding: var(--made-size-space-0point5-x)
      var(--made-component-badge-medium-padding-x);

    &.made-c-badge--large {
      padding: var(--made-size-space-0point5-x)
        var(--made-component-badge-large-padding-x);
    }
  }

  .made-c-badge--number {
    padding: 0 var(--made-size-space-1-x);
  }

  .made-c-badge--status {
    width: var(--made-size-space-2-x);
    height: var(--made-size-space-2-x);
  }

  .made-c-badge--neutral {
    border: 1px solid var(--made-color-border-neutral-strong);
    background: var(--made-color-background-feedback-neutral-strong-default);
  }

  .made-c-badge--brand {
    border: 1px solid var(--made-color-border-brand);
    background: var(--made-color-background-brand-primary-default);
  }

  .made-c-badge--success {
    border: 1px solid var(--made-color-border-success-strong);
    background: var(--made-color-background-feedback-success-strong-default);
  }

  .made-c-badge--danger {
    border: 1px solid var(--made-color-border-danger-strong);
    background: var(--made-color-background-feedback-danger-strong-default);
  }

  .made-c-badge--warning {
    border: 1px solid var(--made-color-border-warning-strong);
    background: var(--made-color-background-feedback-warning-strong-default);
    color: var(--made-component-badge-warning-contrast-high-text-color);
  }

  .made-c-badge--information {
    border: 1px solid var(--made-color-border-information-strong);
    background: var(
      --made-color-background-feedback-information-strong-default
    );
  }

  .made-c-badge--neutral.made-c-badge--low-contrast:not(.made-c-badge--status) {
    border: 1px solid var(--made-color-border-neutral);
    background: var(--made-color-background-feedback-neutral-default);
    color: var(--made-color-text-primary);
  }

  .made-c-badge--success.made-c-badge--low-contrast:not(.made-c-badge--status) {
    border: 1px solid var(--made-color-border-success);
    background: var(--made-color-background-feedback-success-default);
    color: var(--made-color-text-success);
  }

  .made-c-badge--danger.made-c-badge--low-contrast:not(.made-c-badge--status) {
    border: 1px solid var(--made-color-border-danger);
    background: var(--made-color-background-feedback-danger-default);
    color: var(--made-color-text-danger);
  }

  .made-c-badge--warning.made-c-badge--low-contrast:not(.made-c-badge--status) {
    border: 1px solid var(--made-color-border-warning);
    background: var(--made-color-background-feedback-warning-default);
    color: var(--made-color-text-warning);
  }

  .made-c-badge--information.made-c-badge--low-contrast:not(
      .made-c-badge--status
    ) {
    border: 1px solid var(--made-color-border-information);
    background: var(--made-color-background-feedback-information-default);
    color: var(--made-color-text-information);
  }

  /** Positioning when has slot */
  .made-c-badge.made-c-badge--has-slot-content {
    position: absolute;
  }

  .made-c-badge--has-slot-content.made-c-badge--top.made-c-badge--left {
    top: 0;
    left: 0;
    transform: scale(1) translate(-50%, -50%);
    transform-origin: 0% 0%;

    &.made-c-badge--circular-overlap {
      top: 14%;
      left: 14%;
    }
  }

  .made-c-badge--has-slot-content.made-c-badge--top.made-c-badge--right {
    top: 0;
    right: 0;
    transform: scale(1) translate(50%, -50%);
    transform-origin: 100% 0%;

    &.made-c-badge--circular-overlap {
      top: 14%;
      right: 14%;
    }
  }

  .made-c-badge--has-slot-content.made-c-badge--bottom.made-c-badge--left {
    bottom: 0;
    left: 0;
    transform: scale(1) translate(-50%, 50%);
    transform-origin: 0% 100%;

    &.made-c-badge--circular-overlap {
      bottom: 14%;
      left: 14%;
    }
  }

  .made-c-badge--has-slot-content.made-c-badge--bottom.made-c-badge--right {
    bottom: 0;
    right: 0;
    transform: scale(1) translate(50%, 50%);
    transform-origin: 100% 100%;

    &.made-c-badge--circular-overlap {
      bottom: 14%;
      right: 14%;
    }
  }

  /** Visibility */
  .made-c-badge.made-c-badge--invisible,
  .made-c-badge.made-c-badge--has-slot-content.made-c-badge--invisible {
    transform: scale(0);
  }
`;
