// @mc-made/web-components/dist/components/tabs/src/tabs.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
/* eslint-disable @typescript-eslint/await-thenable */
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property, query, state } from 'lit/decorators.js';
import { scrollIntoView } from '../../../internal/scroll.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './tabs.styles.js';
/**
 * @summary Tabs are used to group different but related content, allowing users to navigate
 * views without leaving the page. They always contain at least two items and one
 * tab is active at a time. Tabs can be used on full page layouts or in components
 * such as modals, cards, or side panels.
 * @documentation '../stories/tabs.mdx'
 *
 * @fires change (CustomEvent<{name: string}>) - Emitted when a tab is activated.
 *
 * @slot - The default slot. Used for grouping tab panels in the tabs component.
 * Must be a <made-tab-panel>.
 * @slot nav - Used for grouping tab elements inside the tabs element. Must be a <made-tab>.
 * Each <made-tab> will automatically use this slot.
 *
 * @csspart base - The component's base wrapper.
 * @csspart nav - The navigation container where tabs are slotted.
 * @csspart tabs - The container that wraps the tabs.
 * @csspart body - The container where tab-panels are slotted.
 */
class MadeTabs extends MadeElement {
    constructor() {
        super(...arguments);
        this.tabs = [];
        this.focusableTabs = [];
        this.panels = [];
        this.firstRender = true;
        this.hasStartShadow = false;
        this.hasEndShadow = false;
        /** Sets the active tab. */
        this.active = '';
        /** The placement of the tabs. */
        this.placement = 'top';
        /** Sets the visual appearance of the tabs. */
        this.appearance = 'default';
        /** Renders the tabs at a reduced size. */
        this.compact = false;
        /**
         * Keyboard navigation uses the "manual activation" behavior. If you want to
         * switch to the "selection automatically follows focus" behavior set this to true.
         */
        this.selectionFollowsFocus = false;
        /**
         * Expand tabs to fill the parnet container.
         * Consider using the `full-width` attribute on smaller screens.
         */
        this.fullWidth = false;
        /**
         * Centers the tabs in the availble space.
         * Consider using the `centered` attribute on larger screens.
         */
        this.centered = false;
    }
    connectedCallback() {
        super.connectedCallback();
        // After the first update...
        this.updateComplete.then(async () => {
            this.setupTabsAndPanels();
            this.setAriaLabels();
            await this.syncTabs();
            if (this.active) {
                const tab = this.tabs.find(t => t.panel === this.active);
                if (tab) {
                    this.setActiveTab(tab);
                }
            }
            else {
                this.setActiveTab(this.getActiveTab() ?? this.tabs[0]);
            }
            this.updateShadows();
            // Re-measure after fonts load — the initial getBoundingClientRect() call
            // may have used a fallback font whose glyph widths differ from the brand
            // font, resulting in an incorrectly-sized clip-path indicator.
            document.fonts.ready.then(() => this.updateIndicator());
        });
    }
    disconnectedCallback() {
        super.disconnectedCallback();
    }
    firstUpdated() {
        this.updateShadows();
        this.nav.addEventListener('scroll', () => this.updateShadows());
        // Also listen to resize if needed
        window.addEventListener('resize', () => this.updateShadows());
    }
    updateShadows() {
        if (!this.nav)
            return;
        const scrollOffset = 5;
        const { scrollLeft, clientWidth, clientHeight, scrollWidth, scrollHeight, scrollTop } = this.nav;
        if (['top', 'bottom'].includes(this.placement)) {
            this.hasStartShadow =
                this.dir === 'rtl'
                    ? scrollWidth - (scrollLeft * -1 + clientWidth) > scrollOffset
                    : scrollLeft > scrollOffset;
            this.hasEndShadow =
                this.dir === 'rtl'
                    ? scrollLeft * -1 > scrollOffset
                    : scrollWidth - scrollOffset > scrollLeft + clientWidth;
        }
        else {
            this.hasStartShadow = scrollTop > scrollOffset;
            this.hasEndShadow =
                scrollHeight - scrollOffset > scrollTop + clientHeight;
        }
    }
    async syncTabs() {
        if (this.tabs.length) {
            await Promise.all(this.tabs.map(async (tab) => {
                await tab.updateComplete;
                tab.appearance = this.appearance;
                tab.placement = this.placement;
                tab.compact = this.compact;
            }));
            /** This is necessary to achieve the sliding active tab effect */
            this.cloneTabs();
        }
    }
    getAllTabs() {
        const slot = this.shadowRoot.querySelector('slot[name="nav"]');
        return [...slot.assignedElements()].filter(el => {
            return el.tagName.toLowerCase() === 'made-tab';
        });
    }
    getAllPanels() {
        return [...this.body.assignedElements()].filter(el => el.tagName.toLowerCase() === 'made-tab-panel');
    }
    getActiveTab() {
        return this.tabs.find(el => el.active);
    }
    handleClick(event) {
        this.firstRender = false;
        const target = event.target;
        const tab = target.closest('made-tab');
        const tabGroup = tab?.closest('made-tabs');
        // Ensure the target tab is in this specific tab group instance
        if (tabGroup !== this) {
            return;
        }
        if (tab !== null) {
            this.setActiveTab(tab);
        }
    }
    async handleKeyDown(event) {
        this.firstRender = false;
        await this.updateComplete;
        const target = event.target;
        const tab = target.closest('made-tab');
        const tabGroup = tab?.closest('made-tabs');
        // Is the in this grpoup?
        if (tabGroup !== this) {
            return;
        }
        // Activate a tab
        if (['Enter', ' '].includes(event.key)) {
            if (tab !== null) {
                this.setActiveTab(tab);
                event.preventDefault();
            }
            return;
        }
        // Navigate between tabs; https://www.w3.org/WAI/ARIA/apg/patterns/tabs/
        if ([
            'ArrowLeft',
            'ArrowRight',
            'ArrowUp',
            'ArrowDown',
            'Home',
            'End'
        ].includes(event.key)) {
            const activeEl = this.tabs.find(t => t.matches(':focus'));
            const isRtl = this.dir === 'rtl';
            let nextTab = null;
            if (activeEl?.tagName.toLowerCase() === 'made-tab') {
                if (event.key === 'Home') {
                    nextTab = this.focusableTabs[0];
                }
                else if (event.key === 'End') {
                    nextTab = this.focusableTabs[this.focusableTabs.length - 1];
                }
                else if ((['top', 'bottom'].includes(this.placement) &&
                    event.key === (isRtl ? 'ArrowRight' : 'ArrowLeft')) ||
                    (['start', 'end'].includes(this.placement) && event.key === 'ArrowUp')) {
                    const currentIndex = this.tabs.findIndex(el => el === activeEl);
                    nextTab = this.findNextFocusableTab(currentIndex, 'backward');
                }
                else if ((['top', 'bottom'].includes(this.placement) &&
                    event.key === (isRtl ? 'ArrowLeft' : 'ArrowRight')) ||
                    (['start', 'end'].includes(this.placement) &&
                        event.key === 'ArrowDown')) {
                    const currentIndex = this.tabs.findIndex(el => el === activeEl);
                    nextTab = this.findNextFocusableTab(currentIndex, 'forward');
                }
                if (!nextTab) {
                    return;
                }
                nextTab.tabIndex = 0;
                nextTab.focus({ preventScroll: true });
                if (this.selectionFollowsFocus) {
                    this.setActiveTab(nextTab);
                }
                else {
                    this.tabs.forEach(tabEl => {
                        tabEl.tabIndex = tabEl === nextTab ? 0 : -1;
                    });
                }
                if (['top', 'bottom'].includes(this.placement)) {
                    scrollIntoView(nextTab, this.nav, 'horizontal');
                }
                else {
                    scrollIntoView(nextTab, this.nav, 'vertical');
                }
                event.preventDefault();
            }
        }
    }
    findNextFocusableTab(currentIndex, direction) {
        let nextTab = null;
        const iterator = direction === 'forward' ? 1 : -1;
        let nextIndex = currentIndex + iterator;
        while (currentIndex < this.tabs.length) {
            nextTab = this.tabs[nextIndex] || null;
            if (nextTab === null) {
                // If we are at the end: loop to start; or if we are at the start: loop to end
                if (direction === 'forward') {
                    nextTab = this.focusableTabs[0];
                }
                else {
                    nextTab = this.focusableTabs[this.focusableTabs.length - 1];
                }
                break;
            }
            if (!nextTab.disabled) {
                break;
            }
            nextIndex += iterator;
        }
        return nextTab;
    }
    setActiveTab(tab) {
        if (tab.closest('made-tabs') !== this) {
            return;
        }
        if (tab !== this.activeTab && !tab.disabled) {
            this.active = tab.panel;
            this.activeTab = tab;
            // Set active states on tabs, and manage tabIndex for focus
            this.tabs.forEach(el => {
                el.active = el === this.activeTab;
                el.tabIndex = el === this.activeTab ? 0 : -1;
            });
            // Set active states on panels
            this.panels.forEach(el => (el.active = el.name === this.activeTab?.panel));
            this.updateIndicator();
            // Fire change event if tab changed by a user action
            if (!this.firstRender) {
                this.dispatchEvent(new CustomEvent('change', {
                    detail: { name: this.activeTab.panel },
                    bubbles: true,
                    composed: true
                }));
            }
        }
    }
    getDimensions(el) {
        const { width, height } = el.getBoundingClientRect();
        return { width, height };
    }
    /**
     * Re-measures the active tab and repositions the clip-path indicator.
     */
    updateIndicator() {
        if (!this.activeTab)
            return;
        const isRtl = this.dir === 'rtl';
        const { width: activeTabWidth, height: activeTabHeight } = this.getDimensions(this.activeTab);
        const { width: activeTabLayerWidth } = this.getDimensions(this.activeTabLayer);
        const allTabs = this.getAllTabs();
        const precedingTabs = allTabs.slice(0, allTabs.indexOf(this.activeTab));
        const offset = precedingTabs.reduce((previous, current) => {
            const { width, height } = this.getDimensions(current);
            return { left: previous.left + width, top: previous.top + height };
        }, { left: 0, top: 0 });
        switch (this.placement) {
            case 'top':
            case 'bottom': {
                const position = isRtl
                    ? activeTabLayerWidth - (offset.left + activeTabWidth)
                    : offset.left;
                this.activeTabLayer.style.setProperty('--indicator-width', `${activeTabWidth}px`);
                this.activeTabLayer.style.setProperty('--indicator-x', `${position}px`);
                break;
            }
            case 'start':
            case 'end':
                this.activeTabLayer.style.setProperty('--indicator-height', `${activeTabHeight}px`);
                this.activeTabLayer.style.setProperty('--indicator-y', `${offset.top}px`);
                break;
        }
    }
    setAriaLabels() {
        this.tabsList.setAttribute('aria-orientation', `${this.placement === 'top' || this.placement === 'bottom' ? 'horizontal' : 'vertical'}`);
        // Link each tab with its panel
        this.tabs.forEach(tab => {
            const panel = this.panels.find(el => el.name === tab.panel);
            if (panel) {
                tab.setAttribute('aria-controls', panel.getAttribute('id'));
                panel.setAttribute('aria-labelledby', tab.getAttribute('id'));
            }
        });
    }
    setupTabsAndPanels() {
        this.tabs = this.getAllTabs();
        this.focusableTabs = this.tabs.filter(el => !el.disabled);
        this.panels = this.getAllPanels();
    }
    // Create a layer of cloned tabs in an active state to achieve the sliding effect
    cloneTabs() {
        const slottedTabs = this.navSlot.assignedElements({ flatten: true });
        this.activeTabLayer.innerHTML = '';
        slottedTabs.forEach(tabElement => {
            const clone = tabElement.cloneNode(true);
            const { width } = this.getDimensions(tabElement);
            clone.tabIndex = -1;
            clone.id = '';
            clone.slot = '';
            clone.setAttribute('data-clone', 'true');
            clone.active = true;
            // Ensure width matches when full-width is enabled
            if (this.fullWidth) {
                clone.style.width = `${width}px`;
            }
            this.activeTabLayer.appendChild(clone);
        });
    }
    async updated(changedProperties) {
        if (changedProperties.has('active')) {
            const tab = this.tabs.find(el => el.panel === this.active);
            if (tab) {
                this.setActiveTab(tab);
            }
        }
        if (['appearance', 'placement', 'compact'].some(watchedProp => changedProperties.has(watchedProp))) {
            this.syncTabs();
        }
        if (changedProperties.has('fullWidth')) {
            const tab = this.tabs.find(el => el.panel === this.active);
            this.syncTabs();
            if (tab) {
                await 0;
                this.setActiveTab(tab);
            }
        }
    }
    render() {
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-tab-group': true,
            'made-c-tab-group--top': this.placement === 'top',
            'made-c-tab-group--bottom': this.placement === 'bottom',
            'made-c-tab-group--start': this.placement === 'start',
            'made-c-tab-group--end': this.placement === 'end',
            [`made-c-tab-group--${this.appearance}`]: true,
            'made-c-tab-group--compact': this.compact,
            'made-c-tab-group--start-shadow': this.hasStartShadow,
            'made-c-tab-group--end-shadow': this.hasEndShadow,
            'made-c-tab-group--full-width': this.fullWidth,
            'made-c-tab-group--centered': this.centered
        })}
        @click=${this.handleClick}
        @keydown=${this.handleKeyDown}
      >
        <div class="made-c-tab-group__nav-container" part="nav">
          <div class="made-c-tab-group__nav">
            <div part="tabs" class="made-c-tab-group__tabs" role="tablist">
              <slot name="nav" @slotchange=${this.setupTabsAndPanels}> </slot>
              <div
                class=${classMap({
            'made-c-tab-group__active-layer': true,
            'no-animation': this.firstRender
        })}
                aria-hidden="true"
              ></div>
            </div>
          </div>
        </div>
        <slot
          part="body"
          class="made-c-tab-group__body"
          @slotchange=${this.setupTabsAndPanels}
        ></slot>
      </div>
    `;
    }
}
MadeTabs.styles = [componentStyles, styles];
export default MadeTabs;
__decorate([
    query('.made-c-tab-group')
], MadeTabs.prototype, "tabGroup", void 0);
__decorate([
    query('.made-c-tab-group__tabs')
], MadeTabs.prototype, "tabsList", void 0);
__decorate([
    query('.made-c-tab-group__body')
], MadeTabs.prototype, "body", void 0);
__decorate([
    query('.made-c-tab-group__nav')
], MadeTabs.prototype, "nav", void 0);
__decorate([
    query('slot[name="nav"]')
], MadeTabs.prototype, "navSlot", void 0);
__decorate([
    query('.made-c-tab-group__active-layer')
], MadeTabs.prototype, "activeTabLayer", void 0);
__decorate([
    state()
], MadeTabs.prototype, "firstRender", void 0);
__decorate([
    state()
], MadeTabs.prototype, "hasStartShadow", void 0);
__decorate([
    state()
], MadeTabs.prototype, "hasEndShadow", void 0);
__decorate([
    property({ reflect: true })
], MadeTabs.prototype, "active", void 0);
__decorate([
    property()
], MadeTabs.prototype, "placement", void 0);
__decorate([
    property()
], MadeTabs.prototype, "appearance", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTabs.prototype, "compact", void 0);
__decorate([
    property({ type: Boolean, attribute: 'selection-follows-focus' })
], MadeTabs.prototype, "selectionFollowsFocus", void 0);
__decorate([
    property({ type: Boolean, attribute: 'full-width' })
], MadeTabs.prototype, "fullWidth", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTabs.prototype, "centered", void 0);
