// @mc-made/web-components/dist/components/tabs/src/tabs.styles.js
import { css } from 'lit';
export default css `
  :host {
    --shadow-color: var(--made-color-text-on-color);

    display: block;
    height: 100%;
  }

  .made-c-tab-group__active-layer {
    --indicator-size: var(--made-size-space-1-x);
    --indicator-x: 0px;
    --indicator-y: 0px;
    --indicator-width: 0px;
    --indicator-height: 0px;

    position: absolute;
    display: flex;
    color: var(--text-active-color);
    background-color: transparent;
    clip-path: inset(
      0px calc(100% - (var(--indicator-x) + var(--indicator-width))) 0px
        var(--indicator-x)
    );

    &:not(.no-animation) {
      transition: clip-path var(--made-motion-duration-medium)
        cubic-bezier(0.5, 0, 0.15, 1);
    }

    .made-c-tab-group--minimal &,
    .made-c-tab-group--contained & {
      clip-path: inset(
        0 calc(100% - (var(--indicator-x) + var(--indicator-width))) 0
          var(--indicator-x) round
          var(--made-component-tabs-contained-tab-border-radius)
      );
    }

    .made-c-tab-group--minimal.made-c-tab-group--start &,
    .made-c-tab-group--minimal.made-c-tab-group--end &,
    .made-c-tab-group--contained.made-c-tab-group--start &,
    .made-c-tab-group--contained.made-c-tab-group--end & {
      clip-path: inset(
        var(--indicator-y) 0
          calc(100% - (var(--indicator-y) + var(--indicator-height))) 0 round
          10px
      );
    }

    .made-c-tab-group--top & {
      top: 0;

      & made-tab[appearance='default']::part(base) {
        border-block-end: var(--indicator-size) solid
          var(--made-component-tabs-default-border-color);

        padding-block-end: calc(
          var(--made-component-tabs-default-item-padding-y) - var(
              --indicator-size
            )
        );
      }
    }

    .made-c-tab-group--bottom:not(
        .made-c-tab-group--minimal,
        .made-c-tab-group--contained
      )
      & {
      top: calc(-1 * var(--made-size-space-0point25-x));
    }

    .made-c-tab-group--start &,
    .made-c-tab-group--end & {
      flex-direction: column;
      clip-path: inset(
        var(--indicator-y) 0
          calc(100% - (var(--indicator-y) + var(--indicator-height))) 0
      );
    }

    .made-c-tab-group--start & {
      & made-tab[appearance='default']::part(base) {
        border-inline-end: var(--indicator-size) solid
          var(--made-component-tabs-default-border-color);

        padding-inline-end: calc(
          var(--made-component-tabs-default-item-padding-x) - var(
              --indicator-size
            )
        );
      }
    }

    .made-c-tab-group--end & {
      & made-tab[appearance='default']::part(base) {
        border-inline-start: var(--indicator-size) solid
          var(--made-component-tabs-default-border-color);

        padding-inline-start: calc(
          var(--made-component-tabs-default-item-padding-x) - var(
              --indicator-size
            )
        );
      }
    }
  }

  .made-c-tab-group {
    --text-active-color: var(--made-color-text-brand);

    display: flex;
    border-radius: 0;

    &.made-c-tab-group--minimal {
      --text-active-color: var(--made-color-text-on-color);
    }
  }

  .made-c-tab-group__tabs {
    display: flex;
    position: relative;
    color: var(--made-color-text-secondary);
  }

  .made-c-tab-group__nav-container {
    position: relative;
  }

  .made-c-tab-group__nav {
    display: flex;
    /* Hide scrollbar in Firefox */
    scrollbar-width: none;
  }

  .made-c-tab-group--start .made-c-tab-group__nav-container,
  .made-c-tab-group--end .made-c-tab-group__nav-container {
    display: flex;
  }

  .made-c-tab-group--start,
  .made-c-tab-group--end {
    height: 100%;

    & .made-c-tab-group__nav {
      flex-direction: column;
      flex-grow: 0;
    }
  }

  .made-c-tab-group__body {
    display: block;
  }

  /*
   * Top
   */

  .made-c-tab-group--top {
    flex-direction: column;
  }

  .made-c-tab-group--top .made-c-tab-group__nav-container {
    order: 1;
  }

  .made-c-tab-group--top .made-c-tab-group__nav {
    overflow-x: auto;
  }

  /* Hide scrollbar in Chrome/Safari */
  .made-c-tab-group--top .nav::-webkit-scrollbar {
    width: 0;
    height: 0;
  }

  .made-c-tab-group--top {
    & .made-c-tab-group__tabs {
      position: relative;
      display: flex;
      flex-direction: row;
      flex: 1 1 auto;
      border-block-end: solid var(--made-size-space-0point25-x)
        var(--made-color-border-default);
    }
  }

  .made-c-tab-group--minimal,
  .made-c-tab-group--contained {
    & .made-c-tab-group__tabs {
      display: flex;
      flex-grow: 0;
      border: none;
      padding: var(--made-component-tabs-contained-padding);
      position: relative;
    }

    & .made-c-tab-group__active-layer {
      top: var(--made-component-tabs-contained-padding);
      color: var(--text-active-color);
      background-color: var(--made-component-tabs-default-border-color);
    }
  }

  .made-c-tab-group--contained {
    & .made-c-tab-group__tabs {
      background-color: var(--made-color-background-surface-secondary);
      border-radius: var(--made-component-tabs-contained-border-radius);

      .made-c-tab-group--start &,
      .made-c-tab-group--end & {
        border-radius: var(--made-border-radius-06);
      }
    }

    & .made-c-tab-group__active-layer {
      background-color: var(
        --made-component-tabs-contained-item-background-selected
      );
    }
  }

  .made-c-tab-group--top .made-c-tab-group__body {
    order: 2;
  }

  /*
   * Bottom
   */

  .made-c-tab-group--bottom {
    flex-direction: column;
  }

  .made-c-tab-group--bottom .made-c-tab-group__nav-container {
    order: 2;
  }

  .made-c-tab-group--bottom .made-c-tab-group__nav {
    overflow-x: auto;
  }

  /* Hide scrollbar in Chrome/Safari */
  .made-c-tab-group--bottom .nav::-webkit-scrollbar {
    width: 0;
    height: 0;
  }

  .made-c-tab-group--bottom:not(
      .made-c-tab-group--minimal,
      .made-c-tab-group--contained
    ) {
    & .made-c-tab-group__tabs {
      flex: 1 1 auto;
      border-block-end: none;
      border-block-start: solid var(--made-size-space-0point25-x)
        var(--made-color-border-default);
    }

    & .made-c-tab-group__nav {
      padding-top: var(--made-size-space-0point25-x);
    }

    &
      .made-c-tab-group__active-layer
      made-tab[appearance='default']::part(base) {
      border-block-end: none;
      border-block-start: var(--indicator-size) solid
        var(--made-component-tabs-default-border-color);

      padding-block-start: calc(
        var(--made-component-tabs-default-item-padding-y) +
          var(--made-size-space-0point25-x) - var(--indicator-size)
      );
    }
  }

  .made-c-tab-group--bottom .made-c-tab-group__body {
    order: 1;
  }

  /*
   * Start
   */

  .made-c-tab-group--start {
    flex-direction: row;
  }

  .made-c-tab-group--start .made-c-tab-group__nav-container {
    order: 1;
  }

  .made-c-tab-group--start .made-c-tab-group__nav {
    overflow-y: auto;
  }

  .made-c-tab-group--start {
    & .made-c-tab-group__tabs {
      flex: 1 1 auto;
      flex-direction: column;
    }

    &:not(.made-c-tab-group--minimal, .made-c-tab-group--contained)
      .made-c-tab-group__tabs {
      border-inline-end: solid var(--made-size-space-0point25-x)
        var(--made-color-border-default);
    }
  }

  .made-c-tab-group--start .made-c-tab-group__body {
    flex: 1 1 auto;
    order: 2;
  }

  /*
   * End
   */

  .made-c-tab-group--end {
    flex-direction: row;
  }

  .made-c-tab-group--end .made-c-tab-group__nav-container {
    order: 2;
  }

  .made-c-tab-group--end .made-c-tab-group__nav {
    overflow-y: auto;
  }

  .made-c-tab-group--end {
    & .made-c-tab-group__tabs {
      flex: 0 0 auto;
      flex-direction: column;
    }

    &:not(.made-c-tab-group--minimal, .made-c-tab-group--contained)
      .made-c-tab-group__tabs {
      border-inline-start: solid var(--made-size-space-0point25-x)
        var(--made-color-border-default);
    }
  }

  .made-c-tab-group--end .made-c-tab-group__body {
    flex: 1 1 auto;
    order: 1;
  }

  /* Scroll shadows */

  .made-c-tab-group--start-shadow .made-c-tab-group__nav-container::before,
  .made-c-tab-group--end-shadow .made-c-tab-group__nav-container::after {
    --shadow-angle: 90deg;
    content: '';
    position: absolute;
    pointer-events: none;
    background: linear-gradient(
      var(--shadow-angle),
      var(--shadow-color) 2.94%,
      rgba(255, 255, 255, 0) 100%
    );
  }

  .made-c-tab-group--start-shadow:is(
      .made-c-tab-group--top,
      .made-c-tab-group--bottom
    )
    .made-c-tab-group__nav-container::before {
    z-index: 1;
    left: 0;
    top: 0;
    bottom: 0;
    width: 50px;
  }

  .made-c-tab-group--end-shadow:is(
      .made-c-tab-group--top,
      .made-c-tab-group--bottom
    )
    .made-c-tab-group__nav-container::after {
    right: 0;
    top: 0;
    bottom: 0;
    width: 50px;
    --shadow-angle: 270deg;
  }

  .made-c-tab-group--start-shadow:is(
      .made-c-tab-group--start,
      .made-c-tab-group--end
    )
    .made-c-tab-group__nav-container::before {
    z-index: 1;
    top: 0;
    left: 0;
    right: 0;
    height: 50px;
    --shadow-angle: 180deg;
  }

  .made-c-tab-group--end-shadow:is(
      .made-c-tab-group--start,
      .made-c-tab-group--end
    )
    .made-c-tab-group__nav-container::after {
    bottom: 0;
    left: 0;
    right: 0;
    height: 50px;
    --shadow-angle: 0deg;
  }

  /* Panel styles */

  .made-c-tab-group--start ::slotted(made-tab-panel),
  .made-c-tab-group--end ::slotted(made-tab-panel) {
    --padding: 0 var(--made-size-space-6-x);
  }
  .made-c-tab-group--top ::slotted(made-tab-panel),
  .made-c-tab-group--bottom ::slotted(made-tab-panel) {
    --padding: var(--made-size-space-6-x) 0;
  }

  /* Full width made-c-tab-group__tabs */

  .made-c-tab-group--full-width .made-c-tab-group__tabs {
    width: 100%;
  }

  .made-c-tab-group--full-width ::slotted(made-tab) {
    flex: 1;
  }

  /* Centered made-c-tab-group__tabs */

  .made-c-tab-group--centered .made-c-tab-group__tabs {
    justify-content: center;
  }
`;
