// @mc-made/web-components/dist/components/theme-provider/src/theme-provider.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { property, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './theme-provider.styles.js';
/**
 * @summary Applies a Made theme by using CSS custom properties to set styles on all components in its scope.
 * @documentation '../stories/theme-provider.mdx'
 *
 * @fires mde-change {CustomEvent} - Emitted when the theme changes.
 * @fires mde-error {CustomEvent} - Emitted when the theme doesn't load.
 *
 * @slot - The default slot.
 */
class MadeThemeProvider extends MadeElement {
    constructor() {
        super(...arguments);
        this.systemColorPreference = MadeThemeProvider.colorSchemeQuery
            .matches
            ? 'dark'
            : 'light';
        /** Tracks the last [theme, resolvedColorMode] applied to avoid redundant style recalculations. */
        this.lastApplied = '';
        /** Name of the theme to apply. */
        this.theme = 'mastercard';
        /** Color mode to apply. Defaults to 'auto' which applies system preference. */
        this.colorMode = 'auto';
    }
    static onColorSchemeChange() {
        for (const instance of MadeThemeProvider.connectedInstances) {
            instance.systemColorPreference = MadeThemeProvider.colorSchemeQuery
                .matches
                ? 'dark'
                : 'light';
            void instance.adoptStyles();
        }
    }
    updated(changedProperties) {
        if (changedProperties.has('theme') || changedProperties.has('colorMode')) {
            void this.adoptStyles();
        }
    }
    connectedCallback() {
        super.connectedCallback();
        MadeThemeProvider.connectedInstances.add(this);
        if (!MadeThemeProvider.colorSchemeListenerAttached) {
            MadeThemeProvider.colorSchemeQuery.addEventListener('change', MadeThemeProvider.onColorSchemeChange);
            MadeThemeProvider.colorSchemeListenerAttached = true;
        }
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        MadeThemeProvider.connectedInstances.delete(this);
        if (MadeThemeProvider.connectedInstances.size === 0) {
            MadeThemeProvider.colorSchemeQuery.removeEventListener('change', MadeThemeProvider.onColorSchemeChange);
            MadeThemeProvider.colorSchemeListenerAttached = false;
        }
    }
    static async loadBuiltIn(name) {
        switch (name) {
            case 'unify':
                await import(`../theme-unify.js`);
                break;
            case 'mastercard':
                await import(`../theme-mastercard.js`);
                break;
        }
    }
    async adoptStyles() {
        // If the theme is already registered (built-in or custom), apply it directly.
        if (MadeThemeProvider.themes.has(this.theme)) {
            // If this theme extends a built-in, ensure the base is loaded first.
            const base = MadeThemeProvider.themeBases.get(this.theme);
            if (base && !MadeThemeProvider.themes.has(base)) {
                await MadeThemeProvider.loadBuiltIn(base);
            }
            this.applyStyles();
            return;
        }
        // For built-in themes, lazy-load the stylesheet which registers itself,
        // then apply.
        if (MadeThemeProvider.builtInThemes.includes(this.theme)) {
            await MadeThemeProvider.loadBuiltIn(this.theme);
            this.applyStyles();
            return;
        }
        // Unknown theme — not built-in and not registered via createTheme().
        this.emit('mde-error', {
            detail: `Unknown theme "${this.theme}". Register it first with MadeThemeProvider.createTheme().`
        });
    }
    applyStyles() {
        try {
            const resolvedMode = this.colorMode === 'auto'
                ? this.systemColorPreference
                : this.colorMode;
            // Skip if the effective theme + mode combination hasn't changed.
            const cacheKey = `${this.theme}::${resolvedMode}`;
            if (cacheKey === this.lastApplied)
                return;
            this.lastApplied = cacheKey;
            const sheets = [];
            // Prepend the base theme's sheet when this theme extends a built-in.
            const base = MadeThemeProvider.themeBases.get(this.theme);
            if (base) {
                const baseSheet = MadeThemeProvider.themes.get(base)?.get(resolvedMode);
                if (baseSheet)
                    sheets.push(baseSheet);
            }
            const overrideSheet = MadeThemeProvider.themes
                .get(this.theme)
                ?.get(resolvedMode);
            if (overrideSheet)
                sheets.push(overrideSheet);
            this.shadowRoot.adoptedStyleSheets = sheets;
            this.emit('mde-change');
        }
        catch (error) {
            this.emit('mde-error', {
                detail: `Could not apply theme: ${String(error)}`
            });
        }
    }
    /**
     * @internal Used by built-in theme files to register themselves on import.
     * Consumers should use the standalone {@link createTheme} function instead.
     */
    static registerTheme(name, colorMode, themeStyles) {
        const themeType = new Map();
        themeType.set(colorMode, themeStyles);
        if (MadeThemeProvider.themes.has(name)) {
            MadeThemeProvider.themes.set(name, new Map([...themeType, ...MadeThemeProvider.themes.get(name)]));
        }
        else {
            MadeThemeProvider.themes.set(name, themeType);
        }
    }
    /**
     * @internal Records that a custom theme extends a built-in theme.
     * Called by the standalone {@link createTheme} function.
     */
    static setThemeBase(name, base) {
        MadeThemeProvider.themeBases.set(name, base);
    }
    /**
     * @internal Injects `@font-face` CSS into `document.adoptedStyleSheets`
     * for the given theme name, once per name.
     * Called by the standalone {@link createTheme} function.
     */
    static injectFonts(themeName, css) {
        if (MadeThemeProvider.fontRegistry.has(themeName))
            return;
        const sheet = new CSSStyleSheet();
        sheet.replaceSync(css);
        document.adoptedStyleSheets = [...document.adoptedStyleSheets, sheet];
        MadeThemeProvider.fontRegistry.add(themeName);
    }
    render() {
        return html `<slot></slot> `;
    }
}
MadeThemeProvider.styles = [componentStyles, styles];
MadeThemeProvider.themes = new Map();
/** Tracks which custom themes are based on a built-in theme (populated via {@link setThemeBase}). */
MadeThemeProvider.themeBases = new Map();
/** Tracks which theme names have already had their fonts injected (populated via {@link injectFonts}). */
MadeThemeProvider.fontRegistry = new Set();
// Single shared MQL + listener for all instances (fixes per-instance listener leak).
MadeThemeProvider.colorSchemeQuery = window.matchMedia('(prefers-color-scheme: dark)');
MadeThemeProvider.connectedInstances = new Set();
MadeThemeProvider.colorSchemeListenerAttached = false;
MadeThemeProvider.builtInThemes = [
    'mastercard',
    'unify'
];
export default MadeThemeProvider;
__decorate([
    state()
], MadeThemeProvider.prototype, "systemColorPreference", void 0);
__decorate([
    property({ reflect: true })
], MadeThemeProvider.prototype, "theme", void 0);
__decorate([
    property({ reflect: true, attribute: 'color-mode' })
], MadeThemeProvider.prototype, "colorMode", void 0);
