// @mc-made/web-components/dist/components/checkbox/src/checkbox.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { CheckboxValidator } from '../../../internal/face/validators/checkbox-validator.js';
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormState, getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { mixinOnReportValidity, onReportValidity } from '../../../internal/face/on-report-validity.js';
import { property, query, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/src/icon.component.js';
import styles from './checkbox.styles.js';
/**
 * @summary The Checkbox gives users a clear way to make selections, such as agreeing to
 * terms or choosing multiple items from a list.
 * @documentation '../stories/checkbox.mdx'
 *
 * @dependency made-icon
 *
 * @fires change {Event} - Fired when an alteration to the control's value is committed by the user.
 * @fires input {Event} - Fired when the control receives input.
 * @fires invalid {Event} - Fired when the control has been checked for validity and doesn't satisfy its constraints.
 *
 * @slot - The checkbox's label. Alternatively, you can use the `label` attribute.
 * @slot supporting-text - Text that describes how to use the checkbox. Alternatively, you can use the `supporting-text` attribute.
 *
 * @csspart form-control - The outer wrapper.
 * @csspart input-container - The container that wraps the control, label and supporting text.
 * @csspart control - Checkbox control.
 * @csspart checked-icon - The checked icon element.
 * @csspart checked-path - The svg path of the checked icon.
 * @csspart indeterminate-icon - The indeterminate icon element.
 * @csspart label - The label's wrapper.
 * @csspart supporting-text - The supporting text's wrapper.
 */
const checkboxBaseClass = mixinOnReportValidity(mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement))));
class MadeCheckbox extends checkboxBaseClass {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, 'supporting-text');
        /** Constraint validation error state controlled by `reportValidity()` */
        this.hasNativeError = false;
        /**
         * The validation message displayed from a native error via
         * `reportValidity()`.
         */
        this.nativeErrorText = '';
        /** The checkbox's label. If you need to display HTML, use the `label` slot instead. */
        this.label = '';
        /** The checkbox's supporting text. If you need to display HTML, use the `supporting-text` slot instead. */
        this.supportingText = '';
        /** The current value of the checkbox, submitted as a name/value pair with form data. */
        this.value = 'on';
        /** The checkbox's size. */
        this.size = 'medium';
        /** Toggle the checkbox. */
        this.checked = false;
        /** Set the checkbox to indeterminate state. */
        this.indeterminate = false;
        /** Makes the checkbox readonly. */
        this.readonly = false;
        /** Makes the checkbox a required field. */
        this.required = false;
        /** Use the appearance attribute to change the visual style of the checkbox. */
        this.appearance = 'neutral';
        /**
         * Gets or sets whether or not the text field is in an invalid state.
         *
         * This error state overrides the error state controlled by
         * `reportValidity()`.
         */
        this.error = false;
    }
    get hasError() {
        return this.error || this.hasNativeError;
    }
    handleChange() {
        if (this.readonly)
            return;
        // dispatch 'composed' change event
        this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
    }
    handleInput(event) {
        if (this.readonly)
            return;
        const target = event.target;
        this.checked = target.checked;
        this.indeterminate = target.indeterminate;
        if (this.hasError) {
            this.reportValidity();
        }
    }
    /** Simulates a click on the checkbox. */
    click() {
        this.checkbox.click();
    }
    /** Sets focus on the checkbox. */
    focus(options) {
        this.checkbox.focus(options);
    }
    /** Removes focus from the checkbox. */
    blur() {
        this.checkbox.blur();
    }
    get validationText() {
        if (this.hasError && this.getErrorText()) {
            return this.getErrorText();
        }
        return null;
    }
    get supportingOrValidationText() {
        return this.validationText ? this.validationText : this.supportingText;
    }
    renderSupportingText() {
        if (!this.supportingOrValidationText) {
            return null;
        }
        return html `
      ${this.hasError
            ? html `<span class="made-c-form-control__error-icon">
            <made-icon
              name="exclamation-triangle-solid-12px"
              library="system"
            ></made-icon>
          </span>`
            : nothing}
      <span>${this.supportingOrValidationText}</span>
    `;
    }
    render() {
        const hasLabelTextSlot = this.hasSlotController.test('[default]');
        const hasSupportingTextSlot = this.hasSlotController.test('supporting-text');
        const hasLabel = this.label ? true : !!hasLabelTextSlot;
        const supportingText = this.renderSupportingText();
        const hasSupportingText = supportingText ? true : !!hasSupportingTextSlot;
        return html `
      <div
        part="form-control"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control--medium': this.size === 'medium',
            'made-c-form-control--large': this.size === 'large',
            'made-c-form-control--has-label': hasLabel,
            'made-c-form-control--has-supporting-text': hasSupportingText,
            'made-c-form-control--readonly': this.readonly,
            'made-c-form-control--disabled': this.disabled,
            'made-c-form-control--error': this.hasError
        })}
      >
        <label
          part="input-container"
          class=${classMap({
            'made-c-checkbox': true,
            'made-c-checkbox--error': this.hasError,
            'made-c-checkbox--medium': this.size === 'medium',
            'made-c-checkbox--large': this.size === 'large',
            'made-c-checkbox--checked': this.checked,
            'made-c-checkbox--indeterminate': this.indeterminate,
            'made-c-checkbox--readonly': this.readonly,
            'made-c-checkbox--disabled': this.disabled,
            'made-c-checkbox--neutral': this.appearance === 'neutral',
            'made-c-checkbox--branded': this.appearance === 'branded'
        })}
        >
          <input
            id="checkbox"
            class="made-c-checkbox__input"
            type="checkbox"
            name=${this.name}
            tabindex=${'0' /* This ensures the Checkbox is keyboard navigable in Safari */}
            ?disabled=${this.disabled}
            ?readonly=${this.readonly}
            ?required=${this.required}
            value=${this.value}
            .checked=${this.checked}
            .indeterminate=${this.indeterminate}
            aria-checked=${this.indeterminate
            ? 'mixed'
            : this.checked
                ? 'true'
                : 'false'}
            @change=${this.handleChange}
            @input=${this.handleInput}
          />
          <span part="control" class="made-c-checkbox__control">
            <made-icon
              class="made-c-checkbox__checked-icon"
              name="check-outline"
              library="system"
              part="checked-icon"
              exportparts="checked-path"
            ></made-icon>
            <made-icon
              class="made-c-checkbox__indeterminate-icon"
              name="minus-outline"
              library="system"
              part="indeterminate-icon"
            ></made-icon>
          </span>
          <div class="made-c-checkbox__label-wrapper">
            <div part="label" class="made-c-checkbox__label">
              <slot>${this.label}</slot>
            </div>
            <div
              id="supporting-text"
              part="supporting-text"
              class="made-c-form-control__supporting-text"
              aria-hidden=${supportingText ? 'false' : 'true'}
            >
              <slot name="supporting-text">${supportingText}</slot>
            </div>
          </div>
        </label>
      </div>
    `;
    }
    /**
     * Reset the text field to its default value.
     */
    reset() {
        // The checked property does not reflect, so the original attribute set by
        // the user is used to determine the default value.
        this.checked = this.hasAttribute('checked');
        this.hasNativeError = false;
        this.nativeErrorText = '';
    }
    getErrorText() {
        return this.errorText || this.nativeErrorText;
    }
    [getFormValue]() {
        if (!this.checked || this.indeterminate) {
            return null;
        }
        return this.value;
    }
    [getFormState]() {
        return String(this.checked);
    }
    formResetCallback() {
        this.reset();
    }
    formStateRestoreCallback(state) {
        this.checked = state === 'true';
    }
    // @ts-expect-error 'ts(2425)'
    [createValidator]() {
        return new CheckboxValidator(() => this);
    }
    // @ts-expect-error 'ts(2425)'
    [getValidityAnchor]() {
        return this.checkbox;
    }
    // @ts-expect-error 'ts(2425)'
    [onReportValidity](invalidEvent) {
        // Prevent default pop-up behavior.
        invalidEvent?.preventDefault();
        this.hasNativeError = !!invalidEvent;
        this.nativeErrorText = this.validationMessage;
    }
}
MadeCheckbox.styles = [componentStyles, formControlStyles, styles];
MadeCheckbox.dependencies = { 'made-icon': MadeIcon };
export default MadeCheckbox;
__decorate([
    query('input[type="checkbox"]')
], MadeCheckbox.prototype, "checkbox", void 0);
__decorate([
    state()
], MadeCheckbox.prototype, "hasNativeError", void 0);
__decorate([
    state()
], MadeCheckbox.prototype, "nativeErrorText", void 0);
__decorate([
    property()
], MadeCheckbox.prototype, "label", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeCheckbox.prototype, "supportingText", void 0);
__decorate([
    property()
], MadeCheckbox.prototype, "value", void 0);
__decorate([
    property({ reflect: true })
], MadeCheckbox.prototype, "size", void 0);
__decorate([
    property({ type: Boolean })
], MadeCheckbox.prototype, "checked", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeCheckbox.prototype, "indeterminate", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeCheckbox.prototype, "readonly", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeCheckbox.prototype, "required", void 0);
__decorate([
    property()
], MadeCheckbox.prototype, "appearance", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeCheckbox.prototype, "error", void 0);
__decorate([
    property({ attribute: 'error-text' })
], MadeCheckbox.prototype, "errorText", void 0);
