// @mc-made/web-components/dist/components/drawer/src/drawer.styles.js
import { css } from 'lit';
export default css `
  :host {
    --transition-duration: var(--made-motion-duration-long);
    --padding: 1rem;

    display: contents;
  }

  .made-c-drawer {
    position: fixed;
    inset: 0;
    width: 100%;
    height: 100%;
    max-width: none;
    max-height: none;
    margin: 0;
    padding: 0;
    background: transparent;
    border: none;
    overflow: visible;
  }

  .made-c-drawer:focus {
    outline: none;
  }

  .made-c-drawer::backdrop {
    background-color: color-mix(
      in srgb,
      var(--made-color-background-backdrop) 40%,
      transparent
    );
    opacity: 0;
    transition: opacity var(--transition-duration)
      var(--made-motion-easing-emphasised-ease-out);
  }

  .made-c-drawer__panel--open ~ dialog::backdrop,
  .made-c-drawer:has(.made-c-drawer__panel--open)::backdrop {
    opacity: 1;
  }

  .made-c-drawer__panel {
    position: fixed;
    bottom: 0;
    left: 0;
    right: 0;
    width: 100%;
    min-height: 25dvh;
    max-height: 80dvh;
    display: flex;
    flex-direction: column;
    padding: var(--made-component-drawer-picker-padding-y)
      var(--made-component-drawer-picker-padding-x);
    background-color: var(--made-color-background-surface-default);
    font-family: var(--made-font-family-primary);
    color: var(--made-color-text-primary);
    overflow: hidden;
    box-shadow: var(--made-component-modal-box-shadow);
    border-start-start-radius: var(--made-component-drawer-border-radius);
    border-start-end-radius: var(--made-component-drawer-border-radius);
    border-block-start: 1px solid var(--made-component-drawer-border-color);
    transition: transform var(--transition-duration)
      var(--made-motion-easing-emphasised-ease-out);
    will-change: transform;
    z-index: 1;
    transform: translateY(100%);
    cursor: grab;
    transform-origin: bottom center;
    --_behind-y: 0px;
    --_behind-scale: 1;

    &.made-c-drawer__panel--no-dismiss {
      cursor: default;
    }
  }

  .made-c-drawer__panel--open {
    transform: translateY(var(--_behind-y)) scale(var(--_behind-scale));
  }

  .made-c-drawer__panel--close {
    transform: translateY(100%);
  }

  .made-c-drawer__body {
    flex: 1 1 auto;
    overflow: auto;
    padding: var(--padding);
    -webkit-overflow-scrolling: touch;
  }

  .made-c-drawer__swipe-handle {
    flex: 0 0 auto;
    display: flex;
    justify-content: center;
    margin-block-start: -6px;
  }

  .made-c-drawer__swipe-handle-bar {
    display: block;
    width: 100px;
    height: 6px;
    border-radius: var(--made-border-radius-full);
    color: var(--made-color-border-strong);
  }

  .made-c-drawer__swipe-handle {
    & > button {
      appearance: none;
      background: none;
      border: none;
      padding: 0;
      margin: 0;
      cursor: inherit;
      line-height: 0;

      &:focus {
        outline: none;
      }

      &:focus-visible {
        outline: 2px solid var(--made-color-border-focused);
        outline-offset: 2px;
      }
    }
  }

  .made-c-drawer__panel--dragging {
    transition: none !important;
    transform: var(--_drag-transform) !important;
    cursor: grabbing !important;
  }

  :host(.made-c-drawer--behind) .made-c-drawer__panel {
    --_behind-scale: 0.97;
    /* --_behind-y is set as an inline style computed from actual panel height */
  }
`;
