// @mc-made/web-components/dist/components/drawer/src/drawer.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './drawer.styles.js';
/**
 * @summary Drawers slide up from the bottom of the screen to present secondary
 * content or actions, with optional swipe-to-dismiss gestures.
 * @documentation '../stories/drawer.mdx'
 *
 * @fires mde-show - Emitted when the drawer opens (before the open animation completes).
 * @fires mde-hide - Emitted after the drawer finishes closing.
 * @fires mde-request-close - Emitted when the user attempts to close the drawer
 *   (backdrop click, Escape key, or swipe). Call `event.preventDefault()` to cancel.
 *   The event detail is `{ source: 'backdrop' | 'escape' | 'swipe' }`.
 *
 * @slot - The drawer's main content.
 *
 * @csspart base - The `<dialog>` element.
 * @csspart panel - The inner panel that slides up from the bottom.
 * @csspart body - The scrollable content region inside the panel.
 *
 * @dependency made-icon
 */
class MadeDrawer extends MadeElement {
    constructor() {
        super(...arguments);
        /** Whether the drawer is open. */
        this.open = false;
        /**
         * When set, the drawer can only be closed by an action inside it.
         * Backdrop clicks, Escape key, and swipe gestures are ignored.
         */
        this.noDismiss = false;
        // ─── Swipe state ─────────────────────────────────────────────────────────
        this._swipeStart = null;
        /** Panel's computed translateX/Y at the moment a drag gesture begins. */
        this._dragBaseTransform = null;
        this._isDragging = false;
        /** Deferred pointer capture: set on pointerdown, applied only when drag is confirmed. */
        this._pendingCapture = null;
        // ─── Event handlers ──────────────────────────────────────────────────────
        this._handleCancel = (event) => {
            event.preventDefault();
            this._requestClose('escape');
        };
        this._onDocTouchStart = (e) => {
            // Only the topmost open drawer handles gestures.
            if (MadeDrawer._topDrawer !== this || this.noDismiss)
                return;
            if (!e.composedPath().includes(this.panel))
                return;
            const t = e.touches[0];
            this._swipeStart = { x: t.clientX, y: t.clientY };
            const matrix = new DOMMatrixReadOnly(getComputedStyle(this.panel).transform);
            this._dragBaseTransform = { x: matrix.m41, y: matrix.m42 };
        };
        // ─── Pointer-device swipe (mouse / pen) ──────────────────────────────────
        //
        // Mirrors the touch handlers but uses PointerEvents with setPointerCapture
        // so the gesture is tracked without document-level listeners. Skipped for
        // pointerType === 'touch' because the touch handlers already cover that.
        this._onPanelPointerDown = (e) => {
            if (MadeDrawer._topDrawer !== this ||
                e.pointerType === 'touch' ||
                this.noDismiss)
                return;
            this._swipeStart = { x: e.clientX, y: e.clientY };
            const matrix = new DOMMatrixReadOnly(getComputedStyle(this.panel).transform);
            this._dragBaseTransform = { x: matrix.m41, y: matrix.m42 };
            // Defer pointer capture until the drag gesture is confirmed (dead-zone
            // crossed in _applyDragTransform). Capturing eagerly redirects the
            // synthesised click event away from child elements such as close buttons.
            this._pendingCapture = {
                element: e.currentTarget,
                pointerId: e.pointerId
            };
        };
        this._onPanelPointerMove = (e) => {
            if (!this._swipeStart || e.pointerType === 'touch')
                return;
            const dx = e.clientX - this._swipeStart.x;
            const dy = e.clientY - this._swipeStart.y;
            if (this._applyDragTransform(dx, dy))
                e.preventDefault();
        };
        this._onPanelPointerUp = (e) => {
            if (!this._swipeStart || e.pointerType === 'touch')
                return;
            const dx = e.clientX - this._swipeStart.x;
            const dy = e.clientY - this._swipeStart.y;
            this._finishDrag(dx, dy);
        };
        this._onDocTouchMove = (e) => {
            if (!this._swipeStart)
                return;
            const touch = e.touches[0];
            const dx = touch.clientX - this._swipeStart.x;
            const dy = touch.clientY - this._swipeStart.y;
            if (this._applyDragTransform(dx, dy))
                e.preventDefault();
        };
        this._onDocTouchEnd = (e) => {
            if (!this._swipeStart)
                return;
            const touch = e.changedTouches[0];
            const dx = touch.clientX - this._swipeStart.x;
            const dy = touch.clientY - this._swipeStart.y;
            this._finishDrag(dx, dy);
        };
    }
    static get _topDrawer() {
        return MadeDrawer._openStack[MadeDrawer._openStack.length - 1];
    }
    _registerOpen() {
        const prevTop = MadeDrawer._topDrawer;
        if (!MadeDrawer._openStack.includes(this)) {
            MadeDrawer._openStack.push(this);
        }
        if (prevTop) {
            // Compute the lift needed to expose the drag handle above the front drawer.
            // With transform-origin: bottom center, scale(0.97) drops the top edge by
            // panelHeight * 0.03. Add 1.25rem (~20px) so the handle clears the front panel.
            const behindY = -(prevTop.panel.offsetHeight * (1 - 0.97) + 20);
            prevTop.panel.style.setProperty('--_behind-y', `${behindY}px`);
            prevTop.classList.add('made-c-drawer--behind');
        }
        document.body.style.overflow = 'hidden';
    }
    _registerClose() {
        const idx = MadeDrawer._openStack.indexOf(this);
        if (idx !== -1)
            MadeDrawer._openStack.splice(idx, 1);
        if (MadeDrawer._openStack.length === 0) {
            document.body.style.overflow = '';
        }
        MadeDrawer._topDrawer?.classList.remove('made-c-drawer--behind');
    }
    // Removes the --behind class from the drawer below so its forward animation
    // starts in tandem with this drawer's close animation.
    _beginClose() {
        const idx = MadeDrawer._openStack.indexOf(this);
        const below = idx > 0 ? MadeDrawer._openStack[idx - 1] : undefined;
        if (below) {
            below.classList.remove('made-c-drawer--behind');
            below.panel.style.removeProperty('--_behind-y');
        }
    }
    // Pops this drawer from the stack and releases scroll-lock if empty.
    _finishClose() {
        const idx = MadeDrawer._openStack.indexOf(this);
        if (idx !== -1)
            MadeDrawer._openStack.splice(idx, 1);
        if (MadeDrawer._openStack.length === 0) {
            document.body.style.overflow = '';
        }
        // Clear any height pinned when this drawer opened as a nested drawer.
        this.panel.style.minHeight = '';
        this.panel.style.maxHeight = '';
    }
    // ─── Lifecycle ───────────────────────────────────────────────────────────
    firstUpdated() {
        if (this.open) {
            this.dialog.showModal();
            this._animateIn();
            this._registerOpen();
        }
        this._setupSwipe();
    }
    updated(changedProperties) {
        if (changedProperties.has('open')) {
            if (this.open && !this.dialog.open) {
                this._registerOpen();
                this.dialog.showModal();
                // Double rAF: the first frame lets the browser paint the panel at its
                // initial off-screen transform; the second adds --open to start the
                // CSS transition. This is required when showing from display:none.
                requestAnimationFrame(() => {
                    // Pin nested drawers to match the drawer directly below them.
                    // Reading offsetHeight here (post-showModal, pre-paint) ensures the
                    // panel below is fully laid out before we sample it.
                    // Uses [idx-1] rather than [0] to avoid walking back to stale entries
                    // that may have been left in the stack if a previous story render did
                    // not trigger disconnectedCallback (e.g. Storybook HMR).
                    const idx = MadeDrawer._openStack.indexOf(this);
                    if (idx > 0) {
                        const belowPanel = MadeDrawer._openStack[0]?.panel;
                        const h = belowPanel?.offsetHeight ?? 0;
                        if (h > 0) {
                            this.panel.style.minHeight = `${h}px`;
                            this.panel.style.maxHeight = `${h}px`;
                        }
                    }
                    requestAnimationFrame(() => this._animateIn());
                });
                this.emit('mde-show');
            }
            else if (!this.open && this.dialog.open) {
                this._beginClose();
                this._animateOut(() => {
                    this.dialog.close();
                    this._finishClose();
                    this.emit('mde-hide');
                });
            }
        }
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        document.removeEventListener('touchstart', this._onDocTouchStart);
        document.removeEventListener('touchmove', this._onDocTouchMove);
        document.removeEventListener('touchend', this._onDocTouchEnd);
        this._registerClose();
    }
    // ─── Animation helpers ───────────────────────────────────────────────────
    _animateIn() {
        this.panel.classList.add('made-c-drawer__panel--open');
        this.panel.classList.remove('made-c-drawer__panel--close');
    }
    _animateOut(onDone) {
        this.panel.classList.add('made-c-drawer__panel--close');
        this.panel.classList.remove('made-c-drawer__panel--open');
        let done = false;
        const finish = () => {
            if (done)
                return;
            done = true;
            this.panel.classList.remove('made-c-drawer__panel--close');
            onDone();
        };
        this.panel.addEventListener('transitionend', finish, { once: true });
        // Fallback: when the panel is already at the close position (e.g. after a
        // full drag-to-edge), "before" and "after" transforms are identical so no
        // CSS transition starts and transitionend never fires.
        //
        // We detect this via transitionrun. rAF callbacks run BEFORE style
        // recalculation, so transitionrun fires after any rAF-based check. A
        // 50ms timeout is sufficient — transitionrun fires within one paint frame
        // (~16ms) if a transition is going to start, so after 50ms we know for
        // certain whether one started or not.
        let transitionStarted = false;
        const onTransitionRun = () => {
            transitionStarted = true;
        };
        this.panel.addEventListener('transitionrun', onTransitionRun, {
            once: true
        });
        setTimeout(() => {
            this.panel.removeEventListener('transitionrun', onTransitionRun);
            if (!transitionStarted) {
                this.panel.removeEventListener('transitionend', finish);
                finish();
            }
        }, 50);
    }
    // ─── Close request ───────────────────────────────────────────────────────
    _requestClose(source) {
        if (this.noDismiss)
            return;
        const event = this.emit('mde-request-close', {
            cancelable: true,
            detail: { source }
        });
        if (!event.defaultPrevented) {
            this.open = false;
        }
    }
    _handlePointerDown(event) {
        if (MadeDrawer._topDrawer !== this)
            return;
        if (event.target === this.dialog) {
            this._requestClose('backdrop');
        }
    }
    // ─── Swipe support ───────────────────────────────────────────────────────
    _setupSwipe() {
        document.addEventListener('touchstart', this._onDocTouchStart, {
            passive: true
        });
        document.addEventListener('touchmove', this._onDocTouchMove, {
            passive: false
        });
        document.addEventListener('touchend', this._onDocTouchEnd, {
            passive: true
        });
    }
    // ─── Live drag helpers ───────────────────────────────────────────────────
    /**
     * Applies a real-time drag transform while the gesture is in progress.
     * The drawer is always at the bottom, so only the Y axis is relevant.
     * Dragging down (positive dy) moves toward the close position.
     *
     * Returns `true` when the gesture is on-axis so callers can call
     * `e.preventDefault()` to suppress native scroll.
     */
    _applyDragTransform(_dx, dy) {
        if (!this._dragBaseTransform)
            return false;
        const { y: by } = this._dragBaseTransform;
        const h = this.panel.offsetHeight;
        const DZ = 4; // dead-zone px before drag is considered intentional
        // Clamp: 0 (fully open) ↔ h (fully off-screen at bottom)
        const t = `translateY(${Math.max(0, Math.min(by + dy, h))}px)`;
        const onAxis = this.open ? dy > DZ : dy < -DZ;
        if (onAxis && !this._isDragging) {
            this._isDragging = true;
            this.panel.classList.add('made-c-drawer__panel--dragging');
            // Confirm pointer capture now that a real drag has started.
            if (this._pendingCapture) {
                this._pendingCapture.element.setPointerCapture(this._pendingCapture.pointerId);
                this._pendingCapture = null;
            }
        }
        if (this._isDragging) {
            this.panel.style.setProperty('--_drag-transform', t);
        }
        return onAxis;
    }
    /**
     * Called on touch-end / pointer-up.  Decides whether to complete the
     * gesture (snap open / close) or snap back, then hands control back to the
     * CSS transition system by removing `--dragging`.
     */
    _finishDrag(_dx, dy) {
        this._swipeStart = null;
        this._dragBaseTransform = null;
        this._pendingCapture = null;
        const T = 60; // minimum swipe distance to trigger open/close
        const shouldClose = this.open && dy > T;
        // Remove drag override and re-enable CSS transitions in one tick so the
        // browser animates from the current drag position to the snap target.
        this.panel.style.removeProperty('--_drag-transform');
        this.panel.classList.remove('made-c-drawer__panel--dragging');
        this._isDragging = false;
        if (shouldClose) {
            if (this.noDismiss)
                return; // CSS snaps back automatically
            const event = this.emit('mde-request-close', {
                cancelable: true,
                detail: { source: 'swipe' }
            });
            if (event.defaultPrevented)
                return; // CSS snaps back automatically
            // _animateOut adds --close, removes --open → CSS transitions from drag
            // position to close/peek position.  Set _skipNextAnimation before the
            // property change so updated() doesn't start a second animation.
            this._beginClose();
            this._animateOut(() => {
                this.open = false;
                this.dialog.close();
                this._finishClose();
                this.emit('mde-hide');
            });
        }
        // Snap-back: no state change.  Removing --dragging re-enables the
        // transition which animates back to the CSS resting position automatically.
    }
    // ─── Render ──────────────────────────────────────────────────────────────
    render() {
        return html `
      <dialog
        part="base"
        class="made-c-drawer"
        aria-modal="true"
        @cancel=${this._handleCancel}
        @pointerdown=${this._handlePointerDown}
      >
        <div
          part="panel"
          class=${classMap({
            'made-c-drawer__panel': true,
            'made-c-drawer__panel--no-dismiss': this.noDismiss
        })}
          @pointerdown=${this._onPanelPointerDown}
          @pointermove=${this._onPanelPointerMove}
          @pointerup=${this._onPanelPointerUp}
        >
          <div class="made-c-drawer__swipe-handle">
            <made-icon
              name="bar"
              library="system"
              class="made-c-drawer__swipe-handle-bar"
            ></made-icon>
          </div>
          <div part="body" class="made-c-drawer__body">
            <slot></slot>
          </div>
        </div>
      </dialog>
    `;
    }
}
MadeDrawer.styles = [componentStyles, styles];
// ─── Open drawer registry ────────────────────────────────────────────────
// Tracks open instances in insertion order so that:
//  • actions (backdrop, Escape, swipe) always target the topmost drawer
//  • scroll-lock survives nested open/close cycles
MadeDrawer._openStack = [];
export default MadeDrawer;
__decorate([
    query('dialog')
], MadeDrawer.prototype, "dialog", void 0);
__decorate([
    query('.made-c-drawer__panel')
], MadeDrawer.prototype, "panel", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeDrawer.prototype, "open", void 0);
__decorate([
    property({ attribute: 'no-dismiss', type: Boolean, reflect: true })
], MadeDrawer.prototype, "noDismiss", void 0);
