// @mc-made/web-components/dist/components/slider/src/slider.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import '../../input/made-input.js';
import { classMap } from 'lit/directives/class-map.js';
import { getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { live } from 'lit/directives/live.js';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { property, query, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './slider.styles.js';
const rangeBaseClass = mixinFormAssociated(mixinElementInternals(MadeElement));
/**
 * @summary Sliders let users select a numeric value within a defined range by
 * dragging a thumb along a horizontal track.
 * @documentation '../stories/slider.mdx'
 *
 * @fires input {Event} - Fired continuously while the user drags the thumb or types in the value input.
 * @fires change {Event} - Fired when the value is committed — on thumb release or when the value input loses focus.
 *
 * @slot label - The slider's label. Use this slot when the label needs to contain HTML.
 * @slot min - Replaces the default minimum value label. Accepts icons or other custom content.
 * @slot max - Replaces the default maximum value label. Accepts icons or other custom content.
 *
 * @csspart form-control - The outermost wrapper element.
 * @csspart label - The label element.
 * @csspart control - The row containing the min label, track, max label, and optional value input.
 * @csspart input - The internal `<input type="range">` element.
 * @csspart tooltip - The value tooltip shown above the thumb when `show-value` is set.
 * @csspart min - The minimum value label.
 * @csspart max - The maximum value label.
 * @csspart value-input - The `made-input` element shown when `show-input` is set.
 *
 * @accessibility When no `label` attribute or slot is provided, add an `aria-label`
 * directly to the `<made-slider>` element so screen readers can announce the
 * slider's purpose, e.g. `<made-slider aria-label="Volume">`.
 */
class MadeRange extends rangeBaseClass {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, 'label', 'min', 'max');
        this.hasFocus = false;
        this._pendingEvents = new Set();
        /** The current value of the range input. */
        this.value = 0;
        /** The range's label. If you need to display HTML, use the `label` slot instead. */
        this.label = '';
        /** The minimum value. */
        this.min = 0;
        /** The maximum value. */
        this.max = 100;
        /** The step increment. Set to `'any'` to allow any value. */
        this.step = 1;
        /** Makes the range a required field. */
        this.required = false;
        /**
         * When set, displays a tooltip showing the
         * current value.
         */
        this.showTooltip = false;
        /** Hides the minimum and maximum value labels. */
        this.hideMinMax = false;
        /**
         * When set, renders a compact number input to the right of the slider.
         * The input stays in sync with the slider: dragging the thumb updates the
         * input, and typing in the input moves the thumb.
         */
        this.showInput = false;
        this._handleResize = () => {
            if (this.showTooltip)
                this.positionTooltip();
        };
    }
    updated(changedProperties) {
        if (changedProperties.has('value') ||
            changedProperties.has('min') ||
            changedProperties.has('max') ||
            changedProperties.has('showTooltip')) {
            requestAnimationFrame(() => {
                const pct = this.fillPercentage;
                this.style.setProperty('--range-fill', `${pct}%`);
                if (this.showTooltip)
                    this.positionTooltip();
            });
        }
        if (this._pendingEvents.size > 0) {
            if (this._pendingEvents.has('input')) {
                this.dispatchEvent(new Event('input', { bubbles: true, composed: true }));
            }
            if (this._pendingEvents.has('change')) {
                this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
            }
            this._pendingEvents.clear();
        }
    }
    firstUpdated() {
        requestAnimationFrame(() => {
            const pct = this.fillPercentage;
            this.style.setProperty('--range-fill', `${pct}%`);
            if (this.showTooltip)
                this.positionTooltip();
        });
    }
    connectedCallback() {
        super.connectedCallback();
        window.addEventListener('resize', this._handleResize);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        window.removeEventListener('resize', this._handleResize);
    }
    /**
     * Positions the value tooltip so it is horizontally centred above the thumb.
     * The tooltip uses CSS `translateX(-50%)` so we only need to set
     * `--tooltip-left` to the thumb centre X coordinate.
     */
    positionTooltip() {
        if (!this.input)
            return;
        const trackWidth = this.input.offsetWidth;
        if (!trackWidth)
            return;
        const thumbWidth = this.input.offsetHeight;
        const ratio = this.fillPercentage / 100;
        const thumbCentreX = ratio * (trackWidth - thumbWidth) + thumbWidth / 2;
        // CSS applies translateX(-50%) so left = centre X of the thumb
        this.style.setProperty('--tooltip-left', `${thumbCentreX}px`);
    }
    [getFormValue]() {
        return String(this.value);
    }
    formResetCallback() {
        this.value = Number(this.getAttribute('value') ?? 0);
    }
    formStateRestoreCallback(state, 
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _reason) {
        if (typeof state === 'string') {
            this.value = Number(state);
        }
    }
    get fillPercentage() {
        const min = Number(this.min);
        const max = Number(this.max);
        const value = Number(this.value);
        if (max === min)
            return 0;
        return Math.min(100, Math.max(0, ((value - min) / (max - min)) * 100));
    }
    handleInput(event) {
        event.stopPropagation();
        this._pendingEvents.add('input');
        this.value = Number(event.target.value);
    }
    handleChange(event) {
        event.stopPropagation();
        this._pendingEvents.add('change');
        this.requestUpdate();
    }
    handleFocus() {
        this.hasFocus = true;
    }
    handleBlur() {
        this.hasFocus = false;
    }
    handleValueInput(event) {
        event.stopPropagation();
        const target = event.target;
        if (target.value === '')
            return; // let user clear before retyping
        const num = Number(target.value);
        if (isNaN(num))
            return;
        this._pendingEvents.add('input');
        this.value = Math.min(Number(this.max), Math.max(Number(this.min), num));
    }
    handleValueChange(event) {
        event.stopPropagation();
        const target = event.target;
        const num = Number(target.value);
        if (target.value !== '' && !isNaN(num)) {
            this._pendingEvents.add('change');
            this.value = Math.min(Number(this.max), Math.max(Number(this.min), num));
        }
    }
    /**
     * Safari does not natively handle arrow keys on range inputs reliably.
     * Drive the value change explicitly so behaviour is consistent across browsers.
     */
    handleKeyDown(event) {
        const step = this.step === 'any' ? 1 : Number(this.step);
        const min = Number(this.min);
        const max = Number(this.max);
        let newValue = null;
        switch (event.key) {
            case 'ArrowRight':
            case 'ArrowUp':
                newValue = Math.min(max, Number(this.value) + step);
                break;
            case 'ArrowLeft':
            case 'ArrowDown':
                newValue = Math.max(min, Number(this.value) - step);
                break;
            case 'Home':
                newValue = min;
                break;
            case 'End':
                newValue = max;
                break;
            default:
                return;
        }
        if (newValue !== null && newValue !== Number(this.value)) {
            event.preventDefault();
            this._pendingEvents.add('input');
            this._pendingEvents.add('change');
            this.value = newValue;
        }
    }
    /** Sets focus on the range input. */
    focus(options) {
        this.input.focus(options);
    }
    /** Removes focus from the range input. */
    blur() {
        this.input.blur();
    }
    render() {
        const hasLabel = this.label ? true : !!this.hasSlotController.test('label');
        return html `
      <div
        part="form-control"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control--medium': true
        })}
      >
        ${hasLabel
            ? html ` <label
              part="label"
              class="made-c-form-control__label"
              for="input"
            >
              <slot name="label">${this.label}</slot>
            </label>`
            : nothing}

        <div
          part="control"
          class=${classMap({
            'made-c-range': true,
            'made-c-range--disabled': this.disabled,
            'made-c-range--focused': this.hasFocus
        })}
        >
          ${this.hideMinMax
            ? nothing
            : html ` <span
                part="min"
                class="made-c-range__bound"
                aria-hidden="true"
              >
                <slot name="min">${this.min}</slot>
              </span>`}
          <div class="made-c-range__track-wrapper">
            <input
              part="input"
              id="input"
              type="range"
              class="made-c-range__input"
              min=${this.min}
              max=${this.max}
              step=${ifDefined(this.step)}
              .value=${live(String(this.value))}
              ?disabled=${this.disabled}
              ?required=${this.required}
              tabindex="0"
              aria-valuemin=${this.min}
              aria-valuemax=${this.max}
              aria-valuenow=${this.value}
              @input=${this.handleInput}
              @change=${this.handleChange}
              @focus=${this.handleFocus}
              @blur=${this.handleBlur}
              @keydown=${this.handleKeyDown}
            />
            ${this.showTooltip
            ? html `<div
                  part="tooltip"
                  class="made-c-range__tooltip"
                  aria-hidden="true"
                >
                  <span class="made-c-range__tooltip-text">${this.value}</span>
                </div>`
            : nothing}
          </div>
          ${this.hideMinMax
            ? nothing
            : html ` <span
                part="max"
                class="made-c-range__bound"
                aria-hidden="true"
              >
                <slot name="max">${this.max}</slot>
              </span>`}
          ${this.showInput
            ? html `<made-input
                part="value-input"
                class="made-c-range__value-input"
                type="number"
                no-spinner
                .value=${live(String(this.value))}
                min=${this.min}
                max=${this.max}
                step=${ifDefined(this.step)}
                ?disabled=${this.disabled}
                aria-label=${this.label || 'Value'}
                @input=${this.handleValueInput}
                @change=${this.handleValueChange}
              ></made-input>`
            : nothing}
        </div>
      </div>
    `;
    }
}
MadeRange.styles = [componentStyles, formControlStyles, styles];
export default MadeRange;
__decorate([
    query('input')
], MadeRange.prototype, "input", void 0);
__decorate([
    state()
], MadeRange.prototype, "hasFocus", void 0);
__decorate([
    property({ type: Number })
], MadeRange.prototype, "value", void 0);
__decorate([
    property()
], MadeRange.prototype, "label", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MadeRange.prototype, "min", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MadeRange.prototype, "max", void 0);
__decorate([
    property()
], MadeRange.prototype, "step", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeRange.prototype, "required", void 0);
__decorate([
    property({ type: Boolean, attribute: 'show-tooltip', reflect: true })
], MadeRange.prototype, "showTooltip", void 0);
__decorate([
    property({ type: Boolean, attribute: 'hide-min-max', reflect: true })
], MadeRange.prototype, "hideMinMax", void 0);
__decorate([
    property({ type: Boolean, attribute: 'show-input', reflect: true })
], MadeRange.prototype, "showInput", void 0);
