// @mc-made/web-components/dist/components/slider/src/slider.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: block;

    --inline-size: 280px;
  }

  .made-c-range {
    display: flex;
    align-items: center;
    gap: var(--made-size-space-2-x);
    min-inline-size: var(--inline-size);
  }

  .made-c-range--disabled {
    cursor: not-allowed;
  }

  .made-c-range__bound {
    flex-shrink: 0;
    font-family: var(--made-font-family-primary);
    font-size: var(--made-font-size-body-large);
    font-weight: var(--made-font-weight-regular);
    color: var(--made-color-text-primary);
    line-height: var(--made-line-height-rem-05);
    user-select: none;
  }

  .made-c-range--disabled .made-c-range__bound {
    color: var(--made-color-text-disabled);
  }

  .made-c-form-control__label {
    display: flex;
    align-items: center;
    font-family: var(--made-font-family-primary);
    font-size: var(--made-font-size-body-small);
    color: var(--made-color-text-primary);
    font-weight: var(--made-font-weight-regular);
    margin-bottom: var(--made-size-space-2-x);
  }

  .made-c-range__value-input::part(input) {
    text-align: center;
  }

  .made-c-range__track-wrapper {
    position: relative;
    flex: 1 1 auto;
    min-width: 0;
    overflow: visible;
  }

  .made-c-range__tooltip {
    --tether-size: 6px;

    position: absolute;
    left: var(--tooltip-left, 50%);
    /* Sit above the thumb with a gap equal to the beak height */
    bottom: calc(
      var(--made-component-range-thumb-height) + var(--tether-size) +
        var(--made-size-space-2-x)
    );
    transform: translateX(-50%) translateY(4px);
    opacity: 0;
    background-color: var(--made-color-background-surface-inverse);
    color: var(--made-color-text-on-color);
    padding: var(--made-size-space-2-x) var(--made-size-space-3-x);
    border-radius: var(--made-component-tooltip-border-radius);
    box-shadow: var(--made-component-tooltip-box-shadow);
    font-size: var(--made-font-size-body-small);
    font-family: var(--made-font-family-primary);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--made-line-height-rem-02);
    white-space: nowrap;
    pointer-events: none;
    user-select: none;
    /* overflow: visible needed to render the beak ::after */
    overflow: visible;
    transition:
      opacity var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease),
      transform var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease);
  }

  /* Downward beak pointing at the thumb (mirrors tooltip --top beak) */
  .made-c-range__tooltip::after {
    content: '';
    position: absolute;
    background: inherit;
    pointer-events: none;
    width: 16px;
    height: 9px;
    left: 50%;
    bottom: calc(-1 * var(--tether-size));
    transform: translateX(-50%);
    clip-path: path(
      'M6.43826 8.249396C7.23891 9.250209 8.76109 9.25021 9.56174 8.249397L16 0.20157H0L6.43826 8.249396Z'
    );
  }

  .made-c-range__tooltip-text {
    display: block;
  }

  /* Show on hover, drag (active), keyboard focus, or touch focus (mobile) */
  .made-c-range__input:hover:not(:disabled) ~ .made-c-range__tooltip,
  .made-c-range__input:active:not(:disabled) ~ .made-c-range__tooltip,
  .made-c-range__input:focus-visible ~ .made-c-range__tooltip,
  .made-c-range__input:focus:not(:disabled) ~ .made-c-range__tooltip {
    opacity: 1;
    transform: translateX(-50%) translateY(0);
  }

  .made-c-range__input {
    -webkit-appearance: none;
    appearance: none;
    display: block;
    width: 100%;
    background: transparent;
    cursor: pointer;
    margin: 0;
    padding: 0;
    height: var(--made-component-range-thumb-height);
  }

  .made-c-range__input:disabled {
    cursor: not-allowed;
  }

  /* ===================================
   * Track — WebKit
   * =================================== */

  .made-c-range__input::-webkit-slider-runnable-track {
    height: var(--made-component-range-track-height);
    border-radius: var(--made-component-range-track-border-radius);
    background: linear-gradient(
      to right,
      var(--made-component-range-track-fill-background) var(--range-fill, 0%),
      var(--made-color-border-default) var(--range-fill, 0%)
    );
    cursor: pointer;
    transition: background var(--made-motion-duration-short)
      var(--made-motion-easing-standard-ease);
  }

  .made-c-range__input:disabled::-webkit-slider-runnable-track {
    background: linear-gradient(
      to right,
      var(--made-color-border-strong) var(--range-fill, 0%),
      var(--made-color-border-disabled) var(--range-fill, 0%)
    );
    cursor: not-allowed;
  }

  /* ===================================
   * Track — Firefox
   * =================================== */

  .made-c-range__input::-moz-range-track {
    height: var(--made-component-range-track-height);
    border-radius: var(--made-component-range-track-border-radius);
    background: var(--made-color-border-default);
    cursor: pointer;
  }

  .made-c-range__input::-moz-range-progress {
    height: var(--made-component-range-track-height);
    border-radius: var(--made-component-range-track-border-radius);
    background: var(--made-component-range-track-fill-background);
  }

  .made-c-range__input:disabled::-moz-range-track {
    background: var(--made-color-border-disabled);
    cursor: not-allowed;
  }

  .made-c-range__input:disabled::-moz-range-progress {
    background: var(--made-color-border-strong);
  }

  /* ===================================
   * Thumb — WebKit
   * =================================== */

  .made-c-range__input::-webkit-slider-thumb {
    -webkit-appearance: none;
    width: var(--made-component-range-thumb-width);
    height: var(--made-component-range-thumb-height);
    margin-top: calc(
      (
          var(--made-component-range-track-height) - var(
              --made-component-range-thumb-height
            )
        ) / 2
    );
    border-radius: var(--made-component-range-thumb-border-radius);
    background-color: var(--made-color-background-surface-default);
    border: 2px solid var(--made-component-range-thumb-border-color);
    box-shadow: var(--made-component-range-thumb-box-shadow);
    cursor: pointer;
    transition:
      background-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease),
      border-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease);
  }

  .made-c-range__input:disabled::-webkit-slider-thumb {
    background-color: var(--made-color-background-default);
    border-color: var(--made-color-border-disabled);
    cursor: not-allowed;
    box-shadow: none;
  }

  /* ===================================
   * Thumb — Firefox
   * =================================== */

  .made-c-range__input::-moz-range-thumb {
    width: var(--made-component-range-thumb-width);
    height: var(--made-component-range-thumb-height);
    border-radius: var(--made-component-range-thumb-border-radius);
    background-color: var(--made-color-background-surface-default);
    border: 2px solid var(--made-component-range-thumb-border-color);
    box-shadow: var(--made-component-range-thumb-box-shadow);
    cursor: pointer;
    transition:
      background-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease),
      border-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease);
  }

  .made-c-range__input:disabled::-moz-range-thumb {
    background-color: var(--made-color-background-default);
    border-color: var(--made-color-border-disabled);
    cursor: not-allowed;
    box-shadow: none;
  }

  /* ===================================
   * Focus ring
   * =================================== */

  .made-c-range__input:focus-visible {
    outline: none;
  }

  .made-c-range__input:focus-visible::-webkit-slider-thumb {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: 2px;
  }

  .made-c-range__input:focus-visible::-moz-range-thumb {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: 2px;
  }
`;
