// @mc-made/web-components/dist/components/alert/src/alert.styles.js
import { css } from 'lit';
export default css `
  :host {
    --width: 100%;
    display: contents;
  }

  .made-c-alert {
    --font-size: var(--made-font-size-body-default);
    --line-height: normal;
    --icon-color: var(--made-color-icon-primary);

    display: flex;
    align-items: flex-start;
    padding: var(--made-component-alert-banner-padding-y)
      var(--made-component-alert-banner-padding-x);
    max-width: var(--width);
    border-radius: var(--made-component-alert-border-radius);
    border: 1px solid transparent;
    font-family: var(--made-font-family-primary);
    font-size: var(--font-size);
    line-height: var(--line-height);
    letter-spacing: var(--made-letter-spacing-default);
    color: var(--made-color-text-primary);
    position: relative;
  }

  /* Sentiment variants */
  .made-c-alert--default {
    background: var(--made-color-background-surface-default);
    border-color: var(--made-color-border-default);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-neutral);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      background-color: var(--made-color-background-feedback-neutral-hover);
    }
  }

  .made-c-alert--neutral {
    background: var(--made-color-background-feedback-neutral-strong-default);
    border-color: var(--made-color-border-neutral);
    color: var(--made-color-text-on-color);
    --icon-color: var(--made-color-icon-on-color);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-on-color);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      color: var(--made-color-icon-primary);
      background-color: var(--made-color-background-feedback-neutral-hover);
    }
  }

  .made-c-alert--success {
    background: var(--made-color-background-feedback-success-default);
    border-color: var(--made-color-border-success);
    --icon-color: var(--made-color-icon-success);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-success);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      background-color: var(--made-color-background-feedback-success-hover);
    }
  }

  .made-c-alert--danger {
    background: var(--made-color-background-feedback-danger-default);
    border-color: var(--made-color-border-danger);
    --icon-color: var(--made-color-icon-danger);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-danger);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      background-color: var(--made-color-background-feedback-danger-hover);
    }
  }

  .made-c-alert--warning {
    background: var(--made-color-background-feedback-warning-default);
    border-color: var(--made-color-border-warning);
    --icon-color: var(--made-color-icon-warning);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-warning);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      background-color: var(--made-color-background-feedback-warning-hover);
    }
  }

  .made-c-alert--information {
    background: var(--made-color-background-feedback-information-default);
    border-color: var(--made-color-border-information);
    --icon-color: var(--made-color-icon-information);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-information);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      background-color: var(--made-color-background-feedback-information-hover);
    }
  }

  .made-c-alert--discovery {
    background: var(--made-color-background-feedback-discovery-default);
    border-color: var(--made-color-border-discovery);
    --icon-color: var(--made-color-icon-discovery);

    & .made-c-alert__dismiss-button::part(base) {
      color: var(--made-color-icon-discovery);
    }

    & .made-c-alert__dismiss-button::part(base):hover {
      background-color: var(--made-color-background-feedback-discovery-hover);
    }
  }

  /* Header start slot */
  .made-c-alert__header-start {
    display: flex;
    flex: 0 0 auto;
    align-items: center;
    color: var(--icon-color);
    padding-inline-end: var(--made-size-space-3-x);

    &::slotted(made-icon) {
      font-size: 1.5rem;
    }
  }

  /* Body */
  .made-c-alert__body {
    display: flex;
    flex-direction: column;
    flex: 1;
    min-width: 0;
    letter-spacing: var(--made-letter-spacing-default);
  }

  .made-c-alert__body > * + * {
    margin-block-start: var(--made-size-space-2-x);
  }

  .made-c-alert__header {
    display: flex;
    align-items: center;
  }

  /* Title */
  .made-c-alert__title {
    --font-size: var(--made-component-alert-title-font-size);
    font-weight: var(--made-component-alert-title-font-weight);
    font-size: var(--font-size);
    line-height: var(--made-component-alert-title-line-height);
    letter-spacing: inherit;
    color: inherit;
  }

  /* Timestamp */
  .made-c-alert__timestamp {
    --font-size: var(--made-font-size-body-small);
    font-weight: var(--made-font-weight-regular);
    font-size: var(--font-size);
    line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
    margin-inline-start: var(--made-size-space-2-x);
    letter-spacing: inherit;
    color: var(--made-color-text-secondary);
  }

  /* Header end slot */
  .made-c-alert__header-end {
    display: flex;
    align-items: center;
    flex-shrink: 0;
    margin-inline-start: var(--made-size-space-2-x);
    color: var(--made-color-icon-secondary);

    &::slotted(made-icon) {
      font-size: 1.25rem;
    }
  }

  /* Message */
  .made-c-alert__message {
    --font-size: var(--made-component-alert-subtitle-font-size);
    font-weight: var(--made-component-alert-subtitle-font-weight);
    font-size: var(--font-size);
    line-height: var(--made-component-alert-subtitle-line-height);
    letter-spacing: inherit;
    color: inherit;
  }

  /* Actions */
  .made-c-alert__actions {
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    gap: var(--made-size-space-4-x);
    margin-top: var(--made-size-space-2-x);
  }

  .made-c-alert__dismiss-button {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    font-size: inherit;
    border: none;
    background: none;
    padding: 0;
    padding-inline-start: var(--made-size-space-8-x);
  }

  /* ─── Toast variant ──────────────────────────────────────────── */

  .made-c-alert--toast {
    padding: var(--made-component-alert-toast-padding-y, 1rem)
      var(--made-component-alert-toast-padding-x, 1rem);
    box-shadow: var(--made-component-alert-toast-box-shadow);
  }

  @keyframes alert-slide-in-from-bottom {
    from {
      opacity: 0;
      transform: translateY(0.75rem);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }

  @keyframes alert-slide-out-to-bottom {
    from {
      opacity: 1;
      transform: translateY(0);
    }
    to {
      opacity: 0;
      transform: translateY(0.75rem);
    }
  }

  @keyframes alert-slide-in-from-top {
    from {
      opacity: 0;
      transform: translateY(-0.75rem);
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }

  @keyframes alert-slide-out-to-top {
    from {
      opacity: 1;
      transform: translateY(0);
    }
    to {
      opacity: 0;
      transform: translateY(-0.75rem);
    }
  }

  /* Default (bottom-origin) transitions */
  .made-c-alert--toast.made-c-alert--show {
    animation: alert-slide-in-from-bottom var(--made-motion-duration-default)
      var(--made-motion-easing-emphasised-ease-out) both;
  }

  .made-c-alert--toast.made-c-alert--hide {
    animation: alert-slide-out-to-bottom var(--made-motion-duration-default)
      var(--made-motion-easing-emphasised-ease-in) both;
  }

  /* Top-position placements slide from the opposite direction */
  :host([data-placement^='top']) .made-c-alert--toast.made-c-alert--show {
    animation-name: alert-slide-in-from-top;
  }

  :host([data-placement^='top']) .made-c-alert--toast.made-c-alert--hide {
    animation-name: alert-slide-out-to-top;
  }

  /* Respect the user's motion preference — suppress translation, keep opacity */
  @media (prefers-reduced-motion: reduce) {
    .made-c-alert--toast.made-c-alert--show,
    .made-c-alert--toast.made-c-alert--hide {
      animation-duration: 1ms;
    }
  }
`;
