// @mc-made/web-components/dist/components/alert/src/alert.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIconButton from '../../icon-button/src/icon-button.component.js';
import styles from './alert.styles.js';
/**
 * @summary Alerts display important messages inline or as toast notifications,
 * providing contextual information in response to user actions or system events.
 * @documentation '../stories/alert.mdx'
 *
 * @dependency made-icon-button
 *
 * @fires mde-show       {CustomEvent} - Emitted when the alert begins to show.
 * @fires mde-after-show {CustomEvent} - Emitted after the alert finishes showing.
 * @fires mde-hide       {CustomEvent} - Emitted when the alert begins to hide.
 * @fires mde-after-hide {CustomEvent} - Emitted after the alert finishes hiding.
 *
 * @slot header-start - To display a sentiment icon or avatar.
 * @slot header-end - Optional content (icon) to display at the end of the alert header.
 * @slot default - HTML content for the alert message. Can be used instead of the `alert-message` attribute for more complex content.
 * @slot actions - Optional action elements placed at the bottom of the alert.
 *
 * @csspart base - The component's base wrapper.
 * @csspart header-start - The container that wraps the icon area.
 * @csspart header-end - The container that wraps the icon area.
 * @csspart body - The container that wraps the title and content.
 * @csspart alert-title - The alert title.
 * @csspart alert-timestamp - The alert timestamp.
 * @csspart message - The alert message content.
 * @csspart actions - The container that wraps the actions slot.
 * @csspart dismiss-button - The dismiss button (made-icon-button).
 * @csspart dismiss-button-base - The dismiss button's base part.
 */
class MadeAlert extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'header-start', 'header-end', 'actions');
        this._autoHideStartTime = 0;
        this._autoHideRemaining = 0;
        this._isHovered = false;
        this._isFocused = false;
        this._initialized = false;
        this._isToast = false;
        /** The alert's sentiment, which controls its color. */
        this.sentiment = 'default';
        /** The alert's title. */
        this.alertTitle = '';
        /** The alert's message content. */
        this.alertMessage = '';
        /** The alert's time-stamp */
        this.alertTimestamp = '';
        /** Renders a dismiss button to allow the user to close the alert. */
        this.dismissible = false;
        /**
         * How long (ms) a toast stays visible before auto-dismissing when shown via
         * `toast()`. Set to `0` to disable auto-dismiss.
         */
        this.duration = 5000;
        /** Controls the visibility of the alert. */
        this.open = false;
        /** Placement of the toast in the viewport. */
        this.placement = 'top-right';
        this._pauseAutoHide = () => {
            if (this._autoHideTimer === undefined)
                return;
            clearTimeout(this._autoHideTimer);
            this._autoHideTimer = undefined;
            this._autoHideRemaining -= Date.now() - this._autoHideStartTime;
        };
        this._onMouseEnter = () => {
            this._isHovered = true;
            this._pauseAutoHide();
        };
        this._onMouseLeave = () => {
            this._isHovered = false;
            if (!this._isFocused)
                this._resumeAutoHide();
        };
        this._onFocusIn = () => {
            this._isFocused = true;
            this._pauseAutoHide();
        };
        this._onFocusOut = () => {
            this._isFocused = false;
            if (!this._isHovered)
                this._resumeAutoHide();
        };
        this._resumeAutoHide = () => {
            if (this._autoHideRemaining <= 0 || this._autoHideTimer !== undefined)
                return;
            this._startAutoHideTimer();
        };
    }
    static injectStyles() {
        if (typeof document === 'undefined')
            return;
        // Inject the stylesheet only once per document, keyed by the style element id
        // rather than a JS flag so HMR / multiple bundle instances stay in sync.
        if (!document.getElementById('made-toast-stack-styles')) {
            const style = document.createElement('style');
            style.id = 'made-toast-stack-styles';
            style.textContent = `
      [data-toast-stack] {
        position: fixed;
        z-index: var(--made-z-index-alert-toast);
        display: flex;
        flex-direction: column;
        gap: var(--made-component-alert-toast-gap);
        padding: 1rem;
        min-width: 280px;
        max-width: 480px;
        pointer-events: none;
      }
      [data-toast-stack] > * { pointer-events: auto; }
      [data-toast-stack="top-left"]      { top: 0; left: 0; }
      [data-toast-stack="top-center"]    { top: 0; left: 50%; transform: translateX(-50%); }
      [data-toast-stack="top-right"]     { top: 0; right: 0; }
      [data-toast-stack="bottom-left"]   { bottom: 0; left: 0; flex-direction: column-reverse; }
      [data-toast-stack="bottom-center"] { bottom: 0; left: 50%; transform: translateX(-50%); flex-direction: column-reverse; }
      [data-toast-stack="bottom-right"]  { bottom: 0; right: 0; flex-direction: column-reverse; }
    `;
            document.head.appendChild(style);
        }
        // Eagerly ensure all stack containers exist in the DOM so live regions are
        // present before content is injected — required for reliable announcement in
        // JAWS and some other ATs. Re-uses existing DOM nodes to survive HMR and
        // multiple bundle instances.
        const placements = [
            'top-left',
            'top-center',
            'top-right',
            'bottom-left',
            'bottom-center',
            'bottom-right'
        ];
        for (const placement of placements) {
            if (!MadeAlert._stacks.has(placement)) {
                const existing = document.querySelector(`[data-toast-stack="${placement}"]`);
                if (existing) {
                    MadeAlert._stacks.set(placement, existing);
                }
                else {
                    const el = document.createElement('div');
                    el.role = 'log';
                    el.setAttribute('aria-live', 'polite');
                    el.setAttribute('aria-relevant', 'additions');
                    el.dataset.toastStack = placement;
                    document.body.appendChild(el);
                    MadeAlert._stacks.set(placement, el);
                }
            }
        }
    }
    static getStack(placement) {
        return MadeAlert._stacks.get(placement);
    }
    connectedCallback() {
        super.connectedCallback();
        MadeAlert.injectStyles();
    }
    firstUpdated() {
        this.base.hidden = !this.open;
    }
    updated(changed) {
        if (!this._initialized) {
            // Skip the first update — initial open state is handled by firstUpdated()
            this._initialized = true;
            return;
        }
        if (changed.has('open')) {
            void this.handleOpenChange();
        }
    }
    _waitForAnimation() {
        return new Promise(resolve => {
            requestAnimationFrame(() => {
                const animName = getComputedStyle(this.base).animationName;
                if (!animName || animName === 'none') {
                    resolve();
                }
                else {
                    this.base.addEventListener('animationend', () => resolve(), {
                        once: true
                    });
                }
            });
        });
    }
    async handleOpenChange() {
        if (this.open) {
            // Show
            this.emit('mde-show');
            this.base.hidden = false;
            this.base.classList.add('made-c-alert--show');
            await this._waitForAnimation();
            this.base.classList.remove('made-c-alert--show');
            this.emit('mde-after-show');
        }
        else {
            // Hide
            this.emit('mde-hide');
            this.base.classList.add('made-c-alert--hide');
            await this._waitForAnimation();
            this.base.classList.remove('made-c-alert--hide');
            this.base.hidden = true;
            this.emit('mde-after-hide');
        }
    }
    _startAutoHideTimer() {
        this._autoHideStartTime = Date.now();
        this._autoHideTimer = setTimeout(() => void this.hide(), this._autoHideRemaining);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        clearTimeout(this._autoHideTimer);
        this.removeEventListener('mouseenter', this._onMouseEnter);
        this.removeEventListener('mouseleave', this._onMouseLeave);
        this.removeEventListener('focusin', this._onFocusIn);
        this.removeEventListener('focusout', this._onFocusOut);
        this._isHovered = false;
        this._isFocused = false;
    }
    /** Shows the alert. */
    show() {
        if (this.open)
            return Promise.resolve();
        this.open = true;
        return new Promise(resolve => this.addEventListener('mde-after-show', () => resolve(), { once: true }));
    }
    /** Hides the alert. */
    hide() {
        if (!this.open)
            return Promise.resolve();
        this.open = false;
        return new Promise(resolve => this.addEventListener('mde-after-hide', () => resolve(), { once: true }));
    }
    _cleanupAfterToast(stack, originalParent, originalNextSibling, resolve) {
        this.removeEventListener('mouseenter', this._onMouseEnter);
        this.removeEventListener('mouseleave', this._onMouseLeave);
        this.removeEventListener('focusin', this._onFocusIn);
        this.removeEventListener('focusout', this._onFocusOut);
        this._isHovered = false;
        this._isFocused = false;
        if (stack.contains(this)) {
            stack.removeChild(this);
        }
        resolve();
        this._isToast = false;
        delete this.dataset.placement;
        if (originalParent) {
            originalParent.insertBefore(this, originalNextSibling);
        }
    }
    /**
     * Inserts this alert into a fixed stack at the given `placement`.
     *
     * @param placement - Stack position. Defaults to the `placement` property of the alert.
     */
    async toast(placement = this.placement) {
        return new Promise(resolve => {
            // Reset any stale state from a previous toast cycle
            clearTimeout(this._autoHideTimer);
            this._autoHideTimer = undefined;
            this._isHovered = false;
            this._isFocused = false;
            const stack = MadeAlert.getStack(placement);
            // Save original position so a template-defined element can be restored after hiding
            const originalParent = this.parentElement;
            const originalNextSibling = this.nextSibling;
            stack.appendChild(this);
            this._isToast = true;
            this.dataset.placement = placement;
            // Wait for first render (needed for programmatically-created elements that
            // weren't in the DOM yet — firstUpdated must run before show() is called)
            void this.updateComplete.then(() => {
                // Force a reflow so the entry animation plays from its initial state
                // eslint-disable-next-line @typescript-eslint/no-unused-expressions
                this.clientWidth;
                void this.show();
            });
            if (this.duration > 0) {
                this._autoHideRemaining = this.duration;
                this._startAutoHideTimer();
                this.addEventListener('mouseenter', this._onMouseEnter);
                this.addEventListener('mouseleave', this._onMouseLeave);
                this.addEventListener('focusin', this._onFocusIn);
                this.addEventListener('focusout', this._onFocusOut);
            }
            this.addEventListener('mde-after-hide', () => this._cleanupAfterToast(stack, originalParent, originalNextSibling, resolve), { once: true });
        });
    }
    async handleDismissClick() {
        clearTimeout(this._autoHideTimer);
        await this.hide();
    }
    render() {
        const hasHeaderStartSlot = this.hasSlotController.test('header-start');
        const hasActions = this.hasSlotController.test('actions');
        const hasTimestamp = Boolean(this.alertTimestamp);
        const hasHeaderEndSlot = this.hasSlotController.test('header-end');
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-alert': true,
            'made-c-alert--default': this.sentiment === 'default',
            'made-c-alert--neutral': this.sentiment === 'neutral',
            'made-c-alert--success': this.sentiment === 'success',
            'made-c-alert--danger': this.sentiment === 'danger',
            'made-c-alert--warning': this.sentiment === 'warning',
            'made-c-alert--information': this.sentiment === 'information',
            'made-c-alert--discovery': this.sentiment === 'discovery',
            'made-c-alert--toast': this._isToast,
            'made-c-alert--dismissible': this.dismissible,
            'made-c-alert--has-actions': hasActions
        })}
        role=${ifDefined(this._isToast ? undefined : 'status')}
      >
        <slot
          part="header-start"
          name="header-start"
          class=${classMap({
            'made-c-alert__header-start': hasHeaderStartSlot
        })}
        ></slot>

        <div part="body" class="made-c-alert__body">
          <div class="made-c-alert__header">
            <span part="alert-title" class="made-c-alert__title">
              ${this.alertTitle}
            </span>
            <span
              part="alert-timestamp"
              class=${classMap({ 'made-c-alert__timestamp': hasTimestamp })}
            >
              ${this.alertTimestamp}
            </span>
            <slot
              name="header-end"
              part="header-end"
              class=${classMap({
            'made-c-alert__header-end': hasHeaderEndSlot
        })}
            ></slot>
          </div>

          ${this.alertMessage
            ? html `<div part="message" class="made-c-alert__message">
                ${this.alertMessage}
              </div>`
            : nothing}

          <slot></slot>

          ${hasActions
            ? html `<div part="actions" class="made-c-alert__actions">
                <slot name="actions"></slot>
              </div>`
            : nothing}
        </div>

        ${this.dismissible
            ? html `
              <made-icon-button
                part="dismiss-button"
                exportparts="base:dismiss-button-base"
                class="made-c-alert__dismiss-button"
                aria-label="${this.localize.term('close')}"
                icon-name="x-mark-outline"
                variant="ghost"
                library="system"
                size="small"
                @click=${this.handleDismissClick}
              ></made-icon-button>
            `
            : nothing}
      </div>
    `;
    }
}
MadeAlert.styles = [componentStyles, styles];
MadeAlert.dependencies = {
    'made-icon-button': MadeIconButton
};
MadeAlert._stacks = new Map();
export default MadeAlert;
__decorate([
    query('[part~="base"]')
], MadeAlert.prototype, "base", void 0);
__decorate([
    property({ reflect: true })
], MadeAlert.prototype, "sentiment", void 0);
__decorate([
    property({ type: String, attribute: 'alert-title' })
], MadeAlert.prototype, "alertTitle", void 0);
__decorate([
    property({ type: String, attribute: 'alert-message' })
], MadeAlert.prototype, "alertMessage", void 0);
__decorate([
    property({ type: String, attribute: 'alert-timestamp' })
], MadeAlert.prototype, "alertTimestamp", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeAlert.prototype, "dismissible", void 0);
__decorate([
    property({ type: Number })
], MadeAlert.prototype, "duration", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeAlert.prototype, "open", void 0);
__decorate([
    property({ type: String })
], MadeAlert.prototype, "placement", void 0);
