// @mc-made/web-components/dist/components/radio/src/radio.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var _a;
/* eslint-disable @typescript-eslint/await-thenable */
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormState, getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { html } from 'lit';
import { internals, mixinElementInternals } from '../../../internal/face/element-internals.js';
import { mixinOnReportValidity, onReportValidity } from '../../../internal/face/on-report-validity.js';
import { property, query, state } from 'lit/decorators.js';
import { RadioController } from './radio-controller.js';
import { RadioValidator } from '../../../internal/face/validators/radio-validator.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './radio.styles.js';
const CHECKED = Symbol('checked');
/**
 * @summary Radios allow the user select one option from a set of options.
 * @documentation TODO
 *
 * @fires change {Event} - Fired when the radio is selected.
 * @fires input {Event} - Fired when the radio receives input.
 * @fires mde-invalid {CustomEvent} - Fired when the radio is in an invalid state.
 * @fires mde-reset {CustomEvent} - Fired when the radio is reset by the form.
 *
 * @csspart base - The component's base wrapper.
 * @csspart control - The radio's control in it's unchecked state.
 * @csspart control-checked - The radio's control in it's checked state.
 * @csspart label - The label's wrapper.
 * @csspart supporting-text - The supporting text's wrapper.
 *
 */
const radioBaseClass = mixinOnReportValidity(mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement))));
class MadeRadio extends radioBaseClass {
    constructor() {
        super();
        this.radioController = new RadioController(this);
        /** Constraint validation error state controlled by `reportValidity()` */
        this.hasNativeError = false;
        this[_a] = false;
        /**
         * Whether or not the radio is required. If any radio is required in a group,
         * all radios are implicitly required.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required
         */
        this.required = false;
        /** The element value to use in form submission when checked. */
        this.value = 'on';
        /** The radio's size. */
        this.size = 'medium';
        /**
         * Gets or sets whether or not the text field is in an invalid state.
         *
         * This error state overrides the error state controlled by
         * `reportValidity()`.
         */
        this.error = false;
        /** Set to 'branded' for alternative brand appearance. */
        this.appearance = 'neutral';
        /** Makes the radio readonly. */
        this.readonly = false;
        this.handleClick = async (event) => {
            if (this.checked || this.disabled || this.readonly) {
                return;
            }
            // Wait for a microtask.
            await 0;
            if (event.defaultPrevented) {
                return;
            }
            this.checked = true;
            this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
            this.dispatchEvent(new InputEvent('input', { bubbles: true, composed: true }));
            if (this.hasNativeError) {
                for (const control of this.radioController.controls) {
                    control.reportValidity();
                }
            }
        };
        this.handleKeydown = async (event) => {
            // Wait for a tick.
            await 0;
            if (event.key !== ' ' || event.defaultPrevented) {
                return;
            }
            this.click();
        };
        this.addController(this.radioController);
        this[internals].role = 'radio';
        this.addEventListener('click', this.handleClick);
        this.addEventListener('keydown', this.handleKeydown);
    }
    /** Whether or not the radio is selected. */
    get checked() {
        return this[CHECKED];
    }
    set checked(checked) {
        const wasChecked = this.checked;
        if (wasChecked === checked) {
            return;
        }
        this[CHECKED] = checked;
        if (checked) {
            this[internals].states.add('checked');
        }
        else {
            this[internals].states.delete('checked');
        }
        this.requestUpdate('checked', wasChecked);
        this.radioController.handleCheckedChange();
    }
    get hasError() {
        return this.error || this.hasNativeError;
    }
    updated() {
        this[internals].ariaChecked = String(this.checked);
        // Using aria-disabled since the readonly attribute is not valid and aria-readonly
        // is not currently announced by the majority of screen readers.
        this[internals].ariaDisabled = this.readonly ? 'true' : 'false';
    }
    render() {
        const hasSupportingText = !!this.supportingText;
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-radio': true,
            'made-c-radio--neutral': this.appearance === 'neutral',
            'made-c-radio--branded': this.appearance === 'branded',
            'made-c-radio--checked': this.checked,
            'made-c-radio--disabled': this.disabled,
            'made-c-radio--readonly': this.readonly,
            'made-c-radio--error': !this.disabled && this.hasError,
            'made-c-radio--medium': this.size === 'medium',
            'made-c-radio--large': this.size === 'large',
            'made-c-form-control--has-supporting-text': hasSupportingText
        })}
      >
        <span
          part="${`control${this.checked ? ' control-checked' : ''}`}"
          class="made-c-radio__control"
        >
          ${this.readonly && this.checked
            ? html `<svg
                xmlns="http://www.w3.org/2000/svg"
                viewBox="0 0 16 16"
                fill="none"
                aria-hidden="true"
              >
                <circle
                  cx="8"
                  cy="8"
                  r="3"
                  stroke="currentColor"
                  fill="currentColor"
                  stroke-width="1"
                />
              </svg>`
            : ``}
        </span>
        <div class="made-c-radio__label-container">
          <slot part="label" class="made-c-radio__label"></slot>
          <span
            part="supporting-text"
            id="supporting-text"
            class="made-c-form-control__supporting-text"
            aria-hidden=${hasSupportingText ? 'false' : 'true'}
          >
            ${this.supportingText}
          </span>
        </div>
      </div>
    `;
    }
    [(_a = CHECKED, getFormValue)]() {
        return this.checked ? this.value : null;
    }
    [getFormState]() {
        return String(this.checked);
    }
    formResetCallback() {
        const allRadios = this.radioController.controls;
        // Force all radios to reset state, on first callback,
        // to ensure validity is calculated correctly.
        const index = allRadios.findIndex(radio => radio.value === this.value);
        if (index === 0) {
            this.resetRadios(allRadios);
            if (this.hasNativeError) {
                this.emit('mde-reset');
            }
        }
        this.hasNativeError = false;
    }
    formStateRestoreCallback(state) {
        this.checked = state === 'true';
    }
    // @ts-expect-error ts2425
    [createValidator]() {
        return new RadioValidator(() => {
            if (!this.radioController) {
                // Validation runs on superclass construction, so selection controller
                // might not actually be ready until this class constructs.
                return [this];
            }
            return this.radioController.controls;
        });
    }
    // @ts-expect-error ts2425
    [getValidityAnchor]() {
        return this.radio;
    }
    // @ts-expect-error ts2425
    [onReportValidity](invalidEvent) {
        // Prevent default pop-up behavior.
        invalidEvent?.preventDefault();
        this.hasNativeError = !!invalidEvent;
        if (invalidEvent) {
            this.emit('mde-invalid');
        }
    }
    resetRadios(allRadios) {
        // Set to original state
        for (const radio of allRadios) {
            // The checked property does not reflect, so the original attribute set by
            // the user is used to determine the default value.
            radio.checked = radio.hasAttribute('checked');
        }
    }
}
MadeRadio.styles = [componentStyles, formControlStyles, styles];
export default MadeRadio;
__decorate([
    query('.radio')
], MadeRadio.prototype, "radio", void 0);
__decorate([
    state()
], MadeRadio.prototype, "hasNativeError", void 0);
__decorate([
    property({ type: Boolean })
], MadeRadio.prototype, "checked", null);
__decorate([
    property({ type: Boolean })
], MadeRadio.prototype, "required", void 0);
__decorate([
    property()
], MadeRadio.prototype, "value", void 0);
__decorate([
    property({ reflect: true })
], MadeRadio.prototype, "size", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeRadio.prototype, "supportingText", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeRadio.prototype, "error", void 0);
__decorate([
    property()
], MadeRadio.prototype, "appearance", void 0);
__decorate([
    property({ type: Boolean })
], MadeRadio.prototype, "readonly", void 0);
