// @mc-made/web-components/dist/components/radio/src/radio.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  :host {
    display: block;
  }

  .made-c-radio {
    display: inline-flex;
    align-items: center;
    font-family: var(--made-font-family-primary);
    font-weight: var(--made-font-weight-medium);
    color: var(--made-color-text-primary);
    vertical-align: middle;
    cursor: pointer;
  }

  .made-c-radio.made-c-form-control--has-supporting-text {
    align-items: flex-start;
  }

  .made-c-radio--medium {
    --size: var(--made-size-space-4-x);
    --stroke: 4px;

    font-size: var(--made-font-size-body-small);
  }

  .made-c-radio--large {
    --size: var(--made-size-space-5-x);
    --stroke: 5.25px;

    font-size: var(--made-font-size-body-default);
  }

  .made-c-radio__control {
    all: unset;
    box-sizing: border-box;
    flex: 0 0 auto;
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    width: var(--size);
    height: var(--size);
    border: solid 1px var(--made-component-form-control-border-color);
    box-shadow: none;
    border-radius: 50%;
    background-color: var(--made-component-form-control-background);
    transition: box-shadow var(--made-motion-duration-short)
      var(--made-motion-easing-emphasised-ease-out);
  }

  :host(:focus-visible) {
    outline: 0;
  }

  /* Checked - neutral */

  .made-c-radio--checked.made-c-radio--neutral .made-c-radio__control,
  .made-c-radio--neutral .made-c-radio__control:checked {
    border-color: var(--made-color-border-neutral-strong);
    box-shadow: inset 0 0 0 var(--stroke)
      var(--made-color-border-neutral-strong);
  }

  /* Checked - branded */

  .made-c-radio--checked.made-c-radio--branded .made-c-radio__control,
  .made-c-radio--branded .made-c-radio__control:checked {
    border-color: var(--made-color-border-brand);
    box-shadow: inset 0 0 0 var(--stroke) var(--made-color-border-brand);
  }

  /* Disabled */

  .made-c-radio--disabled,
  .made-c-radio:has(.made-c-radio__control:disabled) {
    cursor: not-allowed;

    & .made-c-radio__control,
    & .made-c-radio__control:disabled {
      border-color: var(--made-color-border-disabled);

      .made-c-radio--checked &,
      & :checked {
        box-shadow: inset 0 0 0 var(--stroke) var(--made-color-border-disabled);
      }
    }

    & .made-c-radio__label {
      color: var(--made-color-text-disabled);
    }
  }

  /* Readonly */

  .made-c-radio--readonly,
  .made-c-radio:has(.made-c-radio__control[readonly]) {
    cursor: default;

    & .made-c-radio__control,
    & .made-c-radio__control[readonly] {
      border-color: var(--made-color-border-disabled);
      background-color: var(--made-component-form-control-background);
      color: var(--made-color-background-brand-primary-default);
      pointer-events: none;
    }

    &.made-c-radio--checked {
      --stroke: unset;
    }
  }

  /* Focused */

  :host(:focus-visible) .made-c-radio__control,
  .made-c-radio__control:focus-visible {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: 2px;
  }

  .made-c-radio__label {
    display: inline-block;
    color: var(--made-color-text-primary);
    font-weight: var(--made-font-weight-regular);
    margin-inline-start: var(--made-size-space-3-x);
    user-select: none;
  }

  .made-c-form-control--has-supporting-text .made-c-radio__label-container {
    display: flex;
    flex-direction: column;
  }

  .made-c-form-control--has-supporting-text
    .made-c-radio__label-container
    .made-c-form-control__supporting-text {
    margin-left: var(--made-size-space-3-x);
    font-weight: var(--made-font-weight-regular);
  }

  /** Error */

  .made-c-radio--error .made-c-radio__control {
    border-color: var(--made-color-border-danger-strong);
  }

  /** Error + checked */
  .made-c-radio--error.made-c-radio--checked .made-c-radio__control,
  .made-c-radio--error .made-c-radio__control:checked {
    border-color: var(--made-color-border-danger-strong);
    box-shadow: inset 0 0 0 var(--stroke) var(--made-color-border-danger-strong);
  }
`;
