// @mc-made/web-components/dist/react/made-radio-group/index.js
import * as React from 'react';
import { createComponent } from '../create-component.js';
import Component from '../../components/radio-group/src/radio-group.component.js';
const tagName = 'made-radio-group';
Component.define(tagName);
/**
 * @summary Radio groups are used to group multiple radio buttons together so they function as a single form control.
 * @documentation '../../components/radio-group/stories/radio-group.mdx'
 *
 * @slot - The default slot.
 * @slot supporting-text - Text that describes the radio group. Alternatively, you can use the `supporting-text` attribute.
 *
 * @csspart form-control - The component's base wrapper.
 * @csspart form-control-label - The component's label.
 * @csspart form-control-input - The default slot wrapper.
 * @csspart form-control-supporting-text - The supporting text's wrapper.
 *
 */
const reactWrapper = createComponent({
    tagName,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change'
    },
    displayName: 'MadeRadioGroup'
});
export default reactWrapper;
