// @mc-made/web-components/dist/utilities/localize.js
// ---------------------------------------------------------------------------
// Translation registry
// ---------------------------------------------------------------------------
/** All registered translations, keyed by lower-cased BCP 47 code. */
const translationRegistry = new Map();
/**
 * Registers a translation so it can be resolved by `LocalizeController`.
 *
 * Call this once per translation file — typically at the end of each
 * `packages/translations/*.ts` file as a side-effect import.
 *
 * @example
 * ```ts
 * // packages/translations/fr.ts
 * import { registerTranslation } from '../utilities/localize.js';
 * const translation: Translation = { $code: 'fr', ... };
 * registerTranslation(translation);
 * export default translation;
 * ```
 */
export function registerTranslation(translation) {
    translationRegistry.set(translation.$code.toLowerCase(), translation);
}
/**
 * Resolves the best matching `Translation` for the given BCP 47 locale string.
 *
 * Resolution order:
 * 1. Exact match (e.g. 'en-GB')
 * 2. Language-prefix match (e.g. 'en' for 'en-GB')
 * 3. English fallback ('en')
 * 4. First registered translation
 */
function resolveTranslation(locale) {
    const code = locale.toLowerCase();
    // 1. Exact match
    if (translationRegistry.has(code)) {
        return translationRegistry.get(code);
    }
    // 2. Language-prefix match (e.g. 'en' from 'en-GB')
    const prefix = code.split('-')[0];
    if (translationRegistry.has(prefix)) {
        return translationRegistry.get(prefix);
    }
    // 3. English fallback
    if (translationRegistry.has('en')) {
        return translationRegistry.get('en');
    }
    // 4. First registered
    return translationRegistry.values().next().value;
}
// ---------------------------------------------------------------------------
// LocalizeController
// ---------------------------------------------------------------------------
/**
 * A Lit `ReactiveController` that resolves translated strings for a component
 * based on the nearest `lang` attribute in the DOM hierarchy.
 *
 * Mount it once on any `MadeElement` (already done in the base class) and use
 * `this.localize.term('key')` inside `render()`.
 *
 * The controller uses a `MutationObserver` to watch for `lang` attribute
 * changes on the host element and all its ancestors, triggering a re-render
 * whenever the active locale changes.
 *
 * @example
 * ```typescript
 * // In a component:
 * render() {
 *   return html`<div aria-label=${this.localize.term('loading')}></div>`;
 * }
 * ```
 *
 * @example Consumer/app usage:
 * ```js
 * // Side-effect import registers the translation
 * import '@mc-made/web-components/translations/fr.js';
 * // Set lang on <html> — all components re-render automatically
 * document.documentElement.lang = 'fr';
 * ```
 */
// Register English inline here to avoid a circular dependency with
// packages/translations/en.ts.
// Translation files (en.ts, fr.ts, etc.) are the canonical source of truth
// for consumers; this ensures `term()` always has a fallback even if no
// translation file is imported.
registerTranslation({
    $code: 'en',
    $name: 'English',
    $dir: 'ltr',
    clearEntry: 'Clear entry',
    close: 'Close',
    hidePassword: 'Hide password',
    loading: 'Loading',
    progress: 'Progress',
    selectDate: 'Select date',
    selectTime: 'Select time',
    showPassword: 'Show password'
});
export class LocalizeController {
    constructor(host) {
        this.host = host;
        host.addController(this);
    }
    hostConnected() {
        // Watch for lang attribute changes anywhere in the ancestor chain.
        this.observer = new MutationObserver(() => this.host.requestUpdate());
        // Observe the host element itself.
        this.observer.observe(this.host, {
            attributes: true,
            attributeFilter: ['lang']
        });
        // Walk up and observe every ancestor element + the document element.
        let node = this.host.parentElement;
        while (node) {
            this.observer.observe(node, {
                attributes: true,
                attributeFilter: ['lang']
            });
            node = node.parentElement;
        }
        // Also watch the document element in case <html lang="..."> changes.
        if (!this.host.parentElement) {
            this.observer.observe(document.documentElement, {
                attributes: true,
                attributeFilter: ['lang']
            });
        }
    }
    hostDisconnected() {
        this.observer?.disconnect();
        this.observer = undefined;
    }
    /**
     * Walks the DOM from the host element upward to find the nearest `lang`
     * attribute, then falls back to `document.documentElement.lang` and
     * finally to `'en'`.
     */
    resolveLocale() {
        // Check the host's own lang property first (set via Lit @property)
        if (this.host.lang) {
            return this.host.lang;
        }
        // Walk up the DOM tree
        let node = this.host.parentElement;
        while (node) {
            const lang = node.getAttribute('lang');
            if (lang)
                return lang;
            node = node.parentElement;
        }
        // Fall back to <html lang>, then the browser/OS preferred language, then 'en'.
        return document.documentElement.lang || navigator.language || 'en';
    }
    /**
     * Returns the translated string for the given key in the currently active locale.
     *
     * If the key is a function (for dynamic/plural terms), call it with the
     * supplied arguments.
     *
     * Falls back to English, then to the raw key name if nothing is found.
     *
     * @param key - A key from the `Translation` interface
     * @param args - Arguments forwarded to function-typed translation values
     */
    term(key, ...args) {
        const locale = this.resolveLocale();
        const translation = resolveTranslation(locale);
        if (!translation) {
            return String(key);
        }
        const value = translation[key];
        if (typeof value === 'function') {
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            return value(...args);
        }
        return String(value ?? key);
    }
    /**
     * Returns the text direction (`'ltr'` or `'rtl'`) for the active locale.
     * Defaults to `'ltr'` when no translation is found.
     */
    dir() {
        const locale = this.resolveLocale();
        return resolveTranslation(locale)?.$dir ?? 'ltr';
    }
}
