// @mc-made/web-components/dist/components/accordion-item/src/accordion-item.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/src/icon.component.js';
import styles from './accordion-item.styles.js';
/**
 * @summary An accordion item is an individual section within an accordion component that can be expanded or collapsed.
 * @documentation '../stories/accordion-item.mdx'
 *
 * @dependency made-icon
 *
 * @fire mde-toggle - Emitted when the accordion item is toggled.
 * @fire mde-show - Emitted when the accordion item is opened.
 * @fire mde-hide - Emitted when the accordion item is closed.
 *
 * @slot - The accordion item content that will be shown/hidden.
 * @slot badge - Used to place a badge in the header area.
 * @slot header-start - An optional leading icon in the header.
 * @slot header-end - An optional trailing icon in the header.
 *
 * @csspart base - The component's base wrapper.
 * @csspart trigger - The clickable trigger area that contains header and subtitle.
 * @csspart header - The header row containing title and icons.
 * @csspart header-content - The container for the header content.
 * @csspart subtitle - The subtitle container.
 * @csspart subtitle-content - The inner wrapper for subtitle content.
 * @csspart icon - The chevron icon that indicates expanded/collapsed state.
 * @csspart content - The collapsible content container.
 */
class MadeAccordionItem extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'badge', 'header-start', 'header-end');
        /** The title text for the accordion item. */
        this.header = '';
        /** The subtitle text for the accordion item. */
        this.subtitle = '';
        /** Controls whether the accordion item is expanded. */
        this.expanded = false;
        /** Disables the accordion item, preventing expansion/collapse. */
        this.disabled = false;
        /** A unique value to identify this accordion item. */
        this.value = '';
        /** The appearance style of the accordion item. */
        this.appearance = 'default';
        this.headerId = `accordion-header-${Math.random().toString(36).substr(2, 9)}`;
        this.contentId = `accordion-content-${Math.random().toString(36).substr(2, 9)}`;
        this._animation = null;
        this._hasCompletedFirstUpdate = false;
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', 'region');
        // Generate a unique value if none provided
        if (!this.value) {
            this.value = `accordion-item-${Math.random().toString(36).substr(2, 9)}`;
        }
    }
    firstUpdated() {
        // Set aria-label with the header text for accessibility
        // Using aria-label instead of aria-labelledby because the host element
        // cannot reference elements in its own shadow DOM
        this.updateAriaLabel();
        // If initially expanded, expose full height so content is visible without animation
        if (this.expanded) {
            this.contentElement.style.height = 'auto';
        }
    }
    updated(changedProperties) {
        if (changedProperties.has('header')) {
            this.updateAriaLabel();
        }
        if (!this._hasCompletedFirstUpdate) {
            this._hasCompletedFirstUpdate = true;
            return;
        }
        if (changedProperties.has('expanded')) {
            if (this.expanded) {
                this._cancelAnimation();
                void this._animateShow();
            }
            else {
                this._cancelAnimation();
                void this._animateHide();
            }
        }
    }
    updateAriaLabel() {
        // Set aria-label to the header text to provide an accessible name for the region
        if (this.header) {
            this.setAttribute('aria-label', this.header);
        }
    }
    /** Shows the accordion item content. */
    show() {
        if (this.disabled || this.expanded)
            return;
        this.expanded = true;
        this.emit('mde-toggle', {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: { value: this.value, expanded: true }
        });
        this.emit('mde-show', { bubbles: true, composed: true });
    }
    /** Hides the accordion item content. */
    hide() {
        if (this.disabled || !this.expanded)
            return;
        this.expanded = false;
        this.emit('mde-toggle', {
            bubbles: true,
            composed: true,
            cancelable: false,
            detail: { value: this.value, expanded: false }
        });
        this.emit('mde-hide', { bubbles: true, composed: true });
    }
    /** Toggles the accordion item between expanded and collapsed. */
    toggle() {
        if (this.disabled)
            return;
        if (this.expanded) {
            this.hide();
        }
        else {
            this.show();
        }
    }
    _cancelAnimation() {
        if (this._animation) {
            try {
                this._animation.commitStyles();
            }
            catch {
                /* ignored */
            }
            this._animation.cancel();
            this._animation = null;
        }
    }
    async _animateShow() {
        const el = this.contentElement;
        // Read the start height (may be mid-close value from a cancelled animation)
        const fromH = el.style.height || '0px';
        // Measure full open height
        el.style.height = 'auto';
        const toH = `${el.getBoundingClientRect().height}px`;
        el.style.height = fromH;
        el.getBoundingClientRect(); // force reflow
        const anim = el.animate([{ height: fromH }, { height: toH }], {
            duration: this._animDuration,
            easing: this._animEasing,
            fill: 'both'
        });
        this._animation = anim;
        try {
            await anim.finished;
        }
        catch {
            return; // cancelled
        }
        if (this._animation !== anim)
            return; // superseded by another call
        if (!this.isConnected)
            return; // element removed from DOM before animation finished
        anim.commitStyles();
        anim.cancel();
        this._animation = null;
        el.style.height = 'auto';
    }
    async _animateHide() {
        const el = this.contentElement;
        // Measure the real pixel height — handles both 'auto' and an explicit value
        const fromH = `${el.getBoundingClientRect().height}px`;
        el.style.height = fromH;
        el.getBoundingClientRect(); // force reflow
        const anim = el.animate([{ height: fromH }, { height: '0px' }], {
            duration: this._animDuration,
            easing: this._animEasing,
            fill: 'both'
        });
        this._animation = anim;
        try {
            await anim.finished;
        }
        catch {
            return; // cancelled
        }
        if (this._animation !== anim)
            return; // superseded
        if (!this.isConnected)
            return; // element removed from DOM before animation finished
        anim.commitStyles();
        anim.cancel();
        this._animation = null;
        el.style.height = '';
    }
    get _animDuration() {
        const val = getComputedStyle(this)
            .getPropertyValue('--made-motion-duration-longest')
            .trim();
        if (val.endsWith('ms'))
            return parseFloat(val);
        if (val.endsWith('s'))
            return parseFloat(val) * 1000;
        return parseFloat(val) || 300;
    }
    get _animEasing() {
        return (getComputedStyle(this)
            .getPropertyValue('--made-motion-easing-emphasised-ease-out')
            .trim() || 'ease');
    }
    handleHeaderClick() {
        if (this.disabled)
            return;
        if (this.expanded) {
            this.hide();
        }
        else {
            this.show();
        }
    }
    handleKeyDown(event) {
        if (this.disabled)
            return;
        // Handle Enter and Space keys
        if (event.key === 'Enter' || event.key === ' ') {
            event.preventDefault();
            this.handleHeaderClick();
        }
    }
    render() {
        const hasHeaderStartSlot = this.hasSlotController.test('header-start');
        const hasHeaderEndSlot = this.hasSlotController.test('header-end');
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-accordion': true,
            'made-c-accordion--contained': this.appearance === 'contained',
            'made-c-accordion--expanded': this.expanded,
            'made-c-accordion--disabled': this.disabled
        })}
      >
        <div
          part="trigger"
          class="made-c-accordion__trigger"
          @click=${this.handleHeaderClick}
          @keydown=${this.handleKeyDown}
          tabindex=${this.disabled ? '-1' : '0'}
          role="button"
          aria-expanded=${this.expanded ? 'true' : 'false'}
          aria-disabled=${this.disabled ? 'true' : 'false'}
          aria-controls=${this.contentId}
        >
          <div part="header" class="made-c-accordion__header">
            <slot
              name="header-start"
              class=${classMap({
            'made-c-accordion__header-start': hasHeaderStartSlot
        })}
            ></slot>
            <div
              part="header-content"
              class="made-c-accordion__header-content"
              id=${this.headerId}
            >
              ${this.header}
            </div>
            <slot
              name="header-end"
              class=${classMap({
            'made-c-accordion__header-end': hasHeaderEndSlot
        })}
            ></slot>
            <div class="made-c-accordion__actions">
              <slot name="badge"></slot>
              <made-icon
                part="icon"
                size="medium"
                class="made-c-accordion__chevron"
                name="chevron-down"
                library="system"
                aria-hidden="true"
              ></made-icon>
            </div>
          </div>
          ${this.subtitle
            ? html `
                <div part="subtitle" class="made-c-accordion__subtitle">
                  <div
                    part="subtitle-content"
                    class="made-c-accordion__subtitle-content"
                  >
                    ${this.subtitle}
                  </div>
                </div>
              `
            : ''}
        </div>
        <div
          part="content"
          class="made-c-accordion__content"
          id=${this.contentId}
          aria-hidden=${this.expanded ? 'false' : 'true'}
        >
          <slot></slot>
        </div>
      </div>
    `;
    }
}
MadeAccordionItem.styles = [componentStyles, styles];
MadeAccordionItem.dependencies = { 'made-icon': MadeIcon };
export default MadeAccordionItem;
__decorate([
    query('.made-c-accordion__trigger')
], MadeAccordionItem.prototype, "headerElement", void 0);
__decorate([
    query('.made-c-accordion__content')
], MadeAccordionItem.prototype, "contentElement", void 0);
__decorate([
    property()
], MadeAccordionItem.prototype, "header", void 0);
__decorate([
    property()
], MadeAccordionItem.prototype, "subtitle", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeAccordionItem.prototype, "expanded", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeAccordionItem.prototype, "disabled", void 0);
__decorate([
    property()
], MadeAccordionItem.prototype, "value", void 0);
__decorate([
    property({ reflect: true })
], MadeAccordionItem.prototype, "appearance", void 0);
