// @mc-made/web-components/dist/components/accordion-item/src/accordion-item.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: block;
  }

  .made-c-accordion {
    font-family: var(--made-font-family-primary);
    background-color: var(--made-color-background-surface-default);
    border-radius: var(--made-border-radius-none);
    border-bottom: 1px solid
      var(--made-component-accordion-contained-border-color);
    max-width: 100%;
  }

  /* Trigger - the clickable area containing header and subtitle */
  .made-c-accordion__trigger {
    display: block;
    cursor: pointer;
    font-family: var(--made-font-family-primary);
    border-radius: var(--made-border-radius-none);
    padding: var(--made-component-accordion-padding-y)
      var(--made-component-accordion-padding-x);
    width: 100%;
    outline: none;
    position: relative;
    overflow: hidden;
    transition:
      background-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease),
      color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease);

    .made-c-accordion--contained & {
      border-radius: var(--made-component-accordion-contained-border-radius);
    }
  }

  .made-c-accordion--contained {
    border: var(--made-component-accordion-contained-border-width, 1px) solid
      var(--made-component-accordion-contained-border-color);
    border-radius: var(--made-component-accordion-contained-border-radius);
  }

  /* Hover state */
  .made-c-accordion__trigger:hover,
  .made-c-accordion__trigger:active {
    background-color: var(--made-color-background-surface-hover);
    color: var(--made-color-text-default);
  }

  .made-c-accordion:not(.made-c-accordion--disabled)
    .made-c-accordion__trigger:hover
    .made-c-accordion__chevron,
  .made-c-accordion:not(.made-c-accordion--disabled)
    .made-c-accordion__trigger:hover
    .made-c-accordion__header-start,
  .made-c-accordion:not(.made-c-accordion--disabled)
    .made-c-accordion__trigger:hover
    .made-c-accordion__header-end {
    color: var(--made-color-icon-primary);
  }

  /* Focus state - using box-shadow for rounded corners */
  .made-c-accordion__trigger:focus-visible {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: 2px;
  }

  /* Disabled state */
  .made-c-accordion--disabled .made-c-accordion__trigger {
    cursor: not-allowed;
    background-color: var(--made-color-background-surface-default);
    color: var(--made-color-text-disabled);
  }

  .made-c-accordion--disabled .made-c-accordion__trigger:hover {
    background-color: var(--made-color-background-surface-default);
    color: var(--made-color-text-disabled);
  }

  .made-c-accordion--disabled .made-c-accordion__chevron,
  .made-c-accordion--disabled .made-c-accordion__header-start,
  .made-c-accordion--disabled .made-c-accordion__header-end {
    color: var(--made-color-icon-disabled);
  }

  .made-c-accordion--disabled .made-c-accordion__header-content {
    color: var(--made-color-text-disabled);
    font-family: var(--made-font-family-primary);
    font-size: var(--made-component-accordion-title-font-size);
    font-style: normal;
    font-weight: var(--made-component-accordion-title-font-weight);
    line-height: var(--made-component-accordion-title-line-height);
    letter-spacing: var(--made-component-accordion-title-letter-spacing);
  }

  .made-c-accordion--disabled .made-c-accordion__subtitle {
    color: var(--made-color-text-disabled);
  }

  .made-c-accordion__header,
  .made-c-accordion__subtitle {
    display: flex;
    align-items: center;
    justify-content: space-between;
    color: var(--made-color-text-primary);
    text-align: left;
    font-family: inherit;
    font-size: inherit;
  }

  .made-c-accordion__subtitle {
    padding-top: 0px;
  }

  .made-c-accordion__header-content {
    font-size: var(--made-component-accordion-title-font-size);
    font-style: normal;
    font-weight: var(--made-component-accordion-title-font-weight);
    line-height: var(--made-component-accordion-title-line-height);
    letter-spacing: var(--made-component-accordion-title-letter-spacing);
  }

  .made-c-accordion__subtitle-content {
    flex: 1;
    font-size: var(--made-component-accordion-subtitle-font-size);
    font-style: normal;
    font-weight: var(--made-component-accordion-subtitle-font-weight);
    line-height: var(--made-component-accordion-subtitle-line-height);
    letter-spacing: var(--made-letter-spacing-default);
    margin-block-start: var(--made-size-space-1-x);
  }

  .made-c-accordion__actions {
    display: flex;
    align-items: center;
    margin-left: auto;
    gap: var(--made-size-space-2-x, 0.5rem);
  }

  .made-c-accordion__chevron {
    color: var(--made-color-icon-primary);
    flex-shrink: 0;
    transition: transform var(--made-motion-duration-longest)
      var(--made-motion-easing-emphasised-ease-out);
  }

  .made-c-accordion--expanded .made-c-accordion__chevron {
    transform: rotate(180deg);
  }

  .made-c-accordion__header-start,
  .made-c-accordion__header-end {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: var(--made-color-icon-primary);
  }

  .made-c-accordion__header-start {
    margin-inline-end: var(--made-size-space-1-x);
  }

  .made-c-accordion__header-end {
    margin-inline-start: var(--made-size-space-1-x);
  }

  .made-c-accordion__content {
    display: block;
    color: var(--made-color-text-secondary);

    /* body/regular/small */
    font-family: inherit;
    font-size: var(--made-font-size-body-small);
    font-style: normal;
    font-weight: var(--made-font-weight-regular);
    line-height: var(--made-line-height-rem-02);
    letter-spacing: var(--made-letter-spacing-default);
    overflow: hidden;
    padding-inline: var(--made-component-accordion-padding-x);
    height: 0; /* JS animation manages transitions; auto is set inline after open */
  }
`;
