// @mc-made/web-components/dist/components/tag/src/tag.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: inline-flex;
  }

  .made-c-tag {
    --_tag-height: 26px;
    --_tag-inline-space: var(--made-size-space-1point5-x);
    display: inline-flex;
    align-items: center;
    justify-content: center;
    column-gap: var(--made-component-tag-gap);
    font-family: var(--made-font-family-primary);
    font-size: var(--made-component-tag-medium-font-size);
    font-weight: var(--made-component-tag-medium-font-weight);
    line-height: var(--made-component-tag-medium-line-height);
    border-radius: var(--made-component-tag-border-radius);
    padding-block: var(--made-size-space-0-x);
    padding-inline: var(--_tag-inline-space);
    min-height: var(--_tag-height);
    border: 1px solid transparent;
    box-shadow: var(--made-component-tag-box-shadow);
  }

  /* Size variants */
  .made-c-tag--small {
    --_tag-height: 20px;
    --_tag-inline-space: var(--made-size-space-1-x);
    font-size: var(--made-component-tag-small-font-size);
    font-weight: var(--made-component-tag-small-font-weight);
    line-height: var(--made-component-tag-small-line-height);
  }

  /* Sentiment variants - High Contrast (default) */
  .made-c-tag--neutral {
    border-color: var(--made-color-border-neutral-strong);
    background: var(--made-color-background-feedback-neutral-strong-default);
    color: var(--made-color-text-inverse);
  }

  .made-c-tag--brand {
    border-color: var(--made-color-border-brand);
    background: var(--made-color-background-brand-primary-default);
    color: var(--made-color-text-inverse);
  }

  .made-c-tag--success {
    border-color: var(--made-color-border-success-strong);
    background: var(--made-color-background-feedback-success-strong-default);
    color: var(--made-color-text-on-color);
  }

  .made-c-tag--danger {
    border-color: var(--made-color-border-danger-strong);
    background: var(--made-color-background-feedback-danger-strong-default);
    color: var(--made-color-text-on-color);
  }

  .made-c-tag--warning {
    border-color: var(--made-color-border-warning-strong);
    background: var(--made-color-background-feedback-warning-strong-default);
    color: var(--made-component-tag-warning-contrast-high-text-color);
  }

  .made-c-tag--information {
    border-color: var(--made-color-border-information-strong);
    background: var(
      --made-color-background-feedback-information-strong-default
    );
    color: var(--made-color-text-on-color);
  }

  /* Sentiment variants - Low Contrast */
  .made-c-tag--neutral.made-c-tag--low-contrast {
    border-color: var(--made-component-tag-border-color);
    background: var(--made-color-background-surface-secondary);
    color: var(--made-color-text-primary);
  }

  .made-c-tag--success.made-c-tag--low-contrast {
    border-color: var(--made-color-border-success);
    background: var(--made-color-background-feedback-success-default);
    color: var(--made-color-text-success);
  }

  .made-c-tag--danger.made-c-tag--low-contrast {
    border-color: var(--made-color-border-danger);
    background: var(--made-color-background-feedback-danger-default);
    color: var(--made-color-text-danger);
  }

  .made-c-tag--warning.made-c-tag--low-contrast {
    border-color: var(--made-color-border-warning);
    background: var(--made-color-background-feedback-warning-default);
    color: var(--made-color-text-warning);
  }

  .made-c-tag--information.made-c-tag--low-contrast {
    border-color: var(--made-color-border-information);
    background: var(--made-color-background-feedback-information-default);
    color: var(--made-color-text-information);
  }

  /* Content wrapper */
  .made-c-tag [part='content'] {
    display: inline-flex;
    align-items: center;
  }

  /* Slot spacing - icons inherit color from parent */
  .made-c-tag--has-start ::slotted([slot='start']) {
    display: inline-flex;
    align-items: center;
    color: currentColor;
  }

  .made-c-tag--has-end ::slotted([slot='end']) {
    display: inline-flex;
    align-items: center;
    color: currentColor;
  }

  /* Dismissible variant - always neutral low contrast */
  .made-c-tag--dismissible {
    border-color: var(--made-component-tag-border-color);
    background: var(--made-color-background-surface-secondary);
    color: var(--made-color-text-primary);
  }

  /* Dismiss button styling */
  .made-c-tag .made-c-tag__dismiss-button {
    margin-right: calc(var(--_tag-inline-space) * -1);
  }

  .made-c-tag .made-c-tag__dismiss-button::part(base) {
    height: var(--_tag-height);
    width: var(--_tag-height);
    outline-offset: -2px;
  }

  .made-c-tag .made-c-tag__dismiss-button::part(base):hover {
    background-color: var(--made-color-background-surface-selected-hover);
  }

  .made-c-tag .made-c-tag__dismiss-button::part(icon) {
    font-size: var(--made-font-size-07);
  }
`;
