// @mc-made/web-components/dist/components/tag/src/tag.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIconButton from '../../icon-button/src/icon-button.component.js';
import styles from './tag.styles.js';
/**
 * @summary Tags categorize and organize items using descriptive keywords for quick recognition and navigation.
 * @documentation '../stories/tag.mdx'
 *
 * @dependency made-icon-button
 *
 * @fires mde-dismiss {CustomEvent} - Fired when the dismiss button is clicked (dismissible tags only).
 *
 * @slot - The tag's main content (required).
 * @slot start - Used to place content before the tag text. Typically used for icons or avatars.
 * @slot end - Used to place content after the tag text. Typically used for icons.
 *
 * @csspart base - The component's base wrapper.
 * @csspart content - The container that wraps the default slot content.
 * @csspart dismiss-button - The dismiss button (when dismissible).
 */
class MadeTag extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'start', 'end');
        /** Alter the contrast of the sentiment attribute. */
        this.contrast = 'high';
        /** Whether the tag can be dismissed. */
        this.isDismissible = false;
        /** The tag's sentiment defines its color. */
        this.sentiment = 'neutral';
        /** Size of the tag. */
        this.size = 'medium';
    }
    /** Returns the active sentiment, overridden to 'neutral' when dismissible. */
    get activeSentiment() {
        return this.isDismissible ? 'neutral' : this.sentiment;
    }
    /** Returns the active size, overridden to 'medium' when dismissible. */
    get activeSize() {
        return this.isDismissible ? 'medium' : this.size;
    }
    /** Returns the active contrast, overridden to 'low' when dismissible. */
    get activeContrast() {
        return this.isDismissible ? 'low' : this.contrast;
    }
    /** Handles the dismiss button click event. */
    handleDismissClick() {
        this.emit('mde-dismiss');
    }
    render() {
        return html `
      <span
        part="base"
        class=${classMap({
            'made-c-tag': true,
            'made-c-tag--neutral': this.activeSentiment === 'neutral',
            'made-c-tag--success': this.activeSentiment === 'success',
            'made-c-tag--danger': this.activeSentiment === 'danger',
            'made-c-tag--warning': this.activeSentiment === 'warning',
            'made-c-tag--information': this.activeSentiment === 'information',
            'made-c-tag--brand': this.activeSentiment === 'brand',
            'made-c-tag--small': this.activeSize === 'small',
            'made-c-tag--low-contrast': this.activeContrast === 'low',
            'made-c-tag--dismissible': this.isDismissible,
            'made-c-tag--has-start': this.hasSlotController.test('start'),
            'made-c-tag--has-end': this.hasSlotController.test('end')
        })}
      >
        <slot name="start"></slot>
        <span part="content">
          <slot></slot>
        </span>
        <slot name="end"></slot>
        ${this.isDismissible
            ? html `
              <made-icon-button
                class="made-c-tag__dismiss-button"
                part="dismiss-button"
                variant="ghost"
                icon-name="navigation/outline/x-mark-outline"
                aria-label="dismiss tag"
                @click=${this.handleDismissClick}
              ></made-icon-button>
            `
            : ''}
      </span>
    `;
    }
}
MadeTag.styles = [componentStyles, styles];
MadeTag.dependencies = {
    'made-icon-button': MadeIconButton
};
export default MadeTag;
__decorate([
    property({ reflect: true })
], MadeTag.prototype, "contrast", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'is-dismissible' })
], MadeTag.prototype, "isDismissible", void 0);
__decorate([
    property({ reflect: true })
], MadeTag.prototype, "sentiment", void 0);
__decorate([
    property({ reflect: true })
], MadeTag.prototype, "size", void 0);
