// @mc-made/web-components/dist/components/accordion/src/accordion.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeAccordionItem from '../../accordion-item/src/accordion-item.component.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './accordion.styles.js';
/**
 * @summary An accordion is a vertically stacked list of headers that reveal or hide associated sections of content.
 * @documentation '../stories/accordion.mdx'
 *
 * @dependency made-accordion-item
 *
 * @fire mde-show - Emitted when an accordion item is expanded.
 * @fire mde-hide - Emitted when an accordion item is collapsed.
 *
 * @slot - One or more made-accordion-item elements.
 *
 * @csspart base - The component's base wrapper.
 */
class MadeAccordion extends MadeElement {
    constructor() {
        super(...arguments);
        /** A convenience property to set the appearance style of each accordion item. */
        this.appearance = 'default';
        /** When true, only one accordion item can be expanded at a time. */
        this.exclusive = false;
        /** A convenience property to disable all accordion items. */
        this.disabled = false;
        this.handleKeyDown = (event) => {
            const target = event.target;
            // Only handle keyboard events from accordion items
            if (target.tagName.toLowerCase() !== 'made-accordion-item') {
                return;
            }
            const items = this.items || [];
            const currentIndex = items.indexOf(target);
            if (currentIndex === -1)
                return;
            let nextIndex = currentIndex;
            switch (event.key) {
                case 'ArrowDown':
                    event.preventDefault();
                    nextIndex = currentIndex + 1;
                    if (nextIndex >= items.length)
                        nextIndex = 0;
                    break;
                case 'ArrowUp':
                    event.preventDefault();
                    nextIndex = currentIndex - 1;
                    if (nextIndex < 0)
                        nextIndex = items.length - 1;
                    break;
                case 'Home':
                    event.preventDefault();
                    nextIndex = 0;
                    break;
                case 'End':
                    event.preventDefault();
                    nextIndex = items.length - 1;
                    break;
                default:
                    return;
            }
            // Focus the next item's header
            if (items[nextIndex]?.headerElement) {
                items[nextIndex].headerElement.focus();
            }
        };
    }
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', 'presentation');
        this.addEventListener('mde-toggle', this.handleItemToggle);
        this.addEventListener('keydown', this.handleKeyDown);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('mde-toggle', this.handleItemToggle);
        this.removeEventListener('keydown', this.handleKeyDown);
    }
    firstUpdated() {
        this.syncItemStates();
    }
    updated(changedProperties) {
        if (changedProperties.has('disabled') ||
            changedProperties.has('appearance')) {
            this.syncItemStates();
        }
    }
    /** Expands all accordion items (only works when exclusive is false). */
    expandAll() {
        if (this.exclusive)
            return;
        this.items?.forEach(item => item.show());
    }
    /** Collapses all accordion items. */
    collapseAll() {
        this.items?.forEach(item => item.hide());
    }
    syncItemStates() {
        this.items?.forEach(item => {
            item.appearance = this.appearance;
            item.disabled = this.disabled;
        });
    }
    handleItemToggle(event) {
        const item = event.target;
        if (this.exclusive && item.expanded) {
            // Close all other items when in exclusive mode
            this.items?.forEach(otherItem => {
                if (otherItem !== item && otherItem.expanded) {
                    otherItem.hide();
                }
            });
        }
        // Re-emit the event from the accordion
        if (item.expanded) {
            this.emit('mde-show', {
                bubbles: true,
                detail: { value: item.value }
            });
        }
        else {
            this.emit('mde-hide', {
                bubbles: true,
                detail: { value: item.value }
            });
        }
    }
    render() {
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-accordion-container': true,
            'made-c-accordion-container--contained': this.appearance === 'contained'
        })}
      >
        <slot @slotchange=${this.syncItemStates}></slot>
      </div>
    `;
    }
}
MadeAccordion.styles = [componentStyles, styles];
MadeAccordion.dependencies = { 'made-accordion-item': MadeAccordionItem };
export default MadeAccordion;
__decorate([
    queryAssignedElements({ selector: 'made-accordion-item' })
], MadeAccordion.prototype, "items", void 0);
__decorate([
    property({ reflect: true })
], MadeAccordion.prototype, "appearance", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeAccordion.prototype, "exclusive", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeAccordion.prototype, "disabled", void 0);
