// @mc-made/web-components/dist/components/avatar/src/avatar.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  :host {
    display: inline-block;
  }

  .made-c-avatar {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    position: relative;
    background-color: var(--made-color-background-surface-selected);
    border-radius: var(--made-border-radius-full);
    font-family: var(--made-font-family-primary);
    font-weight: var(--made-font-weight-medium);
    text-transform: var(--made-text-case-uppercase);
    color: var(--made-color-text-primary);
    box-shadow: var(--made-component-avatar-box-shadow);
    user-select: none;
    vertical-align: middle;
    overflow: hidden;
  }

  .made-c-avatar--xs {
    height: var(--made-size-space-6-x);
    width: var(--made-size-space-6-x);
    font-size: var(--made-font-size-01);
  }

  .made-c-avatar--sm {
    height: var(--made-size-space-8-x);
    width: var(--made-size-space-8-x);
    font-size: var(--made-font-size-02);
  }

  .made-c-avatar--md {
    height: var(--made-size-space-10-x);
    width: var(--made-size-space-10-x);
    font-size: var(--made-font-size-02);
  }

  .made-c-avatar--lg {
    height: var(--made-size-space-12-x);
    width: var(--made-size-space-12-x);
    font-size: var(--made-font-size-06);
  }

  .made-c-avatar--xl {
    height: var(--made-size-space-20-x);
    width: var(--made-size-space-20-x);
    font-size: var(--made-font-size-11);
  }

  .made-c-avatar--xxl {
    height: var(--made-size-space-36-x);
    width: var(--made-size-space-36-x);
    font-size: var(--made-font-size-13);
  }

  .made-c-avatar__img {
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    object-fit: cover;
    overflow: hidden;
  }

  .made-c-avatar--circle,
  .made-c-avatar__img {
    border-radius: 50%;
  }

  .made-c-avatar--squared {
    border-radius: var(--made-component-avatar-squared-border-radius);
  }

  .made-c-avatar__icon {
    display: flex;
    align-items: center;
    justify-content: center;
    position: absolute;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    color: var(--made-color-icon-helper);
  }

  .made-c-avatar made-icon {
    --width: 65%;
    --height: 65%;
  }

  .made-c-avatar__img--icon {
    width: var(--width, 65%);
    height: var(--height, 65%);
  }
`;
