// @mc-made/web-components/dist/components/avatar/src/avatar.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/made-icon.js';
import styles from './avatar.styles.js';
class MadeAvatar extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasError = false;
        /** Shape of the avatar. */
        this.shape = 'circle';
        /** Size of the avatar. */
        this.size = 'medium';
        /** Initials to use as a fallback when no image is available (1-2 characters max recommended). */
        this.initials = '';
        /** Source of the avatar's image. */
        this.image = '';
        // @property() initials =
        /** Use this to describe the avatar for assistive technologies. */
        this.alt = '';
        /** Instruct the browser how to load the image. */
        this.loading = 'eager';
    }
    render() {
        const avatarWithImage = html `
      <img
        part="image"
        class="made-c-avatar__img"
        src=${this.image}
        loading=${this.loading}
        alt=${this.alt}
        @error=${() => (this.hasError = true)}
      />
    `;
        const avatarWithoutImage = this.initials
            ? html ` <div part="initials">${this.initials}</div> `
            : html `
          <div part="icon" aria-hidden="true">
            <slot name="icon" class="made-c-avatar__icon">
              <made-icon library="system" name="user-solid"></made-icon>
            </slot>
          </div>
        `;
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-avatar': true,
            'made-c-avatar--circle': this.shape === 'circle',
            'made-c-avatar--squared': this.shape === 'squared',
            'made-c-avatar--xs': this.size === 'x-small',
            'made-c-avatar--sm': this.size === 'small',
            'made-c-avatar--md': this.size === 'medium',
            'made-c-avatar--lg': this.size === 'large',
            'made-c-avatar--xl': this.size === 'x-large',
            'made-c-avatar--xxl': this.size === 'xx-large'
        })}
        role="img"
        aria-label=${this.alt}
      >
        ${this.image && !this.hasError ? avatarWithImage : avatarWithoutImage}
        <slot></slot>
      </div>
    `;
    }
}
MadeAvatar.styles = [componentStyles, styles];
MadeAvatar.dependencies = {
    'made-icon': MadeIcon
};
export default MadeAvatar;
__decorate([
    state()
], MadeAvatar.prototype, "hasError", void 0);
__decorate([
    property({ reflect: true })
], MadeAvatar.prototype, "shape", void 0);
__decorate([
    property({ reflect: true })
], MadeAvatar.prototype, "size", void 0);
__decorate([
    property()
], MadeAvatar.prototype, "initials", void 0);
__decorate([
    property()
], MadeAvatar.prototype, "image", void 0);
__decorate([
    property()
], MadeAvatar.prototype, "alt", void 0);
__decorate([
    property()
], MadeAvatar.prototype, "loading", void 0);
