// @mc-made/web-components/dist/components/toggle/src/toggle.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { CheckboxValidator } from '../../../internal/face/validators/checkbox-validator.js';
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormState, getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { live } from 'lit/directives/live.js';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './toggle.styles.js';
/**
 * @summary Toggle is used to turn an item on or off.
 * @documentation '../stories/toggle.mdx'
 *
 * @slot (default) - the toggle's label
 * @slot supporting-text - the toggle's supporting text
 *
 * @fires change {Event} - Fired whenever `checked` changes due to user interaction.
 * @fires input {Event} - Fired whenever `checked` changes due to user interaction.
 *
 * @csspart form-control - The outer wrapper.
 * @csspart input-container - The container that wraps the control, label and supporting text.
 * @csspart control - The control that houses the toggle's thumb.
 * @csspart thumb - The toggle's thumb.
 * @csspart label - The toggle's label.
 * @csspart supporting-text - The descriptive text's wrapper.
 *
 * @cssproperty --width - The width of the toggle.
 * @cssproperty --height - The height of the toggle.
 * @cssproperty --thumb-size - The size of the thumb.
 */
const toggleBaseClass = mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement)));
class MadeToggle extends toggleBaseClass {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'supporting-text');
        /**
         * The value associated with this toggle on form submission.
         */
        this.value = 'on';
        /**
         * When true, require the toggle to be checked when participating in
         * form submission.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#validation
         */
        this.required = false;
        /** The toggle's size. */
        this.size = 'medium';
        /**
         * Puts the toggle in the checked state and sets the form submission value to
         * the `value` property.
         */
        this.checked = false;
        /** Additional text as part of the label.  */
        this.supportingText = '';
        /** If true, the toggle will stretch to take up the full width of its container. */
        this.fullWidth = false;
        /** When true reverses the order of the component, putting the label before the toggle.  */
        this.reverse = false;
    }
    handleInput(event) {
        const target = event.target;
        this.checked = target.checked;
    }
    handleChange() {
        // dispatch 'composed' change event
        this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
    }
    handleKeyDown(event) {
        if (event.key === 'Enter') {
            event.preventDefault();
            this.click();
        }
    }
    /** Simulates a click on the toggle. */
    click() {
        this.input.click();
    }
    /** Sets focus on the toggle. */
    focus(options) {
        this.input.focus(options);
    }
    /** Removes focus from the toggle. */
    blur() {
        this.input.blur();
    }
    render() {
        const hasLabel = !!this.hasSlotController.test('[default]');
        const hasSupportingText = this.supportingText
            ? true
            : !!this.hasSlotController.test('supporting-text');
        return html `
      <div
        part="form-control"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control--small': this.size === 'small',
            'made-c-form-control--medium': this.size === 'medium',
            'made-c-form-control--has-label': hasLabel,
            'made-c-form-control--has-supporting-text': hasSupportingText
        })}
      >
        <label
          part="input-container"
          class=${classMap({
            'made-c-toggle': true,
            'made-c-toggle--checked': this.checked,
            'made-c-toggle--disabled': this.disabled,
            'made-c-toggle--small': this.size === 'small',
            'made-c-toggle--medium': this.size === 'medium',
            'made-c-toggle--full-width': this.fullWidth,
            'made-c-toggle--reverse': this.reverse
        })}
        >
          <input
            class="made-c-toggle__input"
            type="checkbox"
            tabindex=${'0' /* This ensures the Toggle is keyboard navigable in Safari */}
            name=${this.name}
            value=${ifDefined(this.value)}
            ?checked=${live(this.checked)}
            ?disabled=${this.disabled}
            ?required=${this.required}
            role="switch"
            aria-checked=${this.checked ? 'true' : 'false'}
            aria-label=${ifDefined(this.ariaLabel)}
            @change=${this.handleChange}
            @input=${this.handleInput}
            @keydown=${this.handleKeyDown}
          />

          <span part="control" class="made-c-toggle__control">
            <span part="thumb" class="made-c-toggle__thumb"></span>
          </span>

          <div part="label" class="made-c-toggle__label">
            <slot></slot>
            <div
              aria-hidden=${hasSupportingText ? 'false' : 'true'}
              class="made-c-form-control__supporting-text"
              id="supporting-text"
              part="supporting-text"
            >
              <slot name="supporting-text">${this.supportingText}</slot>
            </div>
          </div>
        </label>
      </div>
    `;
    }
    [getFormValue]() {
        return this.checked ? this.value : null;
    }
    [getFormState]() {
        return String(this.checked);
    }
    formResetCallback() {
        // The checked property does not reflect, so the original attribute set by
        // the user is used to determine the default value.
        this.checked = this.hasAttribute('checked');
    }
    formStateRestoreCallback(state) {
        this.checked = state === 'true';
    }
    // @ts-expect-error 'ts(2425)'
    [createValidator]() {
        return new CheckboxValidator(() => ({
            checked: this.checked,
            required: this.required
        }));
    }
    // @ts-expect-error 'ts(2425)'
    [getValidityAnchor]() {
        return this.input;
    }
}
MadeToggle.styles = [componentStyles, formControlStyles, styles];
export default MadeToggle;
__decorate([
    query('input')
], MadeToggle.prototype, "input", void 0);
__decorate([
    property({ attribute: 'aria-label' })
], MadeToggle.prototype, "ariaLabel", void 0);
__decorate([
    property()
], MadeToggle.prototype, "value", void 0);
__decorate([
    property({ type: Boolean })
], MadeToggle.prototype, "required", void 0);
__decorate([
    property({ reflect: true })
], MadeToggle.prototype, "size", void 0);
__decorate([
    property({ type: Boolean })
], MadeToggle.prototype, "checked", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeToggle.prototype, "supportingText", void 0);
__decorate([
    property({ type: Boolean, attribute: 'full-width', reflect: true })
], MadeToggle.prototype, "fullWidth", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeToggle.prototype, "reverse", void 0);
