// @mc-made/web-components/dist/components/toggle/src/toggle.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  :host {
    display: inline-block;
  }

  .made-c-toggle {
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: space-between;
    width: max-content;
    font-family: var(--made-font-family-primary);
    font-weight: var(--made-font-weight-regular);
    color: var(--made-color-text-primary);
    vertical-align: middle;
    cursor: pointer;
  }

  .made-c-toggle--small {
    --height: var(--made-size-space-4-x);
    --width: calc(var(--height) * 2);
    --thumb-size: calc(var(--height) - 4px);

    font-size: var(--made-font-size-02);
    line-height: var(--made-line-height-rem-02);
  }

  .made-c-toggle--medium {
    --height: var(--made-size-space-6-x);
    --width: calc(var(--height) * 2);
    --thumb-size: calc(var(--height) - 6px);

    font-size: var(--made-font-size-04);
    line-height: var(--made-line-height-rem-04);
  }

  .made-c-toggle__control {
    flex: 0 0 auto;
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    width: var(--width);
    height: var(--height);
    background-color: var(--made-component-toggle-background-off);
    border: solid 1px var(--made-component-toggle-background-off);
    box-shadow: var(--made-component-toggle-box-shadow);
    border-radius: var(--made-component-toggle-border-radius);
    transition:
      background-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease),
      border-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease);

    & .made-c-toggle__thumb {
      width: var(--thumb-size);
      height: var(--thumb-size);
      background-color: var(--made-color-background-surface-default);
      border-radius: var(--made-component-toggle-thumb-border-radius);
      border: solid 1px var(--made-color-background-surface-default);
      box-shadow: var(--made-component-toggle-thumb-box-shadow);
      translate: calc((var(--width) - var(--height)) / -2);
      transition:
        translate var(--made-motion-duration-short)
          var(--made-motion-easing-standard-ease),
        background-color var(--made-motion-duration-short)
          var(--made-motion-easing-standard-ease),
        border-color var(--made-motion-duration-short)
          var(--made-motion-easing-standard-ease),
        box-shadow var(--made-motion-duration-short)
          var(--made-motion-easing-standard-ease);
    }
  }

  .made-c-toggle__input {
    position: absolute;
    opacity: 0;
    padding: 0;
    margin: 0;
    pointer-events: none;
  }

  /* Disabled */

  .made-c-toggle--disabled,
  .made-c-toggle:has(.made-c-toggle__input:disabled) {
    cursor: not-allowed;
    color: var(--made-color-text-disabled);

    & .made-c-toggle__control {
      opacity: var(--made-opacity-25);
    }

    & .made-c-toggle__label {
      color: var(--made-color-text-disabled);
    }
  }

  /* checked */

  .made-c-toggle--checked .made-c-toggle__control,
  .made-c-toggle__input:checked ~ .made-c-toggle__control {
    background-color: var(--made-component-toggle-background-on);
    border-color: var(--made-component-toggle-background-on);

    & .made-c-toggle__thumb {
      background-color: var(--made-color-background-surface-default);
      border-color: var(--made-color-background-surface-default);
      translate: calc((var(--width) - var(--height)) / 2);
    }
  }

  /* Focus */

  .made-c-toggle:not(.made-c-toggle--disabled)
    .made-c-toggle__input:focus-visible
    ~ .made-c-toggle__control {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: 2px;
  }

  .made-c-toggle__label {
    display: none;
  }

  .made-c-form-control--has-label .made-c-toggle__label {
    display: inline-block;
    margin-inline-start: var(--made-size-space-4-x);
    user-select: none;
  }

  .made-c-form-control--has-supporting-text .made-c-toggle {
    align-items: flex-start;
  }

  /** Full-width */

  :host([full-width]) {
    display: block;
  }

  .made-c-toggle.made-c-toggle--full-width {
    width: 100%;
  }

  /** Reverse position of label and control */

  .made-c-toggle.made-c-toggle--reverse {
    flex-direction: row-reverse;

    & .made-c-toggle__label {
      margin-inline: 0 var(--made-size-space-4-x);
    }
  }
`;
