// @mc-made/web-components/dist/components/progress-bar/src/progress-bar.styles.js
import { css } from 'lit';
export default css `
  :host {
    --fill-color: var(--made-component-progress-progress-bar-background);
    --width: 100%;

    display: block;
    inline-size: var(--width);
  }

  .made-c-progress-bar {
    display: flex;
    flex-direction: column;
    gap: var(--made-size-space-2-x);
    font-family: var(--made-font-family-primary);
    width: 100%;
  }

  /* ─── Label ─────────────────────────────────────────────────── */

  .made-c-progress-bar__label {
    font-size: var(--made-font-size-body-small);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--made-line-height-rem-02);
    letter-spacing: var(--made-letter-spacing-default);
    color: var(--made-color-text-primary);
  }

  /* ─── Track and Fill Container ──────────────────────────────────── */

  .made-c-progress-bar__container {
    position: relative;
    width: 100%;
    border-radius: var(--made-border-radius-full);
    overflow: hidden;
  }

  /* ─── Track ──────────────────────────────────────────────────── */

  .made-c-progress-bar__track {
    position: relative;
    width: 100%;
    height: var(--made-size-space-2-x); /* medium default */
    background-color: var(--made-component-progress-track-background);
    border-radius: var(--made-component-progress-track-border-radius, 9999px);
  }

  :host([size='small']) .made-c-progress-bar__track {
    height: var(--made-size-space-1-x);
  }

  :host([size='large']) .made-c-progress-bar__track {
    height: var(--made-size-space-3-x);
  }

  :host([size='x-large']) .made-c-progress-bar__track {
    height: var(--made-size-space-4-x);
  }

  /* ─── Fill ───────────────────────────────────────────────────── */

  .made-c-progress-bar__fill {
    position: absolute;
    inset: 0;
    width: 100%;
    background-color: var(--fill-color);
    transform-origin: left;
    transform: scaleX(var(--progress-fraction, 0));
    transition: transform var(--made-motion-duration-short)
      var(--made-motion-easing-standard-ease-out);
  }

  /* ─── Indeterminate ─────────────────────────────────────────── */

  .made-c-progress-bar__indeterminate1,
  .made-c-progress-bar__indeterminate2 {
    position: absolute;
    inset: 0;
    background-color: var(--fill-color);
    border-radius: var(--made-border-radius-full);
    transform-origin: left center;
    will-change: transform;
  }

  .made-c-progress-bar__indeterminate1 {
    animation: 2.1s cubic-bezier(0.65, 0.815, 0.735, 0.395) 0s infinite normal
      none running indeterminate1;
  }

  .made-c-progress-bar__indeterminate2 {
    animation: 2.1s cubic-bezier(0.165, 0.84, 0.44, 1) 1.15s infinite normal
      none running indeterminate2;
  }

  /* ─── Supporting text ────────────────────────────────────────── */

  .made-c-progress-bar__supporting-text {
    font-size: var(--made-font-size-body-small);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--made-line-height-rem-02);
    letter-spacing: var(--made-letter-spacing-default);
    color: var(--made-color-text-secondary);
  }

  @keyframes indeterminate1 {
    0% {
      transform: translateX(-135%) scaleX(0.35);
    }
    60% {
      transform: translateX(100%) scaleX(0.9);
    }
    100% {
      transform: translateX(100%) scaleX(0.9);
    }
  }

  @keyframes indeterminate2 {
    0% {
      transform: translateX(-300%) scaleX(1);
    }
    60% {
      transform: translateX(107%) scaleX(0.08);
    }
    100% {
      transform: translateX(107%) scaleX(0.08);
    }
  }
`;
