// @mc-made/web-components/dist/components/progress-bar/src/progress-bar.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { property } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './progress-bar.styles.js';
/**
 * @summary Displays an indicator showing the completion progress of a task.
 * @documentation '../stories/progress-bar.mdx'
 *
 * @slot label - The progress bar's label. Alternatively, use the `label` attribute.
 * @slot supporting-text - Optional descriptive text displayed below the progress bar.
 *
 * @csspart base - The component's base wrapper.
 * @csspart label - The label element.
 * @csspart track - The track (background) element. Carries the `progressbar` ARIA role.
 * @csspart fill - The filled progress indicator.
 * @csspart supporting-text - The supporting text element.
 *
 * @cssproperty --fill-color - Color of the progress fill. Defaults to `var(--made-component-progress-progress-bar-background)`.
 * @cssproperty --width - The width of the progress bar. Defaults to `100%`.
 */
class MadeProgressBar extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, 'label', 'supporting-text');
        /** The current progress value. */
        this.value = 0;
        /** The maximum progress value. */
        this.maxValue = 100;
        /** The label text. Alternatively use the `label` slot. */
        this.label = '';
        /** The size of the progress bar track. */
        this.size = 'medium';
        /** Supporting text displayed below the progress bar. Alternatively use the `supporting-text` slot. */
        this.supportingText = '';
        /** If true, the progress bar will display an indeterminate state. */
        this.indeterminate = false;
    }
    updated(changedProperties) {
        if (changedProperties.has('value')) {
            requestAnimationFrame(() => {
                this.style.setProperty('--progress-fraction', `clamp(0, ${this.value / this.maxValue}, 1)`);
            });
        }
    }
    render() {
        const hasLabel = this.hasSlotController.test('label') || !!this.label;
        const hasSupportingText = this.hasSlotController.test('supporting-text') || !!this.supportingText;
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-progress-bar': true,
            [`made-c-progress-bar--${this.size}`]: true
        })}
      >
        ${hasLabel
            ? html `
              <div
                part="label"
                id="label"
                class=${classMap({
                'made-c-progress-bar__label': true
            })}
              >
                <slot name="label">${this.label}</slot>
              </div>
            `
            : nothing}

        <div class="made-c-progress-bar__container">
          <div
            part="track"
            class="made-c-progress-bar__track"
            role="progressbar"
            aria-valuenow=${this.indeterminate ? nothing : this.value}
            aria-valuemin=${this.indeterminate ? nothing : '0'}
            aria-valuemax=${this.indeterminate ? nothing : this.maxValue}
            aria-labelledby=${hasLabel ? 'label' : nothing}
            aria-describedby=${hasSupportingText ? 'supporting-text' : nothing}
            aria-label=${!hasLabel
            ? this.label || this.localize.term('progress')
            : nothing}
          >
            ${this.indeterminate
            ? html ` <div class="made-c-progress-bar__indeterminate1"></div>
                  <div class="made-c-progress-bar__indeterminate2"></div>`
            : html `
                  <div
                    part="fill"
                    class=${classMap({
                'made-c-progress-bar__fill': true
            })}
                  ></div>
                `}
          </div>
        </div>

        ${hasSupportingText
            ? html `
              <div
                id="supporting-text"
                part="supporting-text"
                class="made-c-progress-bar__supporting-text"
              >
                <slot name="supporting-text">${this.supportingText}</slot>
              </div>
            `
            : nothing}
      </div>
    `;
    }
}
MadeProgressBar.styles = [componentStyles, styles];
export default MadeProgressBar;
__decorate([
    property({ type: Number })
], MadeProgressBar.prototype, "value", void 0);
__decorate([
    property({ type: Number, attribute: 'max-value' })
], MadeProgressBar.prototype, "maxValue", void 0);
__decorate([
    property()
], MadeProgressBar.prototype, "label", void 0);
__decorate([
    property({ reflect: true })
], MadeProgressBar.prototype, "size", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeProgressBar.prototype, "supportingText", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeProgressBar.prototype, "indeterminate", void 0);
