// @mc-made/web-components/dist/components/textarea/src/textarea.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: block;
  }

  .made-c-form-control__label-wrapper {
    display: flex;
    justify-content: space-between;
  }

  .made-c-form-control .made-c-form-control__character-count {
    display: none;
  }

  .made-c-form-control--has-character-count
    .made-c-form-control__character-count {
    display: inline-block;
    margin-left: auto;
    margin-bottom: var(--made-size-space-1-x);
    font-weight: var(--made-font-weight-regular);
    color: var(--made-color-text-helper);
  }

  .made-c-form-control--has-character-count.made-c-form-control--medium
    .made-c-form-control__character-count {
    --font-size: var(--made-font-size-body-small);

    font-size: var(--font-size);
    line-height: calc(var(--font-size) * 1.45);
    line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
  }

  .made-c-form-control--has-character-count.made-c-form-control--large
    .made-c-form-control__character-count {
    --font-size: var(--made-font-size-body-medium);

    font-size: var(--font-size);
    line-height: calc(var(--font-size) * 1.45);
    line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
  }

  .made-c-textarea {
    --adjuster-padding: var(--made-size-space-0point5-x);
    display: grid;
    align-items: center;
    position: relative;
    width: 100%;
    font-family: var(--made-font-family-primary);
    font-size: var(--made-font-size-body-default);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--made-line-height-rem-04);
    letter-spacing: var(--made-letter-spacing-default);
    vertical-align: middle;
    background-color: var(--made-component-form-control-background);
    border: solid var(--made-component-form-control-border-width)
      var(--made-component-form-control-border-color);
    border-radius: var(--made-component-form-control-border-radius);
    padding: var(--adjuster-padding);
    cursor: text;

    &:focus-within {
      outline: 2px solid var(--made-color-border-focused);
      outline-offset: 2px;
    }
  }

  .made-c-textarea.made-c-textarea--disabled,
  .made-c-textarea:has(:disabled) {
    background-color: var(--made-color-background-disabled);
    border-color: var(--made-color-border-disabled);
    color: var(--made-color-text-disabled);
    cursor: not-allowed;
  }

  .made-c-textarea.made-c-textarea--disabled .made-c-textarea__control,
  .made-c-textarea:has(:disabled) .made-c-textarea__control {
    color: var(--made-color-text-disabled);
  }

  .made-c-textarea.made-c-textarea--readonly,
  .made-c-textarea:has(:readonly) {
    background-color: var(--made-color-background-disabled);
    color: var(--made-color-text-placeholder);
  }

  .made-c-textarea__control,
  .made-c-textarea__size-adjuster {
    grid-area: 1 / 1 / 2 / 2;
  }

  .made-c-textarea__size-adjuster {
    visibility: hidden;
    pointer-events: none;
    opacity: 0;
  }

  .made-c-textarea__control {
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    line-height: inherit;
    inline-size: 100%;
    block-size: 100%;
    color: var(--made-color-text-primary);
    border: none;
    background: none;
    box-shadow: none;
    cursor: inherit;
    -webkit-appearance: none;
  }

  .made-c-textarea__control::-webkit-search-decoration,
  .made-c-textarea__control::-webkit-search-cancel-button,
  .made-c-textarea__control::-webkit-search-results-button,
  .made-c-textarea__control::-webkit-search-results-decoration {
    -webkit-appearance: none;
  }

  .made-c-textarea__control::placeholder {
    color: var(--made-color-text-placeholder);
    user-select: none;
    -webkit-user-select: none;
  }

  .made-c-textarea__control:focus {
    outline: none;
  }

  /** Error state */
  .made-c-form-control.made-c-form-control--error .made-c-textarea {
    border-color: var(--made-color-border-danger-strong);
  }

  /** Success state */
  .made-c-form-control.made-c-form-control--success:not(
      .made-c-form-control--error
    )
    .made-c-textarea {
    border-color: var(--made-color-border-success-strong);
  }

  /*
   * Size modifiers
   */
  .made-c-textarea--medium .made-c-textarea__control {
    padding: calc(
        var(--made-component-textarea-medium-padding-y) - var(
            --adjuster-padding
          )
      )
      calc(
        var(--made-component-textarea-medium-padding-x) - var(
            --adjuster-padding
          )
      );
  }

  .made-c-textarea--large .made-c-textarea__control {
    padding: calc(
        var(--made-component-textarea-large-padding-y) - var(--adjuster-padding)
      )
      calc(
        var(--made-component-textarea-large-padding-x) - var(--adjuster-padding)
      );
  }

  /*
   * Resize types
   */

  .made-c-textarea--resize-none .made-c-textarea__control {
    resize: none;
  }

  .made-c-textarea--resize-vertical .made-c-textarea__control {
    resize: vertical;
  }

  .made-c-textarea--resize-auto .made-c-textarea__control {
    height: auto;
    resize: none;
    overflow-y: hidden;
  }
`;
