// @mc-made/web-components/dist/components/textarea/src/textarea.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { live } from 'lit/directives/live.js';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { mixinOnReportValidity, onReportValidity } from '../../../internal/face/on-report-validity.js';
import { property, query, state } from 'lit/decorators.js';
import { TextFieldValidator } from '../../../internal/face/validators/input-validator.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './textarea.styles.js';
/**
 * @summary  The `<made-textarea>` component represents a multi-line plain-text editing
 * control, useful when you want to allow users to enter a sizeable amount of
 * free-form text, for example a comment on a review or feedback form.
 * @documentation '../stories/textarea.mdx'
 *
 *
 * @slot label - The textarea's label. Alternatively, you can use the `label` attribute.
 *
 * @fires change {Event} - Fired when an alteration to the control's value is committed by the user.
 * @fires input {Event} - Fired when the control receives input.
 * @fires invalid {Event} - Fired when the control has been checked for validity and doesn't satisfy its constraints.
 *
 * @csspart container - The outer container.
 * @csspart label - The label's wrapper.
 * @csspart character-count - The character-count's wrapper.
 * @csspart supporting-text - The supporting text's wrapper.
 * @csspart base - The internal controls wrapper.
 * @csspart textarea - The internal `<textarea>` control.
 */
const textareaBaseClass = mixinOnReportValidity(mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement))));
class MadeTextarea extends textareaBaseClass {
    constructor() {
        super(...arguments);
        this.LABEL = 'label';
        this.hasSlotController = new HasSlotController(this, this.LABEL, 'supporting-text');
        this.hasFocus = false;
        /**
         * Whether or not a native error has been reported via `reportValidity()`.
         */
        this.hasNativeError = false;
        /**
         * The validation message displayed from a native error via
         * `reportValidity()`.
         */
        this.nativeErrorText = '';
        this.title = ''; // make reactive to pass through
        /**
         * Gets or sets whether or not the text field is in a visually invalid state.
         *
         * This error state overrides the error state controlled by
         * `reportValidity()`.
         */
        this.error = false;
        /**
         * Gets or sets an explicit visually valid state. Can be used as part of
         * a manual vaidation process.
         *
         * The error state returned by `reportValidity()` will override this.
         */
        this.success = false;
        /** The current value of the input, submitted as a name/value pair with form data. */
        this.value = '';
        /** The input's label. If you need to display HTML, use the `label` slot instead. */
        this.label = '';
        /** The input's size. */
        this.size = 'medium';
        /** Control's how the textarea can be resized. */
        this.resize = 'vertical';
        /** The default number of rows visible in the text area. */
        this.rows = 4;
        /**
         *  The textarea's supporting text. Provide extra details below the control, e.g. usage instructions.
         */
        this.supportingText = '';
        /**
         * Placeholder text to show as a hint when the input is empty.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/placeholder
         */
        this.placeholder = '';
        /**
         * Makes the input readonly.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#readonly
         */
        this.readonly = false;
        /**
         * Makes the input a required field.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required
         */
        this.required = false;
        /** Visual indicators that are appended to the label of a required field. */
        this.requiredIndicator = 'required';
        /** Add the optional visual indicator to a non-required field. */
        this.optional = false;
        /** Hides the required visual indicator on the label, when the text field is required. */
        this.noRequiredIndicator = false;
        /**
         * The minimum length of input that will be considered valid.
         * Set to -1 for none.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#minlength
         */
        this.minlength = -1;
        /**
         * The maximum length of input that will be considered valid.
         * Set to -1 for none.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#maxlength
         */
        this.maxlength = -1;
        /** Shows the character count if input type is 'text' and maxlength is set. */
        this.characterCount = false;
        this.setCustomValidity = (error) => {
            super.setCustomValidity(error);
            if (!error && this.hasNativeError) {
                this.hasNativeError = false;
                this.nativeErrorText = '';
            }
        };
    }
    get hasError() {
        return this.error || this.hasNativeError;
    }
    connectedCallback() {
        super.connectedCallback();
        this.resizeObserver = new ResizeObserver(() => this.setDimensions());
        this.updateComplete.then(() => {
            this.setDimensions();
            this.resizeObserver.observe(this.textarea);
            if (this.textarea && this.value !== this.textarea.value) {
                const value = this.textarea.value;
                this.value = value;
            }
        });
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        if (this.textarea) {
            this.resizeObserver?.unobserve(this.textarea);
        }
    }
    handleBlur() {
        this.hasFocus = false;
    }
    handleChange() {
        this.value = this.textarea.value;
        this.setDimensions();
        // dispatch 'composed' change event
        this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
    }
    handleFocus() {
        this.hasFocus = true;
    }
    handleInput() {
        this.value = this.textarea.value;
        if (this.hasError) {
            this.reportValidity();
        }
    }
    /** Sets focus on the textarea. */
    focus(options) {
        this.textarea.focus(options);
    }
    /** Removes focus from the textarea. */
    blur() {
        this.textarea.blur();
    }
    setDimensions() {
        const parsePx = (value) => parseFloat(value);
        const getBorderWidth = () => parsePx(getComputedStyle(this.base).getPropertyValue('border-width'));
        const getPadding = () => parsePx(getComputedStyle(this.base).getPropertyValue('padding'));
        if (this.resize === 'none') {
            return;
        }
        if (this.resize === 'auto') {
            this.sizeAdjuster.style.height = `${this.textarea.clientHeight}px`;
            this.textarea.style.height = 'auto';
            this.textarea.style.height = `${this.textarea.scrollHeight}px`;
            this.base.style.width = '';
            this.base.style.height = '';
            return;
        }
        if (this.textarea.style.width) {
            const borderWidth = getBorderWidth();
            const padding = getPadding();
            const textareaWidth = parsePx(this.textarea.style.width);
            const width = textareaWidth + borderWidth * 2 + padding * 2;
            this.base.style.width = `${width}px`;
        }
        if (this.textarea.style.height) {
            const borderWidth = getBorderWidth();
            const padding = getPadding();
            const textareaHeight = parsePx(this.textarea.style.height);
            const height = textareaHeight + borderWidth * 2 + padding * 2;
            this.base.style.height = `${height}px`;
        }
    }
    /** Selects all the text in the input.
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/select
     */
    select() {
        this.textarea.select();
    }
    /**
     * Sets the start and end positions of the text selection (0-based).
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setSelectionRange
     */
    setSelectionRange(selectionStart, selectionEnd, selectionDirection = 'none') {
        this.textarea.setSelectionRange(selectionStart, selectionEnd, selectionDirection);
    }
    /**
     * Replaces a range of text with a new string.
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setRangeText
     */
    setRangeText(replacement, start, end, selectMode = 'preserve') {
        const selectionStart = start ?? this.textarea.selectionStart;
        const selectionEnd = end ?? this.textarea.selectionEnd;
        this.textarea.setRangeText(replacement, selectionStart, selectionEnd, selectMode);
        if (this.value !== this.textarea.value) {
            this.value = this.textarea.value;
            this.setDimensions();
        }
    }
    get errorOrSuccessText() {
        if (this.hasError && this.getErrorText()) {
            return this.getErrorText();
        }
        if (this.success && this.successText) {
            return this.successText;
        }
        return null;
    }
    get supportingOrValidationText() {
        return this.errorOrSuccessText
            ? this.errorOrSuccessText
            : this.supportingText;
    }
    renderSupportingText() {
        if (!this.supportingOrValidationText) {
            return null;
        }
        return html `
      ${this.hasError
            ? html `<span>
            <made-icon
              name="exclamation-triangle-solid-12px"
              library="system"
            ></made-icon>
          </span>`
            : this.success
                ? html `<span>
              <made-icon
                name="check-circle-solid-12px"
                library="system"
              ></made-icon>
            </span>`
                : nothing}
      <span>${this.supportingOrValidationText}</span>
    `;
    }
    updated(changedProperties) {
        if (['resize', 'value', 'rows'].some(watchedProp => changedProperties.has(watchedProp))) {
            this.setDimensions();
        }
    }
    render() {
        const hasLabel = this.label
            ? true
            : !!this.hasSlotController.test(this.LABEL);
        const hasCharacterCount = this.maxlength > -1 && this.characterCount;
        const charCount = hasCharacterCount
            ? `${this.value.length}/${this.maxlength}`
            : null;
        const role = this.hasError || this.success ? 'alert' : nothing;
        const supportingText = this.renderSupportingText();
        return html `
      <div
        part="container"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control--medium': this.size === 'medium',
            'made-c-form-control--large': this.size === 'large',
            'made-c-form-control--has-label': hasLabel,
            'made-c-form-control--has-character-count': hasCharacterCount,
            'made-c-form-control--error': !this.disabled && this.hasError,
            'made-c-form-control--success': this.success,
            'made-c-form-control--has-supporting-text': !!supportingText
        })}
      >
        <div class="made-c-form-control__label-wrapper">
          <label
            part="label"
            class="made-c-form-control__label"
            for="textarea"
            aria-hidden=${hasLabel ? 'false' : 'true'}
          >
            <slot name=${this.LABEL}>${this.label}</slot>
          </label>
          <span
            part="character-count"
            class="made-c-form-control__character-count"
            aria-hidden=${hasCharacterCount ? 'false' : 'true'}
          >
            ${charCount}
          </span>
        </div>
        <div class="made-c-form-control-input">
          <div
            part="base"
            class=${classMap({
            'made-c-textarea': true,
            'made-c-textarea--medium': this.size === 'medium',
            'made-c-textarea--large': this.size === 'large',
            'made-c-textarea--disabled': this.disabled,
            'made-c-textarea--readonly': this.readonly,
            'made-c-textarea--focused': this.hasFocus,
            'made-c-textarea--resize-none': this.resize === 'none',
            'made-c-textarea--resize-vertical': this.resize === 'vertical',
            'made-c-textarea--resize-auto': this.resize === 'auto'
        })}
          >
            <textarea
              part="textarea"
              id="textarea"
              class="made-c-textarea__control"
              name=${this.name}
              ?disabled=${this.disabled}
              ?readonly=${this.readonly}
              ?required=${this.required}
              placeholder=${ifDefined(this.placeholder)}
              minlength=${ifDefined(this.minlength)}
              maxlength=${ifDefined(this.maxlength)}
              .value=${live(this.value)}
              rows=${this.rows}
              autocapitalize=${ifDefined(this.autocapitalize)}
              autocomplete=${ifDefined(this.autocomplete)}
              ?autofocus=${this.autofocus}
              enterkeyhint=${ifDefined(this.enterkeyhint)}
              inputmode=${ifDefined(this.inputmode)}
              aria-describedby="supporting-text"
              @change=${this.handleChange}
              @input=${this.handleInput}
              @focus=${this.handleFocus}
              @blur=${this.handleBlur}
            ></textarea>
            <div
              class="made-c-textarea__size-adjuster"
              ?hidden=${this.resize !== 'auto'}
            ></div>
          </div>
        </div>
        <div
          part="supporting-text"
          id="supporting-text"
          class="made-c-form-control__supporting-text"
          role=${role}
        >
          ${supportingText}
        </div>
      </div>
    `;
    }
    /**
     * Reset the text field to its default value.
     */
    reset() {
        this.value = this.getAttribute('value') ?? '';
        this.hasNativeError = false;
        this.nativeErrorText = '';
    }
    getErrorText() {
        return this.error ? this.errorText : this.nativeErrorText;
    }
    [getFormValue]() {
        return this.value;
    }
    formResetCallback() {
        this.reset();
    }
    // eslint-disable-next-line @typescript-eslint/no-shadow
    formStateRestoreCallback(state) {
        this.value = state;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [createValidator]() {
        return new TextFieldValidator(() => ({
            state: this,
            renderedControl: this.textarea
        }));
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [getValidityAnchor]() {
        return this.textarea;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [onReportValidity](invalidEvent) {
        // Prevent default pop-up behavior.
        invalidEvent?.preventDefault();
        this.hasNativeError = !!invalidEvent;
        this.nativeErrorText = this.validationMessage;
    }
}
MadeTextarea.styles = [componentStyles, formControlStyles, styles];
export default MadeTextarea;
__decorate([
    query('textarea')
], MadeTextarea.prototype, "textarea", void 0);
__decorate([
    query('[part~="base"]')
], MadeTextarea.prototype, "base", void 0);
__decorate([
    query('.made-c-textarea__size-adjuster')
], MadeTextarea.prototype, "sizeAdjuster", void 0);
__decorate([
    state()
], MadeTextarea.prototype, "hasFocus", void 0);
__decorate([
    state()
], MadeTextarea.prototype, "hasNativeError", void 0);
__decorate([
    state()
], MadeTextarea.prototype, "nativeErrorText", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "title", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTextarea.prototype, "error", void 0);
__decorate([
    property({ attribute: 'error-text' })
], MadeTextarea.prototype, "errorText", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTextarea.prototype, "success", void 0);
__decorate([
    property({ attribute: 'success-text' })
], MadeTextarea.prototype, "successText", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "value", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "label", void 0);
__decorate([
    property({ reflect: true })
], MadeTextarea.prototype, "size", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "resize", void 0);
__decorate([
    property({ type: Number })
], MadeTextarea.prototype, "rows", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeTextarea.prototype, "supportingText", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTextarea.prototype, "readonly", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTextarea.prototype, "required", void 0);
__decorate([
    property({ attribute: 'required-indicator', reflect: true })
], MadeTextarea.prototype, "requiredIndicator", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTextarea.prototype, "optional", void 0);
__decorate([
    property({
        type: Boolean,
        reflect: true,
        attribute: 'no-required-indicator'
    })
], MadeTextarea.prototype, "noRequiredIndicator", void 0);
__decorate([
    property({ type: Number })
], MadeTextarea.prototype, "minlength", void 0);
__decorate([
    property({ type: Number })
], MadeTextarea.prototype, "maxlength", void 0);
__decorate([
    property({ attribute: 'character-count', type: Boolean })
], MadeTextarea.prototype, "characterCount", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "autocapitalize", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean })
], MadeTextarea.prototype, "autofocus", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "enterkeyhint", void 0);
__decorate([
    property()
], MadeTextarea.prototype, "inputmode", void 0);
