// @mc-made/web-components/dist/react/made-radio/index.js
import * as React from 'react';
import { createComponent } from '../create-component.js';
import Component from '../../components/radio/src/radio.component.js';
const tagName = 'made-radio';
Component.define(tagName);
/**
 * @summary Radios allow the user select one option from a set of options.
 * @documentation '../../components/radio/stories/radio.mdx'
 *
 * @fires change {Event} - Fired when the radio is selected.
 * @fires input {Event} - Fired when the radio receives input.
 * @fires mde-invalid {CustomEvent} - Fired when the radio is in an invalid state.
 * @fires mde-reset {CustomEvent} - Fired when the radio is reset by the form.
 *
 * @csspart base - The component's base wrapper.
 * @csspart control - The radio's control in it's unchecked state.
 * @csspart control-checked - The radio's control in it's checked state.
 * @csspart label - The label's wrapper.
 * @csspart form-control-supporting-text - The supporting text's wrapper.
 *
 */
const reactWrapper = createComponent({
    tagName,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
        onMdeInvalid: 'mde-invalid',
        onMdeReset: 'mde-reset'
    },
    displayName: 'MadeRadio'
});
export default reactWrapper;
