// @mc-made/web-components/dist/components/link/src/link.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { property, query, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './link.styles.js';
/**
 * @summary A link allow users to navigate to a different location. They can be presented in-line inside a paragraph or as standalone text.
 * @documentation '../stories/link.mdx'
 *
 * @slot - The link's label.
 * @slot start - Used to place an content before the link's label. Typically used for icons.
 * @slot end - Used to place content after the link's label. Typically used for icons.
 *
 * @csspart base - The component's base wrapper.
 * @csspart end - The container that wraps the end slot.
 * @csspart label - The link's label.
 * @csspart start - The container that wraps the start slot.
 *
 */
class MadeLink extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'start', 'end');
        this.hasFocus = false;
        /** Alternative ways to underline link. */
        this.decoration = 'underline';
        /** Use the appearance attribute to change the visual style of the link. */
        this.appearance = 'neutral';
        /** Setting true disabled the link. */
        this.disabled = false;
    }
    handleBlur() {
        this.hasFocus = false;
    }
    handleFocus() {
        this.hasFocus = true;
    }
    /** Simulates a click on the link. */
    click() {
        this.link.click();
    }
    /** Sets focus on the link. */
    focus(options) {
        this.link.focus(options);
    }
    /** Removes focus from the link. */
    blur() {
        this.link.blur();
    }
    render() {
        return html `
      <a
        part="base"
        class=${classMap({
            'made-c-link': true,
            'made-c-link--neutral': this.appearance === 'neutral',
            'made-c-link--branded': this.appearance === 'branded',
            'made-c-link--underline': this.decoration === 'underline',
            'made-c-link--subtle': this.decoration === 'subtle',
            'made-c-link--disabled': this.disabled,
            'made-c-link--focused': this.hasFocus,
            'made-c-link--has-start-slot': this.hasSlotController.test('start'),
            'made-c-link--has-end-slot': this.hasSlotController.test('end')
        })}
        href=${ifDefined(this.href)}
        target=${ifDefined(this.target)}
        download=${ifDefined(this.download)}
        rel=${ifDefined(this.rel)}
        referrerpolicy=${ifDefined(this.referrerpolicy)}
        aria-disabled=${this.disabled ? 'true' : 'false'}
        tabindex=${this.disabled ? '-1' : '0'}
        @blur=${this.handleBlur}
        @focus=${this.handleFocus}
      >
        <span part="start" class="made-c-link__start">
          <slot name="start"></slot>
        </span>
        <slot part="label" class="made-c-link__label"></slot>
        <span part="end" class="made-c-link__end">
          <slot name="end"></slot>
        </span>
      </a>
    `;
    }
}
MadeLink.styles = [componentStyles, styles];
export default MadeLink;
__decorate([
    query('.made-c-link')
], MadeLink.prototype, "link", void 0);
__decorate([
    state()
], MadeLink.prototype, "hasFocus", void 0);
__decorate([
    property()
], MadeLink.prototype, "decoration", void 0);
__decorate([
    property()
], MadeLink.prototype, "appearance", void 0);
__decorate([
    property()
], MadeLink.prototype, "href", void 0);
__decorate([
    property({ attribute: 'aria-label' })
], MadeLink.prototype, "ariaLabel", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeLink.prototype, "disabled", void 0);
__decorate([
    property()
], MadeLink.prototype, "download", void 0);
__decorate([
    property()
], MadeLink.prototype, "target", void 0);
__decorate([
    property()
], MadeLink.prototype, "rel", void 0);
__decorate([
    property()
], MadeLink.prototype, "referrerpolicy", void 0);
