// @mc-made/web-components/dist/components/link/src/link.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  :host {
    display: inline-block;
  }

  .made-c-link {
    display: inline-flex;
    align-items: center;
    margin: 0;
    font: inherit;
    font-size: inherit;
    font-weight: inherit;
    line-height: inherit;
    letter-spacing: inherit;
    color: inherit;
    text-decoration-line: var(--made-text-decoration-underline);
    text-decoration-thickness: initial;
    text-decoration-style: initial;
    text-underline-offset: 0.125em;
  }

  .made-c-link:focus {
    outline: none;
  }

  .made-c-link:focus-visible,
  .made-c-link--focused {
    outline: none;
    position: relative;
  }

  .made-c-link:focus-visible::after,
  .made-c-link--focused::after {
    content: '';
    position: absolute;
    inset: -4px;
    border-radius: var(--made-border-radius-default);
    border: 2px solid var(--made-color-border-focused);
    pointer-events: none;
  }

  /* Size modifiers */

  .made-c-link--large {
    font-size: var(--made-component-link-large-font-size);
  }

  /* Appearance modifiers */
  .made-c-link--subtle {
    text-decoration: var(--made-text-decoration-none);
  }

  .made-c-link--subtle:focus-visible,
  .made-c-link--subtle:hover,
  .made-c-link--subtle:active {
    text-decoration-line: var(--made-text-decoration-underline);
  }

  .made-c-link--underline,
  .made-c-link--underline:visited {
    text-decoration: var(--made-text-decoration-underline);
  }

  .made-c-link--underline:focus-visible,
  .made-c-link--underline:hover,
  .made-c-link--underline:active {
    text-decoration-line: var(--made-text-decoration-underline);
  }

  /* Color modifiers */
  .made-c-link--neutral {
    color: var(--made-component-link-neutral-default-text-color);
  }

  .made-c-link--neutral:hover {
    color: var(--made-component-link-neutral-hover-text-color);
  }

  .made-c-link--neutral:active {
    color: var(--made-component-link-neutral-active-text-color);
  }

  .made-c-link--neutral:not(.made-c-link--disabled, :disabled):visited {
    color: var(--made-component-link-neutral-visited-text-color);
  }

  .made-c-link--branded {
    color: var(--made-component-link-default-text-color);
  }

  .made-c-link--branded:hover {
    color: var(--made-component-link-hover-text-color);
  }

  .made-c-link--branded:active {
    color: var(--made-component-link-active-text-color);
  }

  .made-c-link--branded:not(.made-c-link--disabled, :disabled):visited {
    color: var(--made-component-link-visited-text-color);
  }

  /* State modifiers*/
  .made-c-link--disabled {
    color: var(--made-color-text-disabled);
    pointer-events: none;
  }

  /* Spacing modifiers */
  .made-c-link--has-start-slot .made-c-link__label {
    padding-inline-start: var(--made-component-link-gap);
  }

  .made-c-link--has-end-slot .made-c-link__label {
    padding-inline-end: var(--made-component-link-gap);
  }

  .made-c-link__start,
  .made-c-link__end {
    display: none;
  }

  .made-c-link--has-start-slot .made-c-link__start,
  .made-c-link--has-end-slot .made-c-link__end {
    flex: 0 0 auto;
    display: flex;
    align-items: center;
    pointer-events: none;
  }

  .made-c-link__label {
    display: inline-block;
  }
`;
