// @mc-made/web-components/dist/components/input/src/input.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { TextFieldValidator } from '../../../internal/face/validators/input-validator.js';
import { live } from 'lit/directives/live.js';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { mixinOnReportValidity, onReportValidity } from '../../../internal/face/on-report-validity.js';
import { property, query, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/src/icon.component.js';
import MadeIconButton from '../../icon-button/made-icon-button.js';
import styles from './input.styles.js';
/**
 * @summary Inputs collect data from the user.
 * @documentation '../stories/input.mdx'
 *
 * @dependency made-icon
 *
 * @slot label - The input's label. Alternatively, you can use the `label` attribute.
 * @slot start - Optional slot used to place an icon or prefix at the start of the input.
 * @slot end - Optional slot used to place an icon or suffix at the end of the input.
 * @slot clear-icon - An icon to use in lieu of the default clear icon.
 * @slot show-password-icon - An icon to use in lieu of the default show password icon.
 * @slot hide-password-icon - An icon to use in lieu of the default hide password icon.
 *
 * @fires change {Event} - Fired when an alteration to the control's value is committed by the user.
 * @fires input {Event} - Fired when the control receives input.
 * @fires invalid {Event} - Fired when the control has been checked for validity and doesn't satisfy its constraints.
 * @fires mde-clear {CustomEvent} - Fired when the clear button is activated.
 *
 * @csspart form-control - The outer wrapper.
 * @csspart label - The label's wrapper.
 * @csspart character-count - The character-count's wrapper.r.
 * @csspart supporting-text - The help text's wrapper.
 * @csspart input-container - The container that wraps the input and any associated elements.
 * @csspart input - The internal `<input>` control.
 * @csspart start - The container that wraps the start slot.
 * @csspart clear-button - The clear button.
 * @csspart password-toggle-button - The password toggle button.
 * @csspart end - The container that wraps the end slot.
 */
const inputBaseClass = mixinOnReportValidity(mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement))));
class MadeInput extends inputBaseClass {
    constructor() {
        super(...arguments);
        this.LABEL = 'label';
        this.hasSlotController = new HasSlotController(this, this.LABEL, 'supporting-text', 'start');
        this.isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
        this.hasFocus = false;
        this.passwordVisible = false;
        /**
         * Whether or not a native error has been reported via `reportValidity()`.
         */
        this.hasNativeError = false;
        /**
         * The validation message displayed from a native error via
         * `reportValidity()`.
         */
        this.nativeErrorText = '';
        this.title = ''; // make reactive to pass through
        this.__numberInput = Object.assign(document.createElement('input'), {
            type: 'number'
        });
        this.__dateInput = Object.assign(document.createElement('input'), {
            type: 'date'
        });
        /**
         * Gets or sets whether or not the text field is in a visually invalid state.
         *
         * This error state overrides the error state controlled by
         * `reportValidity()`.
         */
        this.error = false;
        /**
         * Gets or sets an explicit visually valid state. Can be used as part of
         * a manual vaidation process.
         *
         * The error state returned by `reportValidity()` will override this.
         */
        this.success = false;
        /**
         * The type of input. Works the same as a native `<input>` element, but only a subset of types are supported. Defaults
         * to `text`.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#input_types
         */
        this.type = 'text';
        /** The current value of the input, submitted as a name/value pair with form data. */
        this.value = '';
        /** The input's label. If you need to display HTML, use the `label` slot instead. */
        this.label = '';
        /** The input's size. */
        this.size = 'medium';
        /**
         * Provides extra details below the text field, including instructions on its usage.
         */
        this.supportingText = '';
        /** Adds a clear button when the input is not empty. */
        this.clearable = false;
        /**
         * Placeholder text to show as a hint when the input is empty.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/placeholder
         */
        this.placeholder = '';
        /**
         * Makes the input readonly.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#readonly
         */
        this.readonly = false;
        /** Adds a button to toggle the password's visibility. Only applies to password types. */
        this.passwordToggle = false;
        /** Hides the browser's built-in increment/decrement spin buttons for number inputs. */
        this.noSpinner = false;
        /**
         * Makes the input a required field.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/required
         */
        this.required = false;
        /** Visual indicators that are appended to the label of a required field. */
        this.requiredIndicator = 'required';
        /** Add the optional visual indicator to a non-required field. */
        this.optional = false;
        /** Hides the required visual indicator on the label, when the text field is required. */
        this.noRequiredIndicator = false;
        /**
         * The minimum length of input that will be considered valid.
         * Set to -1 for none.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#minlength
         */
        this.minlength = -1;
        /**
         * The maximum length of input that will be considered valid.
         * Set to -1 for none.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#maxlength
         */
        this.maxlength = -1;
        /** Shows the character count if input type is 'text' and maxlength is set. */
        this.characterCount = false;
    }
    get hasError() {
        return this.error || this.hasNativeError;
    }
    //
    // NOTE: We use an in-memory input for these getters/setters instead of the one in the template because the properties
    // can be set before the component is rendered.
    //
    /**
     * Gets or sets the current value as a `Date` object. Returns `null` if the value can't be converted. This will use the native `<input type="{{type}}">` implementation and may result in an error.
     */
    get valueAsDate() {
        this.__dateInput.type = this.type;
        this.__dateInput.value = this.value;
        return this.input?.valueAsDate || this.__dateInput.valueAsDate;
    }
    set valueAsDate(newValue) {
        this.__dateInput.type = this.type;
        this.__dateInput.valueAsDate = newValue;
        this.value = this.__dateInput.value;
    }
    /** Gets or sets the current value as a number. Returns `NaN` if the value can't be converted. */
    get valueAsNumber() {
        this.__numberInput.value = this.value;
        return this.input?.valueAsNumber || this.__numberInput.valueAsNumber;
    }
    set valueAsNumber(newValue) {
        this.__numberInput.valueAsNumber = newValue;
        this.value = this.__numberInput.value;
    }
    /* Check min and max attribute values for 'today' keyword and convert if necessary */
    updated(changedProperties) {
        if (changedProperties.has('min') && this.min === 'today') {
            this.min = this.convertTodayToDate();
        }
        if (changedProperties.has('max') && this.max === 'today') {
            this.max = this.convertTodayToDate();
        }
    }
    convertTodayToDate() {
        const today = new Date();
        if (this.type === 'date') {
            return today.toISOString().split('T')[0];
        }
        if (this.type === 'datetime-local') {
            return today.toISOString().split('.')[0];
        }
        return 'today';
    }
    handleBlur() {
        this.hasFocus = false;
    }
    handleChange() {
        this.value = this.input.value;
        // dispatch 'composed' change event
        this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
    }
    handleClearClick(event) {
        event.preventDefault();
        if (this.value !== '') {
            this.value = '';
            this.emit('mde-clear');
        }
        this.input.focus();
    }
    handleFocus() {
        this.hasFocus = true;
    }
    handleInput() {
        this.value = this.input.value;
        if (this.hasError) {
            this.reportValidity();
        }
    }
    handlePasswordToggle() {
        const start = this.input.selectionStart;
        const end = this.input.selectionEnd;
        const dir = this.input.selectionDirection || 'none';
        this.passwordVisible = !this.passwordVisible;
        /** Ensure the cursor remains at the end of the input string */
        requestAnimationFrame(() => {
            this.input.focus({ preventScroll: true });
            this.input.setSelectionRange(start, end, dir);
        });
    }
    /** Prevent toggle from stealing focus */
    handlePointerDown(event) {
        event.preventDefault();
    }
    /** Sets focus on the input. */
    focus(options) {
        this.input.focus(options);
    }
    /** Removes focus from the input. */
    blur() {
        this.input.blur();
    }
    /** Selects all the text in the input.
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/select
     */
    select() {
        this.input.select();
    }
    /**
     * Sets the start and end positions of the text selection (0-based).
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setSelectionRange
     */
    setSelectionRange(selectionStart, selectionEnd, selectionDirection = 'none') {
        this.input.setSelectionRange(selectionStart, selectionEnd, selectionDirection);
    }
    /**
     * Replaces a range of text with a new string.
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/setRangeText
     */
    setRangeText(replacement, start, end, selectMode = 'preserve') {
        const selectionStart = start ?? this.input.selectionStart;
        const selectionEnd = end ?? this.input.selectionEnd;
        this.input.setRangeText(replacement, selectionStart, selectionEnd, selectMode);
        if (this.value !== this.input.value) {
            this.value = this.input.value;
        }
    }
    /**
     * Displays the browser picker for an input element (only works if the browser supports it for the input type).
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/showPicker
     * */
    showPicker() {
        if ('showPicker' in HTMLInputElement.prototype) {
            this.input.showPicker();
        }
    }
    /** Increments the value of a numeric input type by the value of the step attribute.
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/stepUp
     */
    stepUp() {
        this.input.stepUp();
        if (this.value !== this.input.value) {
            this.value = this.input.value;
        }
    }
    /**
     * Decrements the value of a numeric input type by the value of the step attribute.
     *
     * https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement/stepDown
     */
    stepDown() {
        this.input.stepDown();
        if (this.value !== this.input.value) {
            this.value = this.input.value;
        }
    }
    handleKeyDown(event) {
        if (event.key === 'Enter') {
            setTimeout(() => {
                const form = this.form;
                const submitButton = form?.querySelector('button[type="submit"], made-button[type="submit"]');
                submitButton?.click();
            });
        }
    }
    get errorOrSuccessText() {
        if (this.hasError && this.getErrorText()) {
            return this.getErrorText();
        }
        if (this.success && this.successText) {
            return this.successText;
        }
        return null;
    }
    get supportingOrValidationText() {
        return this.errorOrSuccessText
            ? this.errorOrSuccessText
            : this.supportingText;
    }
    renderSupportingText() {
        if (!this.supportingOrValidationText) {
            return null;
        }
        return html `
      ${this.hasError
            ? html `<span class="made-c-form-control__error-icon">
            <made-icon
              name="exclamation-triangle-solid-12px"
              library="system"
            ></made-icon>
          </span>`
            : this.success
                ? html `<span class="made-c-form-control__success-icon">
              <made-icon
                name="check-circle-solid-12px"
                library="system"
              ></made-icon>
            </span>`
                : nothing}
      <span>${this.supportingOrValidationText}</span>
    `;
    }
    render() {
        const hasLabel = this.label
            ? true
            : !!this.hasSlotController.test(this.LABEL);
        const hasClearIcon = this.clearable && !this.disabled && !this.readonly;
        const isClearIconVisible = hasClearIcon && (typeof this.value === 'number' || this.value.length > 0);
        const hasCharacterCount = this.type === 'text' && this.maxlength > -1 && this.characterCount;
        const charCount = hasCharacterCount
            ? `${this.value.length}/${this.maxlength}`
            : null;
        const role = this.hasError || this.success ? 'alert' : null;
        const supportingText = this.renderSupportingText();
        const showCalendar = this.type === 'date' || this.type === 'datetime-local';
        /* Only Chrome has a time picker so hide in Safari. Visibility in Firefox is controlled via css */
        const showClock = this.type === 'time' && !this.isSafari;
        return html `
      <div
        part="form-control"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control--medium': this.size === 'medium',
            'made-c-form-control--large': this.size === 'large',
            'made-c-form-control--has-label': hasLabel,
            'made-c-form-control--has-character-count': hasCharacterCount,
            'made-c-form-control--error': !this.disabled && this.hasError,
            'made-c-form-control--success': this.success,
            'made-c-form-control--has-supporting-text': !!supportingText
        })}
      >
        <div class="made-c-form-control__label-wrapper">
          <label
            part="label"
            class="made-c-form-control__label"
            for="input"
            aria-hidden=${hasLabel ? 'false' : 'true'}
          >
            <slot name=${this.LABEL}>${this.label}</slot>
          </label>
          <span
            part="character-count"
            class="made-c-form-control__character-count"
            aria-hidden=${hasCharacterCount ? 'false' : 'true'}
          >
            ${charCount}
          </span>
        </div>
        <div
          part="input-container"
          class=${classMap({
            'made-c-input': true,
            'made-c-input--medium': this.size === 'medium',
            'made-c-input--large': this.size === 'large',
            'made-c-input--disabled': this.disabled,
            'made-c-input--readonly': this.readonly,
            'made-c-input--focused': this.hasFocus,
            'made-c-input--no-spinner': this.noSpinner,
            'made-c-input--has-start-slot': this.hasSlotController.test('start') || this.type === 'search'
        })}
        >
          <span part="start" class="made-c-input__start">
            ${this.type === 'search'
            ? html `<slot name="start"
                  ><span class="made-c-input__start-search"
                    ><made-icon
                      name="magnifying-glass-outline"
                      library="system"
                    ></made-icon></span
                ></slot>`
            : html ` <slot name="start"></slot>`}
          </span>
          <input
            part="input"
            id="input"
            class="made-c-input__control"
            type=${this.type === 'password' && this.passwordVisible
            ? 'text'
            : this.type}
            title=${this
            .title /* An empty title prevents browser validation tooltips from appearing on hover */}
            name=${this.name}
            ?disabled=${this.disabled}
            ?readonly=${this.readonly}
            ?required=${this.required}
            placeholder=${ifDefined(this.placeholder)}
            minlength=${ifDefined(this.minlength)}
            maxlength=${ifDefined(this.maxlength)}
            min=${ifDefined(this.min)}
            max=${ifDefined(this.max)}
            step=${ifDefined(this.step)}
            .value=${live(this.value)}
            autocapitalize=${ifDefined(this.autocapitalize)}
            autocomplete=${ifDefined(this.autocomplete)}
            ?autofocus=${this.autofocus}
            pattern=${ifDefined(this.pattern)}
            enterkeyhint=${ifDefined(this.enterkeyhint)}
            inputmode=${ifDefined(this.inputmode)}
            aria-describedby="supporting-text"
            @keydown=${this.handleKeyDown}
            @change=${this.handleChange}
            @input=${this.handleInput}
            @focus=${this.handleFocus}
            @blur=${this.handleBlur}
          />
          ${isClearIconVisible
            ? html `
                <made-icon-button
                  part="clear-button"
                  exportparts="base:clear-button-base"
                  class="made-c-input__button"
                  aria-label="${this.localize.term('clearEntry')}"
                  icon-name="x-mark-outline"
                  variant="ghost"
                  library="system"
                  size="small"
                  @click=${this.handleClearClick}
                ></made-icon-button>
              `
            : ''}
          ${this.passwordToggle && !this.disabled
            ? html `
                <made-icon-button
                  part="password-toggle-button"
                  exportparts="base:password-toggle-button-base"
                  class="made-c-input__button"
                  variant="ghost"
                  size="small"
                  icon-name=${this.passwordVisible
                ? 'eye-slash-outline'
                : 'eye-outline'}
                  aria-label=${this.passwordVisible
                ? this.localize.term('hidePassword')
                : this.localize.term('showPassword')}
                  library="system"
                  @click=${this.handlePasswordToggle}
                  @mousedown=${this.handlePointerDown}
                ></made-icon-button>
              `
            : ''}
          ${showCalendar
            ? html `
                <made-icon-button
                  part="calendar-button"
                  class="made-c-input__calendar made-c-input__button"
                  variant="ghost"
                  size="small"
                  icon-name="time/outline/calendar-outline"
                  aria-label=${this.localize.term('selectDate')}
                  @click=${() => this.input.showPicker()}
                ></made-icon-button>
              `
            : ''}
          ${showClock
            ? html `
                <made-icon-button
                  part="clock-button"
                  class="made-c-input__clock made-c-input__button"
                  variant="ghost"
                  size="small"
                  icon-name="time/outline/clock-outline"
                  aria-label=${this.localize.term('selectTime')}
                  @click=${() => this.input.showPicker()}
                ></made-icon-button>
              `
            : ''}
          <span part="end" class="made-c-input__end">
            <slot name="end"></slot>
          </span>
        </div>
        <div
          part="supporting-text"
          id="supporting-text"
          class="made-c-form-control__supporting-text"
          role=${ifDefined(role)}
        >
          ${supportingText}
        </div>
      </div>
    `;
    }
    /**
     * Reset the text field to its default value.
     */
    reset() {
        this.value = this.getAttribute('value') ?? '';
        this.hasNativeError = false;
        this.nativeErrorText = '';
    }
    getErrorText() {
        return this.error ? this.errorText : this.nativeErrorText;
    }
    [getFormValue]() {
        return this.value;
    }
    formResetCallback() {
        this.reset();
    }
    // eslint-disable-next-line @typescript-eslint/no-shadow
    formStateRestoreCallback(state) {
        this.value = state;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [createValidator]() {
        return new TextFieldValidator(() => ({
            state: this,
            renderedControl: this.input
        }));
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [getValidityAnchor]() {
        return this.input;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [onReportValidity](invalidEvent) {
        // Prevent default pop-up behavior.
        invalidEvent?.preventDefault();
        this.hasNativeError = !!invalidEvent;
        this.nativeErrorText = this.validationMessage;
    }
}
MadeInput.styles = [componentStyles, formControlStyles, styles];
MadeInput.dependencies = {
    'made-icon': MadeIcon,
    'made-icon-button': MadeIconButton
};
export default MadeInput;
__decorate([
    query('input')
], MadeInput.prototype, "input", void 0);
__decorate([
    state()
], MadeInput.prototype, "hasFocus", void 0);
__decorate([
    state()
], MadeInput.prototype, "passwordVisible", void 0);
__decorate([
    state()
], MadeInput.prototype, "hasNativeError", void 0);
__decorate([
    state()
], MadeInput.prototype, "nativeErrorText", void 0);
__decorate([
    property()
], MadeInput.prototype, "title", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeInput.prototype, "error", void 0);
__decorate([
    property({ attribute: 'error-text' })
], MadeInput.prototype, "errorText", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeInput.prototype, "success", void 0);
__decorate([
    property({ attribute: 'success-text' })
], MadeInput.prototype, "successText", void 0);
__decorate([
    property({ reflect: true })
], MadeInput.prototype, "type", void 0);
__decorate([
    property()
], MadeInput.prototype, "value", void 0);
__decorate([
    property()
], MadeInput.prototype, "label", void 0);
__decorate([
    property({ reflect: true })
], MadeInput.prototype, "size", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeInput.prototype, "supportingText", void 0);
__decorate([
    property({ type: Boolean })
], MadeInput.prototype, "clearable", void 0);
__decorate([
    property()
], MadeInput.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeInput.prototype, "readonly", void 0);
__decorate([
    property({ attribute: 'password-toggle', type: Boolean })
], MadeInput.prototype, "passwordToggle", void 0);
__decorate([
    property({ attribute: 'no-spinner', type: Boolean })
], MadeInput.prototype, "noSpinner", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeInput.prototype, "required", void 0);
__decorate([
    property({ attribute: 'required-indicator', reflect: true })
], MadeInput.prototype, "requiredIndicator", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeInput.prototype, "optional", void 0);
__decorate([
    property({
        type: Boolean,
        reflect: true,
        attribute: 'no-required-indicator'
    })
], MadeInput.prototype, "noRequiredIndicator", void 0);
__decorate([
    property()
], MadeInput.prototype, "pattern", void 0);
__decorate([
    property({ type: Number })
], MadeInput.prototype, "minlength", void 0);
__decorate([
    property({ type: Number })
], MadeInput.prototype, "maxlength", void 0);
__decorate([
    property()
], MadeInput.prototype, "min", void 0);
__decorate([
    property()
], MadeInput.prototype, "max", void 0);
__decorate([
    property({ attribute: 'character-count', type: Boolean })
], MadeInput.prototype, "characterCount", void 0);
__decorate([
    property()
], MadeInput.prototype, "step", void 0);
__decorate([
    property()
], MadeInput.prototype, "autocapitalize", void 0);
__decorate([
    property()
], MadeInput.prototype, "autocomplete", void 0);
__decorate([
    property({ type: Boolean })
], MadeInput.prototype, "autofocus", void 0);
__decorate([
    property()
], MadeInput.prototype, "enterkeyhint", void 0);
__decorate([
    property()
], MadeInput.prototype, "inputmode", void 0);
