// @mc-made/web-components/dist/react/made-tooltip/index.js
import * as React from 'react';
import { createComponent } from '../create-component.js';
import Component from '../../components/tooltip/src/tooltip.component.js';
const tagName = 'made-tooltip';
Component.define(tagName);
/**
 * @summary Tooltips display contextual information about an element when it receives focus or is hovered over.
 * @documentation '../stories/tooltip.mdx'
 *
 * @fires mde-show - Emitted when the tooltip opens.
 * @fires mde-hide - Emitted when the tooltip closes.
 *
 * @slot - The anchor element that triggers the tooltip.
 *
 * @csspart tooltip - The tooltip's popover container.
 * @csspart header - The tooltip header text.
 * @csspart content - The tooltip body text.
 *
 * @cssproperty --max-width - Controls the maximum width of the tooltip. Default: `18rem`.
 */
const reactWrapper = createComponent({
    tagName,
    elementClass: Component,
    react: React,
    events: {
        onMdeShow: 'mde-show',
        onMdeHide: 'mde-hide'
    },
    displayName: 'MadeTooltip'
});
export default reactWrapper;
