// @mc-made/web-components/dist/components/tab-panel/src/tab-panel.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './tab-panel.styles.js';
let id = 0;
/**
 * @summary A Tab Panel is used inside a Tabs component.
 * A Tab Panel displays the content associated with a given Tab.
 * @documentation '../stories/tab-panel.mdx'
 *
 * @slot - The tab panel's content.
 *
 * @csspart base - The component's base wrapper.
 *
 * @cssproperty --padding - The tab-panels padding.
 */
class MadeTabPanel extends MadeElement {
    constructor() {
        super(...arguments);
        this.tabPanelId = ++id;
        this.elementId = `made-tab-panel-${this.tabPanelId}`;
        /** The tab panel's name. */
        this.name = '';
        /**
         * @internal
         *  When true, the tab panel will be shown. */
        this.active = false;
    }
    connectedCallback() {
        super.connectedCallback();
        this.id = this.id.length > 0 ? this.id : this.elementId;
        this.setAttribute('role', 'tabpanel');
    }
    updated(changedProperties) {
        if (changedProperties.has('active')) {
            this.setAttribute('aria-hidden', this.active ? 'false' : 'true');
        }
    }
    render() {
        return html `
      <slot
        part="base"
        class=${classMap({
            'made-c-tab-panel': true
        })}
      ></slot>
    `;
    }
}
MadeTabPanel.styles = [componentStyles, styles];
export default MadeTabPanel;
__decorate([
    property({ reflect: true })
], MadeTabPanel.prototype, "name", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTabPanel.prototype, "active", void 0);
