// @mc-made/web-components/dist/react/made-accordion-item/index.js
import * as React from 'react';
import { createComponent } from '../create-component.js';
import Component from '../../components/accordion-item/src/accordion-item.component.js';
const tagName = 'made-accordion-item';
Component.define(tagName);
/**
 * @summary An accordion item is an individual section within an accordion component that can be expanded or collapsed.
 * @documentation '../../components/accordion-item/stories/accordion-item.mdx'
 *
 * @dependency made-icon
 *
 * @fires mde-toggle - Emitted when the accordion item is toggled.
 * @fires mde-show - Emitted when the accordion item is opened.
 * @fires mde-hide - Emitted when the accordion item is closed.
 *
 * @slot - The accordion item content that will be shown/hidden.
 * @slot badge - Used to place a badge in the header area.
 * @slot header-start - An optional leading icon in the header.
 * @slot header-end - An optional trailing icon in the header.
 *
 * @csspart base - The component's base wrapper.
 * @csspart trigger - The clickable trigger area that contains header and subtitle.
 * @csspart header - The header row containing title and icons.
 * @csspart header-content - The container for the header content.
 * @csspart subtitle - The subtitle container.
 * @csspart subtitle-content - The inner wrapper for subtitle content.
 * @csspart icon - The chevron icon that indicates expanded/collapsed state.
 * @csspart content - The collapsible content container.
 */
const reactWrapper = createComponent({
    tagName,
    elementClass: Component,
    react: React,
    events: {
        onMdeToggle: 'mde-toggle',
        onMdeShow: 'mde-show',
        onMdeHide: 'mde-hide'
    },
    displayName: 'MadeAccordionItem'
});
export default reactWrapper;
