// @mc-made/web-components/dist/react/made-input/index.js
import * as React from 'react';
import { createComponent } from '../create-component.js';
import Component from '../../components/input/src/input.component.js';
const tagName = 'made-input';
Component.define('made-input');
/**
 * @summary Inputs collect data from the user.
 * @documentation '../../components/input/stories/input.mdx'
 *
 * @dependency made-icon
 *
 * @slot label - The input's label. Alternatively, you can use the `label` attribute.
 * @slot start - Optional slot used to place an icon or prefix at the start of the input.
 * @slot end - Optional slot used to place an icon or suffix at the end of the input.
 * @slot clear-icon - An icon to use in lieu of the default clear icon.
 * @slot show-password-icon - An icon to use in lieu of the default show password icon.
 * @slot hide-password-icon - An icon to use in lieu of the default hide password icon.
 *
 * @fires change {Event} - Fired when an alteration to the control's value is committed by the user.
 * @fires input {Event} - Fired when the control receives input.
 * @fires invalid {Event} - Fired when the control has been checked for validity and doesn't satisfy its constraints.
 * @fires mde-clear {CustomEvent} - Fired when the clear button is activated.
 *
 * @csspart form-control - The form control that wraps the label, input, and help text.
 * @csspart form-control-label - The label's wrapper.
 * @csspart form-control-character-count - The character-count's wrapper.
 * @csspart form-control-input - The input's wrapper.
 * @csspart form-control-supporting-text - The help text's wrapper.
 * @csspart base - The component's base wrapper.
 * @csspart input - The internal `<input>` control.
 * @csspart start - The container that wraps the start slot.
 * @csspart clear-button - The clear button.
 * @csspart password-toggle-button - The password toggle button.
 * @csspart end - The container that wraps the end slot.
 */
const reactWrapper = createComponent({
    tagName,
    elementClass: Component,
    react: React,
    events: {
        onChange: 'change',
        onMdeClear: 'mde-clear'
    },
    displayName: 'MadeInput'
});
export default reactWrapper;
