// @mc-made/web-components/dist/react/made-alert/index.js
import * as React from 'react';
import { createComponent } from '../create-component.js';
import Component from '../../components/alert/src/alert.component.js';
const tagName = 'made-alert';
Component.define('made-alert');
/**
 * @summary Alerts display important messages inline or as toast notifications,
 * providing contextual information in response to user actions or system events.
 * @documentation '../stories/alert.mdx'
 *
 * @dependency made-icon-button
 *
 * @fires mde-show       {CustomEvent} - Emitted when the alert begins to show.
 * @fires mde-after-show {CustomEvent} - Emitted after the alert finishes showing.
 * @fires mde-hide       {CustomEvent} - Emitted when the alert begins to hide.
 * @fires mde-after-hide {CustomEvent} - Emitted after the alert finishes hiding.
 *
 * @slot header-start - To display a sentiment icon or avatar.
 * @slot header-end - Optional content (icon) to display at the end of the alert header.
 * @slot default - HTML content for the alert message. Can be used instead of the `alert-message` attribute for more complex content.
 * @slot actions - Optional action elements placed at the bottom of the alert.
 *
 * @csspart base - The component's base wrapper.
 * @csspart header-start - The container that wraps the icon area.
 * @csspart header-end - The container that wraps the icon area.
 * @csspart body - The container that wraps the title and content.
 * @csspart alert-title - The alert title.
 * @csspart alert-timestamp - The alert timestamp.
 * @csspart message - The alert message content.
 * @csspart actions - The container that wraps the actions slot.
 * @csspart dismiss-button - The dismiss button (made-icon-button).
 * @csspart dismiss-button-base - The dismiss button's base part.
 */
const reactWrapper = createComponent({
    tagName,
    elementClass: Component,
    react: React,
    events: {
        onMdeShow: 'mde-show',
        onMdeAfterShow: 'mde-after-show',
        onMdeHide: 'mde-hide',
        onMdeAfterHide: 'mde-after-hide'
    },
    displayName: 'MadeAlert'
});
export default reactWrapper;
