// @mc-made/web-components/dist/components/otp/src/otp.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { live } from 'lit/directives/live.js';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { mixinOnReportValidity, onReportValidity } from '../../../internal/face/on-report-validity.js';
import { property, query, state } from 'lit/decorators.js';
import { TextFieldValidator } from '../../../internal/face/validators/input-validator.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/src/icon.component.js';
import styles from './otp.styles.js';
/**
 * @summary OTP captures one-time passcodes with a single accessible input and segmented visual slots.
 * @documentation '../stories/otp.mdx'
 *
 * @fires mde-complete {CustomEvent} - Emitted when the entered code reaches the configured length.
 * @fires change {Event} - Fired when the user commits a value change.
 *
 * @fires input {InputEvent} - Fired whenever the value changes from user input.
 * @fires invalid {Event} - Fired when the control has been checked for validity and does not satisfy its constraints.
 *
 * @slot label - The otp field's label. Alternatively, use the `label` attribute.
 * @slot supporting-text - Optional instructional or validation text below the label.
 *
 * @method focus(options?: FocusOptions) - Moves focus to the internal `<input>` element.
 * @method blur() - Removes focus from the internal `<input>` element.
 * @method checkValidity() - Returns `true` if the control satisfies its validation constraints.
 * @method reportValidity() - Checks validity and, if invalid, shows a validation message. Returns `true` if valid.
 * @method setCustomValidity(message: string) - Sets a custom validation message. Pass an empty string to clear.
 *
 * @csspart label - The `<label>` element.
 * @csspart input-container - The wrapper that contains the visual slots and hidden input.
 * @csspart slot - Each visual OTP slot.
 * @csspart slot-character - The visible character inside each slot.
 * @csspart input - The internal `<input>` element.
 * @csspart validation-text - The validation message element.
 */
const otpBaseClass = mixinOnReportValidity(mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement))));
class MadeOtp extends otpBaseClass {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, 'label');
        this.hasFocus = false;
        this.hasNativeError = false;
        this.nativeErrorText = '';
        this.focusedSlotIndex = -1;
        /** The `autocomplete` attribute value to use when `otp` is true. Set to "one-time-code"
         * by default to enable OTP autofill on supported platforms.
         */
        this.autocomplete = 'one-time-code';
        /** The current value of the input. */
        this.value = '';
        /** The label for the input. */
        this.label = '';
        /** The number of characters in the OTP. */
        this.length = 6;
        /** The size of the input. */
        this.size = 'medium';
        /** If used in a form, marks the input as required. */
        this.required = false;
        /** Automatically focuses the input when the component is added to the DOM. */
        this.autofocus = false;
        /** The type of input to use. */
        this.type = 'numeric';
        /** When true, the component is optimized for entering OTPs, including enabling the `one-time-code`
         * autocomplete value on supported platforms.
         */
        this.oneTimeCode = false;
        /** A regex pattern that each character must match. Only applied when `type` is "text". */
        this.pattern = '';
        /** When set, the input loses focus automatically once all slots are filled. */
        this.blurOnComplete = false;
        /** When set, entered characters are replaced with a bullet (•) in the visual slots. */
        this.mask = false;
        /** A space-delimited list of group sizes. Each number defines how many contiguous slots appear in that group, divided by a dot separator. E.g. `"3 2 3"` renders 3 slots, separator, 2 slots, separator, 3 slots. When omitted, all slots form a single group. */
        this.groups = '';
        /** When true, character inputs are visually separated from each other. */
        this.separated = false;
        /** When true, entered characters are converted to uppercase. */
        this.uppercase = false;
    }
    get parsedGroups() {
        if (!this.groups?.trim())
            return [];
        return this.groups
            .trim()
            .split(/\s+/)
            .map(n => Math.max(1, parseInt(n, 10) || 1));
    }
    get normalizedLength() {
        const parsed = this.parsedGroups;
        return parsed.length > 0
            ? parsed.reduce((sum, n) => sum + n, 0)
            : this.length;
    }
    get slotValues() {
        return Array.from({ length: this.normalizedLength }, (_, index) => {
            return this.value[index] ?? '';
        });
    }
    get activeSlotIndex() {
        if (!this.hasFocus || this.disabled) {
            return -1;
        }
        if (this.value.length === this.normalizedLength) {
            return this.focusedSlotIndex;
        }
        return Math.min(this.value.length, this.normalizedLength - 1);
    }
    get computedGroups() {
        const parsed = this.parsedGroups;
        if (parsed.length === 0) {
            return [Array.from({ length: this.normalizedLength }, (_, i) => i)];
        }
        const groups = [];
        let slotIndex = 0;
        for (const size of parsed) {
            groups.push(Array.from({ length: size }, (_, i) => slotIndex + i));
            slotIndex += size;
        }
        return groups;
    }
    updated(changedProperties) {
        if (changedProperties.has('length') ||
            changedProperties.has('inputmode') ||
            changedProperties.has('groups')) {
            const normalizedValue = this.normalizeValue(this.value);
            if (normalizedValue !== this.value) {
                this.value = normalizedValue;
            }
        }
        if (changedProperties.has('value') && this.input) {
            const normalizedValue = this.normalizeValue(this.value);
            if (normalizedValue !== this.value) {
                this.value = normalizedValue;
                return;
            }
            this.input.value = this.value;
            // Restore slot-navigation selection after Lit syncs input.value (which resets the cursor).
            if (this.hasFocus &&
                this.value.length === this.normalizedLength &&
                this.focusedSlotIndex >= 0) {
                this.input.setSelectionRange(this.focusedSlotIndex, this.focusedSlotIndex + 1);
            }
        }
    }
    focus(options) {
        this.input?.focus(options);
    }
    blur() {
        this.input?.blur();
    }
    firstUpdated() {
        if (this.autofocus) {
            this.focus();
        }
    }
    render() {
        const hasLabel = this.hasSlotController.test('label') || !!this.label;
        return html `
      <div
        part="input-container"
        class=${classMap({
            'made-c-otp': true,
            'made-c-otp--large': this.size === 'large',
            'made-c-otp--separated': this.separated,
            'made-c-otp--error': this.hasNativeError,
            'made-c-otp--focused': this.hasFocus,
            'made-c-otp--disabled': this.disabled
        })}
        @click=${this.handleContainerClick}
      >
        <div class="made-c-otp__slots" aria-hidden="true">
          ${this.computedGroups.map((group, groupIndex) => html `
              ${groupIndex > 0
            ? html `<div class="made-c-otp__separator" role="presentation">
                    <made-icon
                      name="otp-separator"
                      library="system"
                      class="made-c-otp__separator-dot"
                    ></made-icon>
                  </div>`
            : nothing}
              <div class="made-c-otp__group">
                ${group.map(slotIndex => html `
                    <div
                      part="slot"
                      class=${classMap({
            'made-c-otp__slot': true,
            'made-c-otp__slot--filled': !!this.slotValues[slotIndex],
            'made-c-otp__slot--active': slotIndex === this.activeSlotIndex
        })}
                    >
                      <span
                        part="slot-character"
                        class="made-c-otp__slot-character"
                        >${this.mask && this.slotValues[slotIndex]
            ? '•'
            : this.slotValues[slotIndex]}</span
                      >
                    </div>
                  `)}
              </div>
            `)}
        </div>

        <input
          part="input"
          id="input"
          class="made-c-otp__input"
          type="text"
          inputmode=${this.type === 'numeric' ? 'numeric' : 'text'}
          title=${this.title}
          name=${this.name}
          ?disabled=${this.disabled}
          ?required=${this.required}
          autocomplete=${this.oneTimeCode ? this.autocomplete : 'off'}
          pattern=${ifDefined(this.pattern || undefined)}
          minlength=${this.normalizedLength}
          maxlength=${this.normalizedLength}
          aria-describedby="validation-text"
          .value=${live(this.value)}
          @keydown=${this.handleKeyDown}
          @input=${this.handleInput}
          @change=${this.handleChange}
          @focus=${this.handleFocus}
          @blur=${this.handleBlur}
          @paste=${this.handlePaste}
        />
      </div>

      ${hasLabel
            ? html `
            <label part="label" class="made-c-otp__label" for="input">
              ${this.label}
            </label>
          `
            : nothing}

      <div
        part="validation-text"
        id="validation-text"
        class="made-c-otp__validation-text"
        ?hidden=${!this.nativeErrorText}
        role=${ifDefined(this.nativeErrorText ? 'alert' : undefined)}
      >
        ${this.nativeErrorText}
      </div>
    `;
    }
    handleContainerClick() {
        if (!this.disabled) {
            this.focus();
        }
    }
    handleKeyDown(event) {
        if (this.value.length < this.normalizedLength)
            return;
        const { key } = event;
        if (key !== 'ArrowLeft' &&
            key !== 'ArrowRight' &&
            key !== 'ArrowUp' &&
            key !== 'ArrowDown') {
            return;
        }
        event.preventDefault();
        const lastIndex = this.normalizedLength - 1;
        const current = this.focusedSlotIndex < 0 ? lastIndex : this.focusedSlotIndex;
        let next = current;
        if (key === 'ArrowLeft')
            next = Math.max(0, current - 1);
        else if (key === 'ArrowRight')
            next = Math.min(lastIndex, current + 1);
        else if (key === 'ArrowUp')
            next = 0;
        else if (key === 'ArrowDown')
            next = lastIndex;
        this.focusedSlotIndex = next;
        this.input.setSelectionRange(next, next + 1);
    }
    handleInput(event) {
        const target = event.target;
        const nextValue = this.normalizeValue(target.value);
        if (target.value !== nextValue) {
            target.value = nextValue;
        }
        this.value = nextValue;
        // Keep the focused slot in sync with where the cursor landed after typing.
        if (nextValue.length === this.normalizedLength) {
            const cursorPos = target.selectionStart ?? this.normalizedLength;
            this.focusedSlotIndex = Math.min(cursorPos, this.normalizedLength - 1);
        }
        if (this.hasNativeError) {
            this.reportValidity();
        }
        if (this.value.length === this.normalizedLength) {
            this.emit('mde-complete');
            if (this.blurOnComplete) {
                this.blur();
            }
        }
    }
    handleChange() {
        if (this.hasNativeError) {
            this.reportValidity();
        }
        this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
    }
    handleFocus() {
        this.hasFocus = true;
        // When the OTP is full, seed the focused slot synchronously so the active
        // highlight appears immediately, then defer the selection so it runs after
        // the browser's default select-all on tab-focus.
        if (this.value.length === this.normalizedLength) {
            this.focusedSlotIndex = this.normalizedLength - 1;
        }
        requestAnimationFrame(() => {
            const len = this.value.length;
            if (len === this.normalizedLength) {
                this.input?.setSelectionRange(this.focusedSlotIndex, this.focusedSlotIndex + 1);
            }
            else {
                this.input?.setSelectionRange(len, len);
            }
        });
    }
    handleBlur() {
        this.hasFocus = false;
        this.focusedSlotIndex = -1;
        this[onReportValidity](null);
    }
    handlePaste(event) {
        event.preventDefault();
        const pasted = event.clipboardData?.getData('text') ?? '';
        const stripped = pasted.replace(/[^a-zA-Z0-9]/g, '');
        this.value = this.normalizeValue(stripped);
        if (this.value.length === this.normalizedLength) {
            this.emit('mde-complete');
            if (this.blurOnComplete) {
                this.blur();
            }
        }
    }
    normalizeValue(value) {
        const sanitizedValue = this.type === 'numeric' ? value.replace(/\D/g, '') : value;
        const casedValue = this.uppercase
            ? sanitizedValue.toUpperCase()
            : sanitizedValue;
        return casedValue.slice(0, this.normalizedLength);
    }
    [getFormValue]() {
        return this.value;
    }
    formResetCallback() {
        this.value = this.getAttribute('value') ?? '';
        this.hasNativeError = false;
        this.nativeErrorText = '';
    }
    formStateRestoreCallback(state) {
        this.value = this.normalizeValue(state);
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [createValidator]() {
        return new TextFieldValidator(() => ({
            state: {
                value: this.value,
                required: this.required,
                minlength: this.normalizedLength,
                maxlength: this.normalizedLength,
                type: 'text',
                pattern: this.type === 'numeric'
                    ? `\\d{${this.normalizedLength}}`
                    : this.pattern,
                min: '',
                max: '',
                step: ''
            },
            renderedControl: this.input ?? null
        }));
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [getValidityAnchor]() {
        return this.input;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [onReportValidity](invalidEvent) {
        invalidEvent?.preventDefault();
        this.hasNativeError = !!invalidEvent;
        this.nativeErrorText = this.validationMessage;
    }
}
MadeOtp.styles = [componentStyles, styles];
MadeOtp.dependencies = {
    'made-icon': MadeIcon
};
export default MadeOtp;
__decorate([
    query('input')
], MadeOtp.prototype, "input", void 0);
__decorate([
    state()
], MadeOtp.prototype, "hasFocus", void 0);
__decorate([
    state()
], MadeOtp.prototype, "hasNativeError", void 0);
__decorate([
    state()
], MadeOtp.prototype, "nativeErrorText", void 0);
__decorate([
    state()
], MadeOtp.prototype, "focusedSlotIndex", void 0);
__decorate([
    property()
], MadeOtp.prototype, "value", void 0);
__decorate([
    property()
], MadeOtp.prototype, "label", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MadeOtp.prototype, "length", void 0);
__decorate([
    property({ reflect: true })
], MadeOtp.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeOtp.prototype, "required", void 0);
__decorate([
    property({ type: Boolean })
], MadeOtp.prototype, "autofocus", void 0);
__decorate([
    property()
], MadeOtp.prototype, "type", void 0);
__decorate([
    property({ type: Boolean, attribute: 'one-time-code' })
], MadeOtp.prototype, "oneTimeCode", void 0);
__decorate([
    property()
], MadeOtp.prototype, "pattern", void 0);
__decorate([
    property({ type: Boolean, attribute: 'blur-on-complete' })
], MadeOtp.prototype, "blurOnComplete", void 0);
__decorate([
    property({ type: Boolean })
], MadeOtp.prototype, "mask", void 0);
__decorate([
    property()
], MadeOtp.prototype, "groups", void 0);
__decorate([
    property({ type: Boolean })
], MadeOtp.prototype, "separated", void 0);
__decorate([
    property({ type: Boolean })
], MadeOtp.prototype, "uppercase", void 0);
