// @mc-made/web-components/dist/components/otp/src/otp.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: inline-block;
    width: fit-content;
  }

  .made-c-otp {
    --font-size: var(--made-font-size-body-default);
    --line-height-fallback: calc(var(--font-size) * 1.45);
    --line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
    position: relative;
    display: inline-block;
    width: fit-content;
    font-family: var(--made-font-family-primary);
    font-size: var(--font-size);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--line-height, --line-height-fallback);
    color: var(--made-color-text-primary);
    cursor: text;
  }

  .made-c-otp.made-c-otp--disabled {
    cursor: not-allowed;
  }

  .made-c-otp__slots {
    display: flex;
    align-items: center;
  }

  .made-c-otp--separated .made-c-otp__group {
    box-shadow: none;
    gap: var(--made-size-space-3-x);
  }

  .made-c-otp--separated .made-c-otp__slot {
    border-right-width: var(--made-component-form-control-border-width);
    border-right-style: solid;
    border-right-color: var(--made-component-form-control-border-color);
    border-radius: var(--made-component-form-control-border-radius);
  }

  .made-c-otp--separated.made-c-otp--error .made-c-otp__slot {
    border-right-color: var(--made-color-border-danger-strong);
  }

  .made-c-otp--separated.made-c-otp--disabled .made-c-otp__slot {
    border-right-color: var(--made-color-border-disabled);
  }

  .made-c-otp__group {
    display: flex;
    align-items: stretch;
    box-shadow: var(--made-component-form-control-box-shadow);
  }

  .made-c-otp__slot {
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    width: var(--made-size-space-10-x);
    height: var(--made-size-space-10-x);
    border-width: var(--made-component-form-control-border-width);
    border-style: solid;
    border-color: var(--made-component-form-control-border-color);
    border-right: none;
    border-radius: 0;
    background-color: var(--made-component-form-control-background);
    overflow: hidden;
    box-shadow: 0 0 0 0 var(--made-color-border-focused);
    transition:
      box-shadow var(--made-motion-duration-short) cubic-bezier(0.4, 0, 0.2, 1),
      border-color var(--made-motion-duration-short)
        cubic-bezier(0.4, 0, 0.2, 1);
  }

  .made-c-otp__group .made-c-otp__slot:first-child {
    border-top-left-radius: var(--made-component-form-control-border-radius);
    border-bottom-left-radius: var(--made-component-form-control-border-radius);
  }

  .made-c-otp__group .made-c-otp__slot:last-child {
    border-right-width: var(--made-component-form-control-border-width);
    border-right-style: solid;
    border-right-color: var(--made-component-form-control-border-color);
    border-top-right-radius: var(--made-component-form-control-border-radius);
    border-bottom-right-radius: var(
      --made-component-form-control-border-radius
    );
  }

  .made-c-otp__slot-character {
    font-variant-numeric: tabular-nums;
    font-size: var(--made-font-size-body-default);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--made-line-height-rem-04);
    letter-spacing: var(--made-letter-spacing-default);
    color: var(--made-color-text-primary);
  }

  .made-c-otp__slot--active {
    border-color: transparent;
    box-shadow: 0 0 0 2px var(--made-color-border-focused);
    z-index: 10;

    .made-c-otp--separated & {
      border-right-color: transparent;
    }
  }

  /* Blinking caret shown in the active slot when it has no character */
  .made-c-otp__slot--active:not(.made-c-otp__slot--filled)::after {
    content: '';
    display: block;
    position: absolute;
    width: 1px;
    height: 1em;
    background-color: var(--made-color-text-primary);
    border-radius: 1px;
    animation: made-otp-caret 1.25s ease-out infinite;
  }

  @keyframes made-otp-caret {
    0%,
    70%,
    to {
      opacity: 1;
    }
    20%,
    50% {
      opacity: 0;
    }
  }

  .made-c-otp__group .made-c-otp__slot--active:first-child {
    border-top-left-radius: var(--made-component-form-control-border-radius);
    border-bottom-left-radius: var(--made-component-form-control-border-radius);
  }

  .made-c-otp__group .made-c-otp__slot--active:last-child {
    border-right-color: transparent;
    border-top-right-radius: var(--made-component-form-control-border-radius);
    border-bottom-right-radius: var(
      --made-component-form-control-border-radius
    );
  }

  .made-c-otp__separator {
    display: flex;
    align-items: center;
    justify-content: center;
    flex-shrink: 0;
    width: var(--made-size-space-10-x);
    height: var(--made-size-space-10-x);
    padding: var(--made-size-space-2-x);
  }

  .made-c-otp__separator-dot {
    font-size: 4px;
    color: var(--made-color-icon-primary);
  }

  .made-c-otp.made-c-otp--error .made-c-otp__slot {
    border-color: var(--made-color-border-danger-strong);
  }

  .made-c-otp.made-c-otp--error
    .made-c-otp__group
    .made-c-otp__slot:last-child {
    border-right-color: var(--made-color-border-danger-strong);
  }

  .made-c-otp.made-c-otp--disabled .made-c-otp__slot {
    background-color: var(--made-color-background-disabled);
    border-color: var(--made-color-border-disabled);
    color: var(--made-color-text-disabled);
  }

  .made-c-otp.made-c-otp--disabled
    .made-c-otp__group
    .made-c-otp__slot:last-child {
    border-right-color: var(--made-color-border-disabled);
  }

  .made-c-otp__input {
    position: absolute;
    inset: 0;
    width: 100%;
    height: 100%;
    margin: 0;
    padding: 0;
    border: none;
    opacity: 0;
    cursor: inherit;
  }

  .made-c-otp__input:focus {
    outline: none;
  }

  .made-c-otp__label {
    display: flex;
    align-items: center;
    margin-top: var(--made-size-space-2-x);
    font-family: var(--made-font-family-primary);
    color: var(--made-color-text-primary);
    font-weight: var(--made-font-weight-regular);
    font-size: var(--made-font-size-body-small);
    justify-content: center;
  }

  :host([size='large']) .made-c-otp__label {
    font-size: var(--made-font-size-body-default);
  }

  .made-c-otp__validation-text {
    display: flex;
    align-items: center;
    gap: var(--made-size-space-1-x);
    margin-top: var(--made-size-space-1-x);
    font-family: var(--made-font-family-primary);
    font-size: var(--made-font-size-body-small);
    font-weight: var(--made-font-weight-regular);
    color: var(--made-color-text-secondary);
  }

  .made-c-otp--large .made-c-otp__slot {
    width: var(--made-size-space-12-x);
    height: var(--made-size-space-12-x);
  }

  .made-c-otp--large .made-c-otp__slot-character {
    font-size: var(--made-font-size-body-large);
    line-height: var(--made-line-height-rem-05);
  }

  .made-c-otp--large .made-c-otp__separator {
    width: var(--made-size-space-12-x);
    height: var(--made-size-space-12-x);
  }
`;
