// @mc-made/web-components/dist/components/card/src/card-text.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  :host {
    --color: initial;
    display: block;
    position: relative;
  }

  .made-c-card__text {
    display: flex;
    flex-direction: column;
    gap: var(--made-size-space-2-x);
    padding: var(--made-component-card-padding-y)
      var(--made-component-card-padding-x);
    font-family: var(--made-font-family-primary);
  }

  .made-c-card__title-container {
    display: flex;
    flex-direction: column;
    gap: var(--made-size-space-2-x);

    &.made-c-card__title--top {
      flex-direction: column-reverse;
    }
  }

  .made-c-card__eyebrow {
    font-family: var(--made-eyebrow-small-font-family);
    font-weight: var(--made-eyebrow-small-font-weight);
    font-size: var(--made-font-size-02);
    line-height: var(--made-eyebrow-small-line-height);
    letter-spacing: var(--made-letter-spacing-eyebrow-small);
    text-transform: var(--made-text-case-uppercase);
    color: var(--color, var(--made-color-text-helper));
    margin: 0;
  }

  .made-c-card__title {
    font-family: var(--made-heading-h2-font-family);
    font-weight: var(--made-font-weight-h2);
    font-size: var(--made-font-size-h2);
    line-height: 1.25;
    letter-spacing: var(--made-letter-spacing-h2);
    color: var(--color, var(--made-color-text-primary));
    margin: 0;
  }

  .made-c-card__paragraph {
    font-family: var(--made-body-regular-medium-font-family);
    font-weight: var(--made-font-weight-regular);
    font-size: var(--made-font-size-body-default);
    line-height: var(--made-line-height-rem-04);
    letter-spacing: var(--made-letter-spacing-default);
    color: var(--color, var(--made-color-text-secondary));
    margin: 0;
    padding: 0;
  }
`;
