// @mc-made/web-components/dist/components/card/src/card.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html, literal } from 'lit/static-html.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './card.styles.js';
/**
 * @summary Cards can be used to group related information into flexible containers.
 * @documentation '../stories/card.mdx'
 *
 * @slot - The card's content.
 *
 * @csspart base - The component's base wrapper.
 *
 */
class MadeCard extends MadeElement {
    constructor() {
        super();
        this.title = ''; // make reactive to pass through
        /** Make the whole card a primary action area. */
        this.clickable = false;
        /** Give the card a horizontal layout. */
        this.horizontal = false;
        /** Disables a clickable card. */
        this.disabled = false;
        /**
         * When using `href`, this attribute will map to the underlying link's `rel` attribute. Unlike regular links, the
         * default is `noreferrer noopener` to prevent security exploits. However, if you're using `target` to point to a
         * specific tab/window, this will prevent that from working correctly. You can remove or change the default value by
         * setting the attribute to an empty string or a value of your choice, respectively.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Reference/Attributes/rel
         */
        this.rel = 'noreferrer noopener';
        // Prevent click events when disabled
        this.addEventListener('click', (e) => {
            if (this.disabled) {
                e.preventDefault();
                e.stopImmediatePropagation();
            }
        }, { capture: true });
    }
    isLink() {
        return this.href ? true : false;
    }
    /** Simulates a click on the button. */
    click() {
        if (this.clickable) {
            this.button?.click();
        }
    }
    /** Sets focus on the button. */
    focus(options) {
        if (this.clickable) {
            this.button?.focus(options);
        }
    }
    /** Removes focus from the button. */
    blur() {
        if (this.clickable) {
            this.button?.blur();
        }
    }
    get touchTarget() {
        const isLink = this.isLink();
        const tag = isLink ? literal `a` : literal `button`;
        /* eslint-disable lit/no-invalid-html */
        /* eslint-disable lit/binding-positions */
        return html `
          <${tag}
            part="base"
            class=${classMap({
            'made-c-card--clickable': true,
            'made-c-card--disabled': this.disabled
        })}
             title=${this.title /* An empty title prevents browser validation tooltips from appearing on hover */}
            ?disabled=${ifDefined(isLink ? undefined : this.disabled)}
            type=${ifDefined(!isLink ? 'button' : undefined)}
            href=${ifDefined(isLink ? this.href : undefined)}
            target=${ifDefined(isLink ? this.target : undefined)}
            rel=${ifDefined(isLink ? this.rel : undefined)}
            role=${ifDefined(isLink ? undefined : 'button')}
            aria-disabled=${this.disabled ? 'true' : 'false'}
            aria-label=${ifDefined(this.ariaLabel)}
            tabindex=${this.disabled ? '-1' : '0'}
          >
            <slot></slot>
          </${tag}>
          `;
        /* eslint-enable lit/no-invalid-html */
        /* eslint-enable lit/binding-positions */
    }
    render() {
        return this.clickable
            ? html `
          <div
            part="base"
            class=${classMap({
                'made-c-card': true,
                'made-c-card--horizontal': this.horizontal
            })}
          >
            ${this.touchTarget}
          </div>
        `
            : html `
          <div
            part="base"
            class=${classMap({
                'made-c-card': true,
                'made-c-card--horizontal': this.horizontal
            })}
          >
            <slot></slot>
          </div>
        `;
    }
}
MadeCard.styles = [componentStyles, styles];
export default MadeCard;
__decorate([
    query('.made-c-card--clickable')
], MadeCard.prototype, "button", void 0);
__decorate([
    property()
], MadeCard.prototype, "title", void 0);
__decorate([
    property({ type: Boolean })
], MadeCard.prototype, "clickable", void 0);
__decorate([
    property({ type: Boolean })
], MadeCard.prototype, "horizontal", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeCard.prototype, "disabled", void 0);
__decorate([
    property()
], MadeCard.prototype, "href", void 0);
__decorate([
    property()
], MadeCard.prototype, "target", void 0);
__decorate([
    property()
], MadeCard.prototype, "rel", void 0);
__decorate([
    property({ attribute: 'aria-label' })
], MadeCard.prototype, "ariaLabel", void 0);
