// @mc-made/web-components/dist/components/button/src/button.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { setupFormSubmitter } from '../../../internal/face/form-submitter.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, literal } from 'lit/static-html.js';
import { ifDefined } from 'lit/directives/if-defined.js';
import { internals, mixinElementInternals } from '../../../internal/face/element-internals.js';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeLoader from '../../loader/made-loader.js';
import styles from './button.styles.js';
/**
 * @summary Buttons represent actions that are available to the user.
 * @documentation '../stories/button.mdx'
 *
 * @slot - The button's label.
 * @slot start - Used to place content before button text. Typically used for icons.
 * @slot end - Used to place content after button text. Typically used for icons.
 *
 * @csspart base - The component's base wrapper.
 * @csspart start - The container that wraps the start slot.
 * @csspart label - The button's label.
 * @csspart end - The container that wraps the end slot.
 */
const buttonBaseClass = mixinElementInternals(MadeElement);
class MadeButton extends buttonBaseClass {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'start', 'end');
        this.title = ''; // make reactive to pass through
        /** The button's theme variant. */
        this.variant = 'primary';
        /** The button's size. */
        this.size = 'medium';
        /** Disables the button. */
        this.disabled = false;
        /**
         * The type of button. Note that the default value is `button` instead of `submit`, which is opposite of how native
         * `<button>` elements behave. When the type is `submit`, the button will submit the surrounding form.
         */
        this.type = 'button';
        /**
         * When using `href`, this attribute will map to the underlying link's `rel` attribute. Unlike regular links, the
         * default is `noreferrer noopener` to prevent security exploits. However, if you're using `target` to point to a
         * specific tab/window, this will prevent that from working correctly. You can remove or change the default value by
         * setting the attribute to an empty string or a value of your choice, respectively.
         */
        this.rel = 'noreferrer noopener';
        /** Controls whether the button is in loading state. */
        this.loading = false;
        /**
         * If true, the button will take up the full width of its container.
         */
        this.fullWidth = false;
    }
    /**
     * The associated form element with which this element's value will submit.
     */
    get form() {
        return this[internals].form;
    }
    isLink() {
        return this.href ? true : false;
    }
    /** Simulates a click on the button. */
    click() {
        this.button.click();
    }
    /** Sets focus on the button. */
    focus(options) {
        this.button.focus(options);
    }
    /** Removes focus from the button. */
    blur() {
        this.button.blur();
    }
    render() {
        const isLink = this.isLink();
        const tag = isLink ? literal `a` : literal `button`;
        const hasStartSlot = this.hasSlotController.test('start');
        const hasEndSlot = this.hasSlotController.test('end');
        /* eslint-disable lit/no-invalid-html */
        /* eslint-disable lit/binding-positions */
        return html `
      <${tag}
        part="base"
        class=${classMap({
            'made-c-button': true,
            'made-c-button--primary': this.variant === 'primary',
            'made-c-button--secondary': this.variant === 'secondary',
            'made-c-button--destructive': this.variant === 'destructive',
            'made-c-button--ghost': this.variant === 'ghost',
            'made-c-button--small': this.size === 'small',
            'made-c-button--medium': this.size === 'medium',
            'made-c-button--large': this.size === 'large',
            'made-c-button--has-label': this.hasSlotController.test('[default]'),
            'made-c-button--full-width': this.fullWidth,
            'made-c-button--loading': this.loading
        })}
        ?disabled=${ifDefined(isLink ? undefined : this.disabled)}
        type=${ifDefined(isLink ? undefined : this.type)}
        title=${this.title /* An empty title prevents browser validation tooltips from appearing on hover */}
        name=${ifDefined(isLink ? undefined : this.name)}
        value=${ifDefined(isLink ? undefined : this.value)}
        href=${ifDefined(isLink ? this.href : undefined)}
        target=${ifDefined(isLink ? this.target : undefined)}
        download=${ifDefined(isLink ? this.download : undefined)}
        rel=${ifDefined(isLink ? this.rel : undefined)}
        role=${ifDefined(isLink ? undefined : 'button')}
        aria-disabled=${this.disabled ? 'true' : 'false'}
        tabindex=${this.disabled ? '-1' : '0'}
      >
        <slot name="start" part="start" class=${classMap({ 'made-c-button__start': hasStartSlot })}></slot>
        <slot part="label" class="made-c-button__label"></slot>
        <slot name="end" part="end" class=${classMap({ 'made-c-button__end': hasEndSlot })}></slot>
        ${this.loading ? html `<made-loader hide-track></made-loader>` : ''}
      </${tag}>
    `;
        /* eslint-enable lit/no-invalid-html */
        /* eslint-enable lit/binding-positions */
    }
}
MadeButton.styles = [componentStyles, styles];
(() => {
    setupFormSubmitter(MadeButton);
})();
MadeButton.formAssociated = true;
MadeButton.dependencies = {
    'made-loader': MadeLoader
};
export default MadeButton;
__decorate([
    query('.made-c-button')
], MadeButton.prototype, "button", void 0);
__decorate([
    property()
], MadeButton.prototype, "title", void 0);
__decorate([
    property({ reflect: true })
], MadeButton.prototype, "variant", void 0);
__decorate([
    property({ reflect: true })
], MadeButton.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeButton.prototype, "disabled", void 0);
__decorate([
    property({ reflect: true })
], MadeButton.prototype, "type", void 0);
__decorate([
    property()
], MadeButton.prototype, "name", void 0);
__decorate([
    property()
], MadeButton.prototype, "value", void 0);
__decorate([
    property()
], MadeButton.prototype, "href", void 0);
__decorate([
    property()
], MadeButton.prototype, "target", void 0);
__decorate([
    property()
], MadeButton.prototype, "rel", void 0);
__decorate([
    property()
], MadeButton.prototype, "download", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeButton.prototype, "loading", void 0);
__decorate([
    property({ type: Boolean, attribute: 'full-width' })
], MadeButton.prototype, "fullWidth", void 0);
