// @mc-made/web-components/dist/components/tooltip/src/tooltip.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './tooltip.styles.js';
class MadeTooltip extends MadeElement {
    constructor() {
        super(...arguments);
        this.anchorHasFocus = false;
        this.clickedViaPointer = false;
        this.supportsAnchor = 'anchorName' in document.documentElement.style;
        /** Tooltip header */
        this.header = '';
        /** Tooltip content */
        this.content = '';
        /** Position of the tooltip in relation to its anchor: top, bottom, left, right */
        this.position = 'top';
        /** Delay in ms before showing the tooltip on hover */
        this.delayDuration = 0;
        /** Whether to show a pointer (beak) on the tooltip */
        this.pointer = false;
        /** Whether the tooltip is currently visible */
        this.open = false;
        /** The interaction(s) that trigger the tooltip. Space-separated list of: hover, focus, click */
        this.trigger = 'focus hover';
        /** Disables the tooltip so it will not show */
        this.disabled = false;
        /** Appearance attribute inverts the tooltip's background and text colors */
        this.appearance = 'default';
        this._handleScrollResize = () => {
            this._updatePosition();
        };
        this.handleMouseOver = () => {
            clearTimeout(this.hoverTimeout);
            this.hoverTimeout = window.setTimeout(() => this.show(), this.delayDuration);
        };
        this.handleMouseOut = () => {
            // Only cancel a pending delayed show. Do NOT hide here — handleDocumentPointerMove
            // is already listening once the tooltip is open and will close it when the pointer
            // leaves both the anchor and the tooltip (including the beak / safe-area).
            clearTimeout(this.hoverTimeout);
            this.hoverTimeout = undefined;
        };
        this.handleTouchTap = (event) => {
            // Only handle non-mouse pointers (touch / stylus). Mouse pointers are handled
            // by mouseover/mouseout. Toggle open state on each tap.
            if (event.pointerType === 'mouse')
                return;
            if (this.open) {
                this.hide();
            }
            else {
                this.show();
            }
        };
        this.handleMouseDown = () => {
            this.clickedViaPointer = true;
            if (this.open) {
                this.hide();
            }
        };
        this.handleDocumentPointerMove = (event) => {
            // composedPath() is only valid synchronously — capture it before the rAF
            const path = event.composedPath();
            // rAF-gate: skip if a frame is already queued — avoids running at 60-120Hz
            if (this._pointerMoveFrame !== undefined)
                return;
            this._pointerMoveFrame = requestAnimationFrame(() => {
                this._pointerMoveFrame = undefined;
                if (!this.anchor || this.anchorHasFocus || !this.hasTrigger('hover'))
                    return;
                // Touch pointers don't linger — the pointer leaves as soon as the finger
                // lifts, which is handled by handleTouchTap. Skip hide for touch/stylus.
                if (event.pointerType !== 'mouse')
                    return;
                const isHoveringAnchor = path.includes(this.anchor);
                const isHoveringTooltip = path.includes(this.tooltipPopover);
                if (!isHoveringAnchor && !isHoveringTooltip) {
                    this.hide();
                }
            });
        };
        this.handleDocumentKeyDown = (event) => {
            // Hide the tooltip when escape is pressed
            if (event.key === 'Escape' && this.open) {
                event.preventDefault();
                this.handleBlur();
            }
        };
        this.handleWindowBlur = () => {
            // Window blur fires when focus leaves the page entirely (e.g. another frame/iframe).
            // In that case no blur event fires on the anchor, so anchorHasFocus stays true and
            // handleDocumentPointerMove never closes the tooltip. Reset state and hide.
            if (this.anchorHasFocus) {
                this.anchorHasFocus = false;
                this.hide();
            }
        };
        this.handlePointerUp = () => {
            if (this.open) {
                this.hide();
            }
            else {
                this.show();
            }
        };
        this.handleFocus = () => {
            if (this.clickedViaPointer) {
                this.clickedViaPointer = false;
                return;
            }
            this.anchorHasFocus = true;
            this.show();
        };
        this.handleBlur = () => {
            this.anchorHasFocus = false;
            this.hide();
        };
    }
    firstUpdated() {
        const tooltipId = `made-tooltip-${++MadeTooltip._idCounter}`;
        this.id = tooltipId;
        const slot = this.slotEl;
        this.anchor = slot.assignedElements()[0];
        if (!this.anchor)
            return;
        /** Set up event listeners */
        if (this.hasTrigger('hover')) {
            this.anchor.addEventListener('mouseover', this.handleMouseOver);
            this.anchor.addEventListener('mouseout', this.handleMouseOut);
            this.anchor.addEventListener('pointerdown', this.handleTouchTap);
        }
        if (this.hasTrigger('focus')) {
            this.anchor.addEventListener('mousedown', this.handleMouseDown);
            this.anchor.addEventListener('focus', this.handleFocus);
            this.anchor.addEventListener('blur', this.handleBlur);
        }
        if (this.hasTrigger('click')) {
            this.anchor.addEventListener('click', this.handlePointerUp);
        }
        /** Should be using aria-describedby here but aria-labelledby provides a better experience with popovers. */
        this.anchor.setAttribute('aria-labelledby', tooltipId);
        if (this.open) {
            this._doShow();
        }
    }
    updated(changedProperties) {
        if (!changedProperties.has('open'))
            return;
        if (this.open && !this.tooltipPopover.matches(':popover-open')) {
            this._doShow();
        }
        else if (!this.open && this.tooltipPopover.matches(':popover-open')) {
            this._doHide();
        }
    }
    disconnectedCallback() {
        clearTimeout(this.hoverTimeout);
        this.hoverTimeout = undefined;
        /** Clean up event listeners */
        this.anchor?.removeEventListener('mouseover', this.handleMouseOver);
        this.anchor?.removeEventListener('mouseout', this.handleMouseOut);
        this.anchor?.removeEventListener('pointerdown', this.handleTouchTap);
        this.anchor?.removeEventListener('mousedown', this.handleMouseDown);
        this.anchor?.removeEventListener('focus', this.handleFocus);
        this.anchor?.removeEventListener('blur', this.handleBlur);
        this.anchor?.removeEventListener('click', this.handlePointerUp);
        document.removeEventListener('pointermove', this.handleDocumentPointerMove);
        document.removeEventListener('keydown', this.handleDocumentKeyDown);
        window.removeEventListener('scroll', this._handleScrollResize, {
            capture: true
        });
        window.removeEventListener('resize', this._handleScrollResize);
        window.removeEventListener('blur', this.handleWindowBlur);
        super.disconnectedCallback();
    }
    hasTrigger(trigger) {
        return this.trigger.split(' ').includes(trigger);
    }
    _updatePosition() {
        // After the browser has had a chance to reposition (anchor or fallback),
        // re-detect which side the tooltip actually ended up on so the beak class stays correct.
        requestAnimationFrame(() => {
            this._detectActualPosition();
            if (!this.supportsAnchor)
                this._positionFallback();
        });
    }
    _detectActualPosition() {
        if (!this.anchor)
            return;
        const anchorRect = this.anchor.getBoundingClientRect();
        const tooltipRect = this.tooltipPopover.getBoundingClientRect();
        const anchorCenterX = anchorRect.left + anchorRect.width / 2;
        const anchorCenterY = anchorRect.top + anchorRect.height / 2;
        const tooltipCenterX = tooltipRect.left + tooltipRect.width / 2;
        const tooltipCenterY = tooltipRect.top + tooltipRect.height / 2;
        const dx = tooltipCenterX - anchorCenterX;
        const dy = tooltipCenterY - anchorCenterY;
        let detected;
        if (Math.abs(dy) >= Math.abs(dx)) {
            detected = dy < 0 ? 'top' : 'bottom';
        }
        else {
            detected = dx < 0 ? 'left' : 'right';
        }
        // Only re-render if the actual position differs from the preferred one.
        // Always compare against this.position (not calculatedPosition) so we don't
        // trigger a spurious update when reopening after a previous flip.
        if (detected !== this.position) {
            this.calculatedPosition = detected;
            void this.requestUpdate();
        }
        else {
            // Browser placed it in the preferred position — clear any stale flip.
            this.calculatedPosition = null;
        }
    }
    _animateClose() {
        // animate fade-out
        this.tooltipPopover
            .animate([
            { opacity: 1, transform: 'translateY(0) translateX(0)' },
            {
                opacity: 0,
                transform: 'translateY(var(--move-y)) translateX(var(--move-x))'
            }
        ], {
            duration: 150,
            easing: 'ease'
        })
            .finished.then(() => this.tooltipPopover.hidePopover()); // Actually close it after animation
    }
    _positionFallback() {
        const anchor = this.anchor;
        const tooltip = this.tooltipPopover;
        const anchorRect = anchor.getBoundingClientRect();
        const tooltipRect = tooltip.getBoundingClientRect();
        const gap = 8;
        let top;
        let left;
        switch (this.position) {
            case 'bottom':
                top = anchorRect.bottom + gap;
                left = anchorRect.left + anchorRect.width / 2 - tooltipRect.width / 2;
                break;
            case 'left':
                top = anchorRect.top + anchorRect.height / 2 - tooltipRect.height / 2;
                left = anchorRect.left - tooltipRect.width - gap;
                break;
            case 'right':
                top = anchorRect.top + anchorRect.height / 2 - tooltipRect.height / 2;
                left = anchorRect.right + gap;
                break;
            case 'top':
            default:
                top = anchorRect.top - tooltipRect.height - gap;
                left = anchorRect.left + anchorRect.width / 2 - tooltipRect.width / 2;
                break;
        }
        tooltip.style.setProperty('--fallback-top', `${top}px`);
        tooltip.style.setProperty('--fallback-left', `${left}px`);
    }
    _doShow() {
        this.tooltipPopover.showPopover();
        // After paint: detect which side the browser actually placed the tooltip on (may differ
        // from this.position if @position-try flipped it), then update the beak class.
        // For non-anchor browsers the fallback positioning also runs here.
        this._updatePosition();
        document.addEventListener('pointermove', this.handleDocumentPointerMove);
        document.addEventListener('keydown', this.handleDocumentKeyDown);
        window.addEventListener('blur', this.handleWindowBlur);
        window.addEventListener('scroll', this._handleScrollResize, {
            passive: true,
            capture: true
        });
        window.addEventListener('resize', this._handleScrollResize, {
            passive: true
        });
        this.emit('mde-show');
    }
    _doHide() {
        this.calculatedPosition = null;
        this._animateClose();
        document.removeEventListener('pointermove', this.handleDocumentPointerMove);
        document.removeEventListener('keydown', this.handleDocumentKeyDown);
        window.removeEventListener('blur', this.handleWindowBlur);
        window.removeEventListener('scroll', this._handleScrollResize, {
            capture: true
        });
        window.removeEventListener('resize', this._handleScrollResize);
        this.emit('mde-hide');
    }
    /** Shows the tooltip */
    show() {
        if (this.disabled)
            return;
        this.open = true;
    }
    /** Hides the tooltip */
    hide() {
        this.open = false;
    }
    render() {
        const anchorPosition = this.calculatedPosition || this.position;
        return html `
      <slot></slot>
      <div
        class=${classMap({
            'made-c-tooltip': true,
            'made-c-tooltip--has-pointer': this.pointer,
            [`made-c-tooltip--${anchorPosition}`]: true,
            'made-c-tooltip--inverted': this.appearance === 'inverted'
        })}
        id="tooltip-popover"
        part="tooltip"
        role="tooltip"
        aria-label=${this.header
            ? `${this.header}: ${this.content}`
            : this.content}
        popover="manual"
      >
        ${this.header
            ? html `<div class="made-c-tooltip__header" part="header">
              ${this.header}
            </div>`
            : ''}
        <div class="made-c-tooltip__content" part="content">
          ${this.content}
        </div>
      </div>
    `;
    }
}
MadeTooltip.styles = [componentStyles, styles];
MadeTooltip._idCounter = 0;
export default MadeTooltip;
__decorate([
    query('#tooltip-popover')
], MadeTooltip.prototype, "tooltipPopover", void 0);
__decorate([
    query('slot')
], MadeTooltip.prototype, "slotEl", void 0);
__decorate([
    property({ type: String })
], MadeTooltip.prototype, "header", void 0);
__decorate([
    property({ type: String })
], MadeTooltip.prototype, "content", void 0);
__decorate([
    property({ type: String })
], MadeTooltip.prototype, "position", void 0);
__decorate([
    property({ type: Number, attribute: 'delay-duration' })
], MadeTooltip.prototype, "delayDuration", void 0);
__decorate([
    property({ type: Boolean })
], MadeTooltip.prototype, "pointer", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTooltip.prototype, "open", void 0);
__decorate([
    property({ type: String })
], MadeTooltip.prototype, "trigger", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTooltip.prototype, "disabled", void 0);
__decorate([
    property({ type: String })
], MadeTooltip.prototype, "appearance", void 0);
