// @mc-made/web-components/dist/components/tooltip/src/tooltip.styles.js
import { css } from 'lit';
export default css `
  :host {
    --max-width: 18rem;

    display: contents;
    position: relative;
  }

  .made-c-tooltip {
    --tether-size: 6px;

    position-anchor: --anchor;
    position: fixed;
    position-area: top;
    margin: 0;
    border: 0px solid;
    inset: auto;
    background-color: var(--made-color-background-surface-inverse);
    padding: var(--made-size-space-2-x) var(--made-size-space-3-x);
    border-radius: var(--made-component-tooltip-border-radius);
    box-shadow: var(--made-component-tooltip-box-shadow);
    color: var(--made-color-text-on-color);
    max-width: var(--max-width);
    font-size: var(--made-font-size-body-small);
    font-family: var(--made-font-family-primary);
    line-height: var(--made-line-height-rem-02);
    letter-spacing: var(--made-letter-spacing-default);
    font-weight: var(--made-font-weight-regular);
    pointer-events: none;

    /* need this to see the tether */
    overflow: visible;
  }

  .made-c-tooltip.made-c-tooltip--inverted {
    background-color: var(--made-color-background-surface-default);
    color: var(--made-color-text-primary);
  }

  .made-c-tooltip__header {
    font-weight: var(--made-font-weight-bold);
  }

  .made-c-tooltip:popover-open {
    animation: fade-in var(--made-motion-duration-short)
      var(--made-motion-easing-emphasised-ease-out);
  }

  /* Hover safe area */
  .made-c-tooltip::before {
    content: '';
    position: absolute;
    inset: calc(-1 * var(--made-size-space-2-x));
    background: transparent;
    pointer-events: auto;
  }

  .made-c-tooltip.made-c-tooltip--has-pointer::after {
    content: '';
    position: absolute;
    background: inherit;
    pointer-events: none;
  }

  /* Top beak (16×9, points DOWN toward anchor) — base at y=2 overlaps tooltip by 2px */
  .made-c-tooltip--top.made-c-tooltip--has-pointer::after {
    width: 16px;
    height: 9px;
    left: 50%;
    bottom: calc(-1 * var(--tether-size));
    transform: translateX(-50%);
    clip-path: path(
      'M6.43826 8.249396C7.23891 9.250209 8.76109 9.25021 9.56174 8.249397L16 0.20157H0L6.43826 8.249396Z'
    );
  }

  /* Bottom beak (16×9, points UP toward anchor) — base at y=7 overlaps tooltip by 2px */
  .made-c-tooltip--bottom.made-c-tooltip--has-pointer::after {
    width: 16px;
    height: 9px;
    left: 50%;
    top: calc(-1 * var(--tether-size));
    transform: translateX(-50%);
    clip-path: path(
      'M6.43826 0.750604C7.23891 -0.250209 8.76109 -0.25021 9.56174 0.750603L16 8.79843H0L6.43826 0.750604Z'
    );
  }

  /* Left beak (9×16, points RIGHT toward anchor) — base at x=2 overlaps tooltip by 2px */
  .made-c-tooltip--left.made-c-tooltip--has-pointer::after {
    width: 9px;
    height: 16px;
    top: 50%;
    right: calc(-1 * var(--tether-size));
    transform: translateY(-50%);
    clip-path: path(
      'M8.249396 6.43826C9.250209 7.23891 9.25021 8.76109 8.249397 9.56174L0.20157 16L0.20157 0L8.249396 6.43826Z'
    );
  }

  /* Right beak (9×16, points LEFT toward anchor) — base at x=7 overlaps tooltip by 2px */
  .made-c-tooltip--right.made-c-tooltip--has-pointer::after {
    width: 9px;
    height: 16px;
    top: 50%;
    left: calc(-1 * var(--tether-size));
    transform: translateY(-50%);
    clip-path: path(
      'M0.750604 9.56174C-0.250209 8.76109 -0.25021 7.23891 0.750603 6.43826L8.79843 0L8.79843 16L0.750604 9.56174Z'
    );
  }

  @keyframes fade-in {
    0% {
      opacity: 0;
      transform: translateY(var(--move-y)) translateX(var(--move-x));
    }

    100% {
      opacity: 1;
      transform: translateY(0) translateX(0);
    }
  }

  .made-c-tooltip--top {
    --move-x: 0;
    --move-y: 0.25em;
    position-area: top;
    margin: 0 0 var(--made-size-space-2-x) 0;
    position-try: --try-bottom;
  }

  .made-c-tooltip--bottom {
    --move-x: 0;
    --move-y: -0.25em;
    position-area: bottom;
    margin: var(--made-size-space-2-x) 0 0 0;
    position-try: --try-top;
  }

  .made-c-tooltip--left {
    --move-x: 0.25em;
    --move-y: 0;
    position-area: left;
    margin: 0 var(--made-size-space-2-x) 0 0;
    position-try: --try-right;
  }

  .made-c-tooltip--right {
    --move-x: -0.25em;
    --move-y: 0;
    position-area: right;
    margin: 0 0 0 var(--made-size-space-2-x);
    position-try: --try-left;
  }

  @position-try --try-top {
    position-area: top;
    margin: 0 0 var(--made-size-space-2-x) 0;
  }

  @position-try --try-bottom {
    position-area: bottom;
    margin: var(--made-size-space-2-x) 0 0 0;
  }

  @position-try --try-left {
    position-area: left;
    margin: 0 var(--made-size-space-2-x) 0 0;
  }

  @position-try --try-right {
    position-area: right;
    margin: 0 0 0 var(--made-size-space-2-x);
  }

  ::slotted(*) {
    anchor-name: --anchor;
  }

  /** Fallback positioning for browsers that don't support CSS anchor positioning 
   * Can be removed once support is widespread (see https://caniuse.com/mdn-css_properties_position_anchor)
   * and mature.
   * 
  */

  @supports not (anchor-name: --anchor) {
    /* Positioning is computed in JS (_positionFallback) after showPopover().
       Default off-screen prevents a flash at 0,0 before the first rAF fires. */
    .made-c-tooltip {
      position: fixed;
      top: var(--fallback-top, -9999px);
      left: var(--fallback-left, -9999px);
    }
  }
`;
