// @mc-made/web-components/dist/styles/form-control.styles.js
/* Do not edit directly, this file was auto-generated. */
import { css } from 'lit';
export default css `
  /* Block: made-c-form-control */
  :host {
    display: block;
  }

  /* Form container wrapper */
  .made-c-form-control {
    position: relative;
    border: none;
    padding: 0;
    margin: 0;
  }

  .made-c-form-control__label,
  .made-c-form-control__supporting-text {
    display: none;
  }

  /* Modifier: made-c-form-control--has-label */
  .made-c-form-control--has-label .made-c-form-control__label {
    display: flex;
    align-items: center;
    font-family: var(--made-font-family-primary);
    color: var(--made-color-text-primary);
    font-weight: var(--made-font-weight-regular);
    margin-bottom: var(--made-size-space-1-x);
  }

  /* Modifier: made-c-form-control--has-supporting-text */
  .made-c-form-control--has-supporting-text
    .made-c-form-control__supporting-text {
    display: flex;
    font-family: var(--made-font-family-primary);
    align-items: center;
    gap: var(--made-size-space-1-x);
    color: var(--made-color-text-secondary);
    margin-top: var(--made-size-space-1-x);
    font-size: var(--made-font-size-body-small);
    font-weight: var(--made-font-weight-regular);
  }

  /* Font sizes by size modifier */
  .made-c-form-control--medium .made-c-form-control__label {
    --font-size: var(--made-font-size-body-small);
    font-size: var(--font-size);
    line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
  }

  .made-c-form-control--large .made-c-form-control__label {
    --font-size: var(--made-font-size-body-default);
    font-size: var(--font-size);
    line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
  }

  /* Required indicator for label */
  :host([optional]:not([required]))
    .made-c-form-control--has-label
    .made-c-form-control__label::after {
    content: '\\00a0(optional)';
    color: var(--made-color-text-secondary);
  }

  :host([required][required-indicator='required']:not([no-required-indicator]))
    .made-c-form-control--has-label
    .made-c-form-control__label::after {
    content: '\\00a0(required)';
    color: var(--made-color-text-secondary);
  }

  :host([required][required-indicator='*']:not([no-required-indicator]))
    .made-c-form-control--has-label
    .made-c-form-control__label::after {
    content: '\\00a0*';
    color: var(--made-color-text-secondary);
  }

  /* States for supporting text */
  .made-c-form-control--error .made-c-form-control__supporting-text {
    color: var(--made-color-text-danger);
  }

  .made-c-form-control--success .made-c-form-control__supporting-text {
    color: var(--made-color-text-success);
  }

  /* Layout for container to change to vertical or horizontal (Radio Group) */
  .made-c-form-control-input {
    display: flex;
  }

  .made-c-form-control--medium .made-c-form-control-input {
    gap: var(--made-size-space-3-x);
  }

  .made-c-form-control--large .made-c-form-control-input {
    gap: var(--made-size-space-4-x);
  }

  .made-c-form-control-input--column {
    flex-direction: column;
  }

  .made-c-form-control-input--row {
    flex-direction: row;
  }
`;
