// @mc-made/web-components/dist/components/tab/src/tab.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './tab.styles.js';
let id = 0;
/**
 * @summary A Tab is used inside a Tabs component.
 * A Tab represents a single selectable item that displays a corresponding Tab Panel.
 *
 * @documentation '../stories/tab.mdx'
 *
 * @slot - The tab's label.
 * @slot start - Used to place content before tab label. Typically used for icons.
 * @slot end - Used to place content after tab label. Typically used for icons.
 *
 * @csspart base - The component's base wrapper.
 * @csspart start - The container that wraps the start slot.
 * @csspart end - The container that wraps the end slot.
 *
 */
class MadeTab extends MadeElement {
    constructor() {
        super(...arguments);
        this.tabId = ++id;
        this.elementId = `made-tab-${this.tabId}`;
        this.hasSlotController = new HasSlotController(this, '[default]', 'start', 'end');
        /**
         * The name of the tab panel this tab is associated with.
         * The panel must be located in the same tab group.
         */
        this.panel = '';
        /**
         * @internal
         * Draws the tab in an active state.
         */
        this.active = false;
        /** Disables the tab and prevents selection. */
        this.disabled = false;
        /**
         * @internal
         * Sets the tab appearance.
         */
        this.appearance = 'default';
        /**
         * @internal
         * Renders the tab at a reduced size.
         */
        this.compact = false;
        /**
         * @internal
         * Sets the tab placement.
         */
        this.placement = 'top';
        /**
         * @internal
         * Sets the tabIndex.
         */
        this.tabIndex = 0;
    }
    connectedCallback() {
        // Automatically place in made-tabs nav slot
        this.slot ||= 'nav';
        super.connectedCallback();
        this.setAttribute('role', 'tab');
    }
    updated(changedProperties) {
        if (!this.dataset.clone) {
            if (changedProperties.has('active')) {
                this.setAttribute('aria-selected', this.active ? 'true' : 'false');
            }
            if (changedProperties.has('disabled')) {
                this.setAttribute('aria-disabled', this.disabled ? 'true' : 'false');
                if (this.disabled && !this.active) {
                    this.tabIndex = -1;
                }
                else {
                    this.tabIndex = 0;
                }
            }
        }
    }
    render() {
        // Set ID if not provided
        this.id = this.id?.length > 0 ? this.id : this.elementId;
        const hasStartSlot = this.hasSlotController.test('start');
        const hasEndSlot = this.hasSlotController.test('end');
        return html `
      <div
        part="base"
        class=${classMap({
            'made-c-tab': true,
            'made-c-tab--active': this.active,
            'made-c-tab--default': this.appearance === 'default',
            'made-c-tab--minimal': this.appearance === 'minimal',
            'made-c-tab--contained': this.appearance === 'contained',
            'made-c-tab--compact': this.compact,
            [`made-c-tab--placement-${this.placement}`]: true,
            'made-c-tab--disabled': this.disabled
        })}
      >
        <slot
          name="start"
          part="start"
          class=${classMap({
            'made-c-tab__start': hasStartSlot
        })}
        ></slot>
        <slot></slot>
        <slot
          name="end"
          part="end"
          class=${classMap({
            'made-c-tab__end': hasEndSlot
        })}
        ></slot>
      </div>
    `;
    }
}
MadeTab.styles = [componentStyles, styles];
export default MadeTab;
__decorate([
    query('.tab')
], MadeTab.prototype, "tab", void 0);
__decorate([
    property({ reflect: true })
], MadeTab.prototype, "panel", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTab.prototype, "active", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTab.prototype, "disabled", void 0);
__decorate([
    property({ type: String, reflect: true })
], MadeTab.prototype, "appearance", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeTab.prototype, "compact", void 0);
__decorate([
    property({ type: String, reflect: true })
], MadeTab.prototype, "placement", void 0);
__decorate([
    property({ type: Number, reflect: true })
], MadeTab.prototype, "tabIndex", void 0);
