// @mc-made/web-components/dist/components/tab/src/tab.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: inline-block;
  }

  .made-c-tab {
    position: relative;
    display: inline-flex;
    align-items: center;
    width: 100%;
    font-family: var(--made-font-family-primary);
    color: inherit;
    font-size: var(--made-component-tabs-default-font-size);
    font-weight: var(--made-component-tabs-default-font-weight);
    line-height: var(--made-component-tabs-default-line-height);
    letter-spacing: var(--made-component-tabs-default-letter-spacing);
    padding-block: var(--made-component-tabs-default-item-padding-y);
    padding-inline: var(--made-component-tabs-default-item-padding-x);
    border-color: var(--made-component-tabs-default-border-color);
    background-color: transparent;
    border-bottom: none;
    white-space: nowrap;
    user-select: none;
    -webkit-user-select: none;
    cursor: pointer;
    -webkit-tap-highlight-color: rgba(0, 0, 0, 0);

    ::slotted(*) {
      line-height: inherit;
    }

    .made-c-tab__start,
    .made-c-tab__end {
      display: inline-flex;
      align-items: center;
      justify-content: center;

      &::slotted(made-icon) {
        --width: var(--made-size-space-5-x);
        --height: var(--made-size-space-5-x);
      }
    }

    .made-c-tab__start {
      margin-inline-end: var(--made-size-space-1-x);
    }

    .made-c-tab__end {
      margin-inline-start: var(--made-size-space-1-x);
    }

    &.made-c-tab--placement-top,
    &.made-c-tab--placement-bottom {
      justify-content: center;
    }

    &.made-c-tab--minimal,
    &.made-c-tab--contained {
      font-size: var(--made-component-tabs-contained-font-size);
      font-weight: var(--made-component-tabs-contained-font-weight);
      line-height: var(--made-component-tabs-contained-line-height);
      letter-spacing: var(--made-component-tabs-contained-letter-spacing);
      border-radius: var(--made-component-tabs-contained-tab-border-radius);
      padding-block: var(--made-component-tabs-contained-item-padding-y);
      padding-inline: var(--made-component-tabs-contained-item-padding-x);
      box-shadow: var(--made-component-tabs-contained-item-box-shadow);

      &.made-c-tab--placement-start,
      &.made-c-tab--placement-end {
        border-radius: 10px;
      }
    }

    &.made-c-tab--compact {
      font-size: var(--made-font-size-02, 0.75rem);
      font-weight: var(--made-font-weight-medium, 500);
      line-height: var(--made-line-height-rem-02, 1rem);
      letter-spacing: var(--made-component-tabs-contained-letter-spacing);
      padding-block: var(--made-size-space-1point5-x, 0.375rem);
      padding-inline: var(--made-size-space-2-x, 0.5rem);
    }
  }

  .made-c-tab:hover:not(.made-c-tab--active, .made-c-tab--disabled) {
    color: var(--made-color-text-primary);
  }

  :host(:focus) {
    outline: transparent;
  }

  :focus-visible {
    outline: transparent;
  }

  :host(:focus-visible) .made-c-tab {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: -5px;

    &.made-c-tab--minimal,
    &.made-c-tab--contained {
      outline-offset: 2px;
    }
  }

  .made-c-tab--active:not(.made-c-tab--minimal) {
    color: var(--text-active-color);
  }

  .made-c-tab.made-c-tab--disabled {
    color: var(--made-color-text-disabled);
    cursor: not-allowed;
  }
`;
