// @mc-made/web-components/dist/internal/face/validators/select-validator.js
/**
 * - Adapted from CheckboxValidator to support native <select> constraint validation.
 */
/* istanbul ignore file */
import { Validator } from './validator.js';
/**
 * A validator that provides constraint validation that emulates
 * `<select required>` validation.
 */
export class SelectValidator extends Validator {
    computeValidity(state) {
        if (!this.selectControl) {
            // Lazily create the platform select with two options:
            // index 0 = empty option (for required validation),
            // index 1 = non-empty option (for a selected value).
            this.selectControl = document.createElement('select');
            const emptyOption = document.createElement('option');
            emptyOption.value = '';
            this.selectControl.appendChild(emptyOption);
            const valueOption = document.createElement('option');
            this.selectControl.appendChild(valueOption);
        }
        this.selectControl.required = state.required;
        if (state.value === '') {
            this.selectControl.options[0].selected = true;
            this.selectControl.options[1].selected = false;
        }
        else {
            this.selectControl.options[1].value = state.value;
            this.selectControl.options[0].selected = false;
            this.selectControl.options[1].selected = true;
        }
        return {
            validity: this.selectControl.validity,
            validationMessage: this.selectControl.validationMessage
        };
    }
    equals(prev, next) {
        return prev.value === next.value && prev.required === next.required;
    }
    copy({ value, required }) {
        return { value, required };
    }
}
