// @mc-made/web-components/dist/components/modal/src/modal.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: none;
    --toggle-duration: var(--made-motion-duration-longest);
  }

  :host([open]) {
    display: block;
  }

  .made-c-modal {
    display: flex;
    flex-direction: column;
    top: 0;
    right: 0;
    bottom: 0;
    left: 0;
    color: var(--made-color-text-primary);
    font-family: var(--made-font-family-primary);
    max-width: calc(100% - var(--made-size-space-8-x));
    max-height: calc(100% - var(--made-size-space-8-x));
    background-color: var(--made-color-background-surface-default);
    border-radius: var(--made-component-modal-border-radius);
    border-color: var(--made-component-modal-border-color);
    border-style: solid;
    border-width: 0;
    box-shadow: var(--made-component-modal-box-shadow);
    padding: 0;
    margin: auto;

    &.made-c-modal--show {
      animation: show-modal var(--toggle-duration) cubic-bezier(0, 0, 0.58, 1);

      &::backdrop {
        animation: show-backdrop var(--toggle-duration)
          cubic-bezier(0, 0, 0.58, 1);
      }
    }

    &.made-c-modal--hide {
      animation: hide-modal var(--toggle-duration) cubic-bezier(0, 0, 0.58, 1);

      &::backdrop {
        animation: hide-backdrop var(--toggle-duration)
          cubic-bezier(0, 0, 0.58, 1);
      }
    }
  }

  .made-c-modal:focus {
    outline: none;
  }

  made-icon-button.made-c-modal__close-button::part(base) {
    border-radius: var(--made-border-radius-04);
    height: var(--made-size-space-6-x);
    width: var(--made-size-space-6-x);
  }

  .made-c-modal__media-header {
    position: relative;

    & ~ made-icon-button.made-c-modal__close-button {
      position: absolute;
      top: var(--made-size-space-6-x);
      right: var(--made-size-space-6-x);

      &::part(base) {
        background-color: rgba(255, 255, 255, 0.5);

        &:hover {
          background-color: var(--made-color-background-surface-hover);
        }
      }
    }
  }

  .made-c-modal__header {
    flex: 0 0 auto;
    display: flex;
    flex-wrap: nowrap;
    align-items: center;
    padding-inline-start: var(--made-component-modal-top-bar-padding-left);
    padding-inline-end: var(--made-component-modal-top-bar-padding-right);
    padding-block-start: var(--made-component-modal-top-bar-padding-top);
    padding-block-end: var(--made-component-modal-top-bar-padding-bottom);
  }

  .made-c-modal__title {
    flex: 1 1 auto;
    font-family: inherit;
    --font-size: var(--made-component-modal-title-font-size);
    font-size: var(--font-size);
    font-weight: var(--made-component-modal-title-font-weight);
    line-height: calc(var(--font-size) * 1.45);
    line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
    margin: 0;
  }

  .made-c-modal__body {
    padding: var(--made-size-space-6-x);
    overflow: auto;
    -webkit-overflow-scrolling: touch;

    & ::slotted(*) {
      --font-size: var(--made-component-modal-content-font-size);
      font-size: var(--font-size);
      line-height: calc(var(--font-size) * 1.45);
      line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));
      font-weight: var(--made-component-modal-content-font-weight);
    }
  }

  .made-c-modal__footer {
    flex: 0 0 auto;
    display: flex;
    flex-wrap: wrap;
    gap: var(--made-component-modal-footer-gap);
    justify-content: end;
    padding-block: var(--made-component-modal-footer-padding-y);
    padding-inline: var(--made-component-modal-footer-padding-x);
  }

  .made-c-modal::backdrop {
    background-color: color-mix(
      in srgb,
      var(--made-color-background-backdrop) 40%,
      transparent
    );
  }

  @keyframes show-modal {
    from {
      opacity: 0;
      transform: translateY(var(--made-size-space-4-x));
    }
    to {
      opacity: 1;
      transform: translateY(0);
    }
  }
  @keyframes hide-modal {
    from {
      opacity: 1;
      transform: translateY(0);
    }
    to {
      opacity: 0;
      transform: translateY(var(--made-size-space-4-x));
    }
    to {
    }
  }

  @keyframes show-backdrop {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }
  @keyframes hide-backdrop {
    from {
      opacity: 1;
    }
    to {
      opacity: 0;
    }
  }

  @media screen and (min-width: 640px) {
    .made-c-modal--x-small {
      width: 480px;
    }
    .made-c-modal--small {
      width: 640px;
    }
    .made-c-modal--medium {
      width: 800px;
    }
    .made-c-modal--large {
      width: calc(100% - var(--made-size-space-20-x));
    }
  }
`;
