// @mc-made/web-components/dist/components/modal/src/modal.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeDivider from '../../divider/src/divider.component.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIconButton from '../../icon-button/src/icon-button.component.js';
import styles from './modal.styles.js';
/**
 * @summary Modal’s focus the user’s attention exclusively on one task or piece of
 * information via a window that sits on top of the page content.
 *
 * @documentation '../stories/modal.mdx'
 *
 * @dependency made-icon-button
 * @dependency made-divider
 *
 * @fires close - Emitted when the modal is requested to close.
 *
 * @slot - The default slot - Slot the modal's main content.
 * @slot label - Slot for the modal's header label. Alternatively, you can use the `label` attribute.
 * @slot media-header - Slot for media content in the header.
 * @slot footer - Slot for the modal's footer content, typically action buttons.
 *
 * @csspart body - The modal's body section.
 * @csspart modal - The component's base wrapper.
 * @csspart header - The modal's header section.
 * @csspart title - The modal's title element.
 * @csspart close-button - The modal's close button.
 *
 * @cssproperty --toggle-duration - The animation duration when opening and closing the modal.
 */
class MadeModal extends MadeElement {
    constructor() {
        super(...arguments);
        this.hasSlotController = new HasSlotController(this, '[default]', 'label', 'media-header');
        /** Indicates whether or not the modal is open. Toggle this attribute to show and hide the modal. */
        this.open = false;
        /** Controls the visibility of the dismiss button. */
        this.noDismissButton = false;
        /**
         * The modal's label as displayed in the header. You should always include a relevant label, as it is required for
         * proper accessibility. If you need to display HTML, use the `label` slot instead.
         */
        this.label = '';
        /** The size of the modal. */
        this.size = 'small';
        /** When true, the header divider is not displayed. */
        this.noHeaderDivider = false;
        /** When true, the footer divider is not displayed. */
        this.noFooterDivider = false;
        /** When true, clicking on the backdrop will close the modal. */
        this.noDismiss = false;
        this.handleCancel = (event) => {
            event.preventDefault();
            this.close();
        };
    }
    firstUpdated() {
        if (this.open) {
            this.modal.showModal();
        }
    }
    updated(changedProperties) {
        if (changedProperties.has('open')) {
            if (this.open && !this.modal.open) {
                this.show();
            }
            else if (!this.open && this.modal.open) {
                this.open = true;
                this.close();
            }
        }
    }
    show() {
        this.modal.showModal();
        this.modal.classList.add('made-c-modal--show');
        const onEnd = () => {
            this.modal.classList.remove('made-c-modal--show');
        };
        this.modal.focus();
        this.modal.addEventListener('animationend', onEnd, { once: true });
    }
    close() {
        this.dispatchEvent(new Event('close', { bubbles: true, composed: true, cancelable: true }));
        this.modal.classList.add('made-c-modal--hide');
        const onEnd = () => {
            this.modal.classList.remove('made-c-modal--hide');
            this.open = false;
            this.modal.close();
        };
        this.modal.addEventListener('animationend', onEnd, { once: true });
    }
    handlePointerdown(event) {
        if (event.target === this.modal && !this.noDismiss) {
            this.close();
        }
    }
    handleDialogClick(event) {
        const target = event.target;
        const button = target.closest('[data-dialog="close"]');
        // Close when a button with [data-dialog="close"] is clicked
        if (button) {
            event.stopPropagation();
            this.close();
        }
    }
    getTitle() {
        return html `
      <h2 id="modal-title" part="title" class="made-c-modal__title">
        <slot name="label"> ${this.label} </slot>
      </h2>
    `;
    }
    render() {
        const mediaHeader = this.hasSlotController.test('media-header');
        const label = this.label || this.hasSlotController.test('label');
        const noHeader = !mediaHeader && !label;
        const showDismiss = !this.noDismissButton && (label || mediaHeader);
        const showFooter = this.hasSlotController.test('footer');
        const showBody = this.hasSlotController.test('[default]');
        return html `
      <dialog
        part="modal"
        class=${classMap({
            'made-c-modal': true,
            'made-c-modal--x-small': this.size === 'x-small',
            'made-c-modal--small': this.size === 'small',
            'made-c-modal--medium': this.size === 'medium',
            'made-c-modal--large': this.size === 'large'
        })}
        @cancel=${this.handleCancel}
        @pointerdown=${this.handlePointerdown}
        @click=${this.handleDialogClick}
      >
        ${!noHeader
            ? html `
              <header
                part="header"
                class=${mediaHeader ? '' : 'made-c-modal__header'}
              >
                ${mediaHeader
                ? html `
                      <div class="made-c-modal__media-header">
                        <slot name="media-header"></slot>
                      </div>
                    `
                : label
                    ? this.getTitle()
                    : ''}
                ${showDismiss
                ? html `
                      <made-icon-button
                        variant="ghost"
                        part="close-button"
                        size="small"
                        class="made-c-modal__close-button"
                        icon-name="x-mark-outline"
                        library="system"
                        aria-label=${this.localize.term('close')}
                        @click=${this.close}
                      ></made-icon-button>
                    `
                : ''}
              </header>
              ${!this.noHeaderDivider
                ? html `<made-divider></made-divider>`
                : ''}
            `
            : ''}
        ${showBody
            ? html `
              <div part="body" class="made-c-modal__body">
                ${mediaHeader && label ? this.getTitle() : ''}
                <slot></slot>
              </div>
            `
            : ''}
        ${showFooter
            ? html `
              ${!this.noFooterDivider
                ? html `<made-divider></made-divider>`
                : ''}
              <footer part="footer" class="made-c-modal__footer">
                <slot name="footer"></slot>
              </footer>
            `
            : ''}
      </dialog>
    `;
    }
}
MadeModal.styles = [componentStyles, styles];
MadeModal.dependencies = {
    'made-icon-button': MadeIconButton,
    'made-divider': MadeDivider
};
export default MadeModal;
__decorate([
    query('.made-c-modal')
], MadeModal.prototype, "modal", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeModal.prototype, "open", void 0);
__decorate([
    property({ attribute: 'no-dismiss-button', type: Boolean, reflect: true })
], MadeModal.prototype, "noDismissButton", void 0);
__decorate([
    property({ reflect: true })
], MadeModal.prototype, "label", void 0);
__decorate([
    property()
], MadeModal.prototype, "size", void 0);
__decorate([
    property({ type: Boolean, attribute: 'no-header-divider' })
], MadeModal.prototype, "noHeaderDivider", void 0);
__decorate([
    property({ type: Boolean, attribute: 'no-footer-divider' })
], MadeModal.prototype, "noFooterDivider", void 0);
__decorate([
    property({ type: Boolean, attribute: 'no-dismiss' })
], MadeModal.prototype, "noDismiss", void 0);
