// @mc-made/web-components/dist/components/radio-group/src/radio-group.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html, nothing } from 'lit';
import { property, queryAssignedElements, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/src/icon.component.js';
import styles from './radio-group.styles.js';
/**
 * @summary Radio groups are used to group multiple radio buttons together so they function as a single form control.
 * @documentation '../stories/radio-group.mdx'
 *
 * @slot - The default slot.
 * @slot supporting-text - Text that describes the radio group. Alternatively, you can use the `supporting-text` attribute.
 *
 * @csspart container - The component's base wrapper.
 * @csspart label - The component's label.
 * @csspart layout-container - The default slot wrapper.
 * @csspart supporting-text - The supporting text's wrapper.
 *
 */
class MadeRadioGroup extends MadeElement {
    get hasError() {
        return this.error || this.hasNativeError;
    }
    constructor() {
        super();
        /**
         * Whether or not a native error has been reported via `reportValidity()`.
         */
        this.hasNativeError = false;
        /**
         * The validation message displayed from a native error via
         * `reportValidity()`.
         */
        this.nativeErrorText = '';
        /** Name applied to each radio. Each group should have it's own unique name. */
        this.name = 'radio-group';
        /** The radio group's supporting text.  */
        this.supportingText = '';
        /** Size applied to each radio. */
        this.size = 'medium';
        /** Ensure a radio is selected before allowing the form to submit. */
        this.required = false;
        /**
         * Gets or sets whether or not the text field is in a visually invalid state.
         *
         * This error state overrides the error state controlled by
         * `reportValidity()`.
         */
        this.error = false;
        /** Removes the asterisk from the label when required. */
        this.noAsterisk = false;
        /** Set to 'branded' for alternative brand appearance. */
        this.appearance = 'neutral';
        /** Changes the layout of the radio's. */
        this.direction = 'column';
        this.addEventListener('change', this.handleChange);
        this.addEventListener('mde-invalid', this.handleInvalid);
        this.addEventListener('mde-reset', this.handleReset);
    }
    firstUpdated() {
        this.radios.forEach(radio => {
            if (radio.value === this.value) {
                radio.setAttribute('checked', '');
            }
        });
    }
    updated() {
        this.syncRadios();
    }
    async syncRadios() {
        if (this.radios.length) {
            await Promise.all(
            // Sync properties
            this.radios.map(async (radio) => {
                await radio.updateComplete;
                radio.name = this.name;
                radio.size = this.size;
                radio.required = this.required;
                radio.error = this.hasError;
                radio.appearance = this.appearance;
            }));
        }
    }
    handleSlotChange() {
        if (window.customElements.get('made-radio')) {
            this.syncRadios();
        }
        else {
            window.customElements
                .whenDefined('made-radio')
                .then(() => this.handleSlotChange());
        }
    }
    handleChange(event) {
        this.value = event.target.value;
        this.hasNativeError = false;
    }
    handleInvalid(event) {
        event.stopPropagation();
        if (!this.hasNativeError) {
            this.hasNativeError = true;
            this.nativeErrorText = this.radios[0].validationMessage;
        }
    }
    handleReset() {
        this.hasNativeError = false;
        this.value = '';
    }
    getErrorText() {
        return this.hasError
            ? this.error
                ? this.errorText
                : this.nativeErrorText
            : null;
    }
    get supportingOrValidationText() {
        return this.getErrorText() || this.supportingText;
    }
    renderSupportingText() {
        if (!this.supportingOrValidationText) {
            return null;
        }
        return html `
      ${this.hasError
            ? html `<span>
            <made-icon
              name="exclamation-triangle-solid-12px"
              library="system"
            ></made-icon>
          </span>`
            : nothing}
      <span>${this.supportingOrValidationText}</span>
    `;
    }
    render() {
        const hasLabel = !!this.label;
        const supportingText = this.renderSupportingText();
        const role = this.hasError ? 'alert' : nothing;
        return html `
      <fieldset
        part="container"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control-radio-group': true,
            'made-c-form-control--medium': this.size === 'medium',
            'made-c-form-control--large': this.size === 'large',
            'made-c-form-control--has-label': hasLabel,
            'made-c-form-control--has-supporting-text': !!supportingText,
            'made-c-form-control--error': this.hasError
        })}
        role="radiogroup"
        aria-labelledby="label"
        aria-describedby="supporting-text"
      >
        <label
          part="label"
          id="label"
          class="made-c-form-control__label"
          aria-hidden=${hasLabel ? 'false' : 'true'}
        >
          <slot name="label">${this.label}</slot>
        </label>
        <div
          part="layout-container"
          class=${classMap({
            'made-c-form-control-input': true,
            'made-c-form-control-input--column': this.direction === 'column',
            'made-c-form-control-input--row': this.direction === 'row'
        })}
        >
          <slot @slotchange=${this.handleSlotChange}></slot>
        </div>

        <div
          part="supporting-text"
          id="supporting-text"
          class="made-c-form-control__supporting-text"
          aria-hidden=${supportingText ? 'false' : 'true'}
          role=${role}
        >
          <slot name="supporting-text">${supportingText}</slot>
        </div>
      </fieldset>
    `;
    }
}
MadeRadioGroup.styles = [componentStyles, formControlStyles, styles];
MadeRadioGroup.dependencies = { 'made-icon': MadeIcon };
export default MadeRadioGroup;
__decorate([
    queryAssignedElements()
], MadeRadioGroup.prototype, "radios", void 0);
__decorate([
    state()
], MadeRadioGroup.prototype, "hasNativeError", void 0);
__decorate([
    state()
], MadeRadioGroup.prototype, "nativeErrorText", void 0);
__decorate([
    property()
], MadeRadioGroup.prototype, "label", void 0);
__decorate([
    property()
], MadeRadioGroup.prototype, "name", void 0);
__decorate([
    property()
], MadeRadioGroup.prototype, "value", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeRadioGroup.prototype, "supportingText", void 0);
__decorate([
    property()
], MadeRadioGroup.prototype, "size", void 0);
__decorate([
    property({ type: Boolean })
], MadeRadioGroup.prototype, "required", void 0);
__decorate([
    property({ type: Boolean })
], MadeRadioGroup.prototype, "error", void 0);
__decorate([
    property({ attribute: 'error-text' })
], MadeRadioGroup.prototype, "errorText", void 0);
__decorate([
    property({ type: Boolean })
], MadeRadioGroup.prototype, "noAsterisk", void 0);
__decorate([
    property()
], MadeRadioGroup.prototype, "appearance", void 0);
__decorate([
    property()
], MadeRadioGroup.prototype, "direction", void 0);
