// @mc-made/web-components/dist/internal/made-element.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement } from 'lit';
import { LocalizeController } from '../utilities/localize.js';
import { property } from 'lit/decorators.js';
/**
 * @summary The base class for all Made web components.
 *
 * @description
 * MadeElement extends LitElement to provide common functionality across all Made components:
 * - Automatic component registration to prevent duplicate definitions
 * - Reactive internationalization properties (dir, lang)
 * - Consistent event emission patterns
 * - Made-specific lifecycle patterns
 *
 * All Made components should extend this class instead of LitElement directly.
 *
 * @example
 * ```typescript
 * export default class MyComponent extends MadeElement {
 *   static styles = [componentStyles, myStyles];
 *
 *   render() {
 *     return html`<div>My component</div>`;
 *   }
 * }
 * ```
 *
 * @since 0.1.0
 */
class MadeElement extends LitElement {
    /**
     * Emits a custom event with consistent configuration.
     *
     * All Made component events use this method to ensure consistent behavior:
     * - Events bubble up through the DOM
     * - Events are composed (cross shadow DOM boundaries)
     * - Events are not cancelable by default
     *
     * @param name - The name of the event to emit
     * @param options - Additional event options to merge with defaults
     * @returns The dispatched CustomEvent
     *
     * @example
     * ```typescript
     * // Basic event
     * this.emit('mde-change');
     *
     * // Event with data
     * this.emit('mde-selection-change', {
     *   detail: { selectedItems: this.selectedItems }
     * });
     *
     * // Cancelable event
     * this.emit('mde-before-close', {
     *   cancelable: true,
     *   detail: { reason: 'user-action' }
     * });
     * ```
     */
    emit(name, options) {
        const event = new CustomEvent(name, {
            bubbles: true,
            cancelable: false,
            composed: true,
            detail: {},
            ...options
        });
        this.dispatchEvent(event);
        return event;
    }
    /**
     * Safely registers a custom element, preventing duplicate registration errors.
     *
     * This method checks if an element is already registered before attempting
     * to define it. This prevents the "CustomElementRegistry.define: the name
     * is already used" error that occurs when components are imported multiple times.
     *
     * @param tag - The custom element tag name (e.g., 'made-button')
     * @param customElement - The element class to register (defaults to this class)
     *
     * @example
     * ```typescript
     * // Register the component
     * MadeButton.define('made-button');
     *
     * // Register with different class
     * MadeElement.define('custom-element', MyCustomClass);
     * ```
     *
     * @internal This method is primarily used by the component registration system
     */
    static define(tag, customElement = this) {
        const currentlyRegisteredElement = customElements.get(tag);
        if (!currentlyRegisteredElement) {
            customElements.define(tag, customElement);
        }
    }
    /**
     * Automatically registers component dependencies during construction.
     *
     * This constructor iterates through the static dependencies object and
     * ensures all dependent components are registered before the component
     * is fully initialized. This prevents issues where a component tries to
     * render a dependency that hasn't been defined yet.
     *
     * @internal This is called automatically during component instantiation
     */
    constructor() {
        super();
        /**
         * A reactive controller that resolves translated strings for the component's
         * current locale, falling back to English when no matching translation is found.
         *
         * Use `this.localize.term('key')` in your `render()` method to access translated strings.
         *
         * @example
         * ```typescript
         * render() {
         *   return html`<div aria-label=${this.localize.term('loading')}></div>`;
         * }
         * ```
         */
        this.localize = new LocalizeController(this);
        Object.entries(this.constructor.dependencies).forEach(([tag, component]) => {
            this.constructor.define(tag, component);
        });
    }
}
/**
 * Static registry of component dependencies.
 *
 * Components that render other Made components in their shadow DOM should
 * declare them as dependencies. This ensures that dependent components are
 * automatically registered when the parent component is used.
 *
 * @example
 * ```typescript
 * export default class MadeButton extends MadeElement {
 *   static dependencies = {
 *     'made-loader': MadeLoader,
 *     'made-icon': MadeIcon
 *   };
 * }
 * ```
 *
 * When MadeButton is instantiated, MadeLoader and MadeIcon will be
 * automatically registered if they haven't been already.
 */
MadeElement.dependencies = {};
export default MadeElement;
__decorate([
    property()
], MadeElement.prototype, "dir", void 0);
__decorate([
    property()
], MadeElement.prototype, "lang", void 0);
