// @mc-made/web-components/dist/components/divider/src/divider.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { property } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import styles from './divider.styles.js';
/**
 * @summary A divider is a thin line that visually separates or groups elements.
 * @documentation '../stories/divider.mdx'
 *
 * @cssproperty --width - The thickness of the divider.
 * @cssproperty --inset-size - The amount of inset applied to the divider.
 */
class MadeDivider extends MadeElement {
    constructor() {
        super(...arguments);
        /** Set the Orientation of the divider. */
        this.direction = 'horizontal';
        /** Width or thickness of the divider */
        this.width = 'small';
        /** Indents the divider equally on both sides. */
        this.inset = false;
        /** Indents the divider only on the leading side. */
        this.insetStart = false;
        /** Indents the divider only on the trailing side. */
        this.insetEnd = false;
    }
}
MadeDivider.styles = [componentStyles, styles];
export default MadeDivider;
__decorate([
    property({ reflect: true })
], MadeDivider.prototype, "direction", void 0);
__decorate([
    property({ reflect: true })
], MadeDivider.prototype, "width", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeDivider.prototype, "inset", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'inset-start' })
], MadeDivider.prototype, "insetStart", void 0);
__decorate([
    property({ type: Boolean, reflect: true, attribute: 'inset-end' })
], MadeDivider.prototype, "insetEnd", void 0);
