// @mc-made/web-components/dist/components/select/src/select.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: block;
  }

  .made-c-select {
    --font-size: var(--made-font-size-body-default);
    --line-height-fallback: calc(var(--font-size) * 1.45);
    --line-height: round(down, calc(var(--font-size) * 1.45), var(--base-4));

    position: relative;
    display: inline-flex;
    align-items: stretch;
    width: 100%;

    & .made-c-select__chevron made-icon {
      --width: var(--line-height);
      --height: var(--line-height);
    }
    font-family: var(--made-font-family-primary);
    font-size: var(--font-size);
    font-weight: var(--made-font-weight-regular);
    line-height: var(--line-height, var(--line-height-fallback));
    letter-spacing: var(--made-letter-spacing-default);
    color: var(--made-color-text-primary);
    background-color: var(--made-component-form-control-background);
    border: var(--made-component-form-control-border-width) solid
      var(--made-component-form-control-border-color);
    border-radius: var(--made-component-form-control-border-radius);
    vertical-align: middle;
    overflow: hidden;
    cursor: pointer;
    box-shadow: var(--made-component-form-control-box-shadow);
    transition:
      color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease-in),
      border var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease-in),
      box-shadow var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease-in),
      background-color var(--made-motion-duration-short)
        var(--made-motion-easing-standard-ease-in);
  }

  .made-c-select.made-c-select--focused:not(.made-c-select--disabled),
  .made-c-select:has(.made-c-select__control:focus) {
    outline: 2px solid var(--made-color-border-focused);
    outline-offset: 2px;
  }

  .made-c-form-control.made-c-form-control--error .made-c-select {
    border-color: var(--made-color-border-danger-strong);
  }

  .made-c-form-control.made-c-form-control--success:not(
      .made-c-form-control--error
    )
    .made-c-select {
    border-color: var(--made-color-border-success-strong);
  }

  .made-c-select__control {
    flex: 1 1 auto;
    font-family: inherit;
    font-size: inherit;
    font-weight: inherit;
    line-height: inherit;
    box-sizing: content-box;
    height: var(--made-size-space-6-x);
    width: 100%;
    min-width: 0;
    border: none;
    background: transparent;
    box-shadow: none;
    padding: 0;
    margin: 0;
    cursor: inherit;
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    color: inherit;
    border-radius: inherit;
  }

  .made-c-select__control:focus {
    outline: none;
  }

  /* Chevron indicator — absolutely positioned so the <select> fills full container
   * width (native dropdown width) and clicks on the chevron pass through to
   * the <select> element via pointer-events: none. */

  .made-c-select__chevron {
    position: absolute;
    inset-block: 0;
    inset-inline-end: 0;
    display: inline-flex;
    align-items: center;
    pointer-events: none;
    color: var(--made-color-icon-primary);
  }

  /*
   * Size modifiers
   */

  .made-c-select--medium .made-c-select__control {
    padding: var(--made-component-form-control-medium-padding-y)
      var(--made-component-form-control-medium-padding-x);
    /* Right padding makes room for the absolutely-positioned chevron */
    padding-inline-end: calc(
      var(--line-height) + var(--made-component-form-control-medium-padding-x) +
        var(--made-size-space-1-x)
    );
  }

  .made-c-select--medium .made-c-select__chevron {
    padding-inline: var(--made-size-space-0point5-x)
      var(--made-component-form-control-medium-padding-x);
  }

  .made-c-select--large .made-c-select__control {
    padding: var(--made-component-form-control-large-padding-y)
      var(--made-component-form-control-large-padding-x);
    padding-inline-end: calc(
      var(--line-height) + var(--made-component-form-control-large-padding-x) +
        var(--made-size-space-1-x)
    );
  }

  .made-c-select--large .made-c-select__chevron {
    padding-inline: var(--made-size-space-0point5-x)
      var(--made-component-form-control-large-padding-x);
  }

  /* Disabled state */

  .made-c-select.made-c-select--disabled,
  .made-c-select:has(.made-c-select__control:disabled) {
    background-color: var(--made-color-background-disabled);
    border-color: var(--made-color-border-disabled);
    cursor: not-allowed;

    & .made-c-select__chevron {
      color: var(--made-color-icon-disabled);
    }
  }

  .made-c-select.made-c-select--disabled .made-c-select__control,
  .made-c-select__control:disabled {
    color: var(--made-color-text-disabled);
  }

  /* Placeholder */

  .made-c-select__control::-moz-placeholder {
    color: var(--made-color-text-placeholder);
    -moz-user-select: none;
    user-select: none;
  }

  .made-c-select__control::placeholder {
    color: var(--made-color-text-placeholder);
    user-select: none;
  }

  .made-c-select.made-c-select--disabled
    .made-c-select__control::-moz-placeholder {
    color: var(--made-color-text-disabled);
  }

  .made-c-select.made-c-select--disabled .made-c-select__control::placeholder {
    color: var(--made-color-text-disabled);
  }

  /* hr styling */

  .made-c-select hr {
    color: var(--made-color-border-default);
  }
`;
