// @mc-made/web-components/dist/components/select/src/select.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { createValidator, getValidityAnchor, mixinConstraintValidation } from '../../../internal/face/constraint-validation.js';
import { getFormValue, mixinFormAssociated } from '../../../internal/face/form-associated.js';
import { HasSlotController } from '../../../internal/slot.js';
import { html, nothing } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { live } from 'lit/directives/live.js';
import { mixinElementInternals } from '../../../internal/face/element-internals.js';
import { mixinOnReportValidity, onReportValidity } from '../../../internal/face/on-report-validity.js';
import { property, query, state } from 'lit/decorators.js';
import { SelectValidator } from '../../../internal/face/validators/select-validator.js';
import componentStyles from '../../../styles/component.styles.js';
import formControlStyles from '../../../styles/form-control.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/src/icon.component.js';
import styles from './select.styles.js';
/**
 * @summary Selects allow users to choose one option from a dropdown list.
 * @documentation '../stories/select.mdx'
 *
 * @dependency made-icon
 *
 * @slot - Place `<option>`, `<optgroup>`, and `<hr>` elements here. `<hr>` renders
 *   as a visual separator in the dropdown list.
 * @slot label - The select's label. Alternatively, use the `label` attribute.
 * @slot supporting-text - Supporting text displayed below the control. Alternatively,
 *   use the `supporting-text` attribute.
 *
 * @fires change {Event} - Fired when the user commits a new selection.
 * @fires invalid {Event} - Fired when the control has been checked for validity and
 *   does not satisfy its constraints.
 *
 * @method focus(options?: FocusOptions) - Moves focus to the internal `<select>` element.
 * @method blur() - Removes focus from the internal `<select>` element.
 * @method checkValidity() - Returns `true` if the control satisfies its validation constraints.
 * @method reportValidity() - Checks validity and, if invalid, shows a validation message. Returns `true` if valid.
 * @method setCustomValidity(message: string) - Sets a custom validation message. Pass an empty string to clear.
 *
 * @csspart form-control - The outermost wrapper `<div>` that contains the label, control, and supporting text.
 * @csspart label - The `<label>` element.
 * @csspart select-container - The `<div>` that wraps the `<select>` and the chevron icon.
 * @csspart select - The internal `<select>` element.
 * @csspart chevron - The `<span>` wrapping the chevron `<made-icon>`.
 * @csspart supporting-text - The `<div>` that holds the supporting text or validation message.
 */
const selectBaseClass = mixinOnReportValidity(mixinConstraintValidation(mixinFormAssociated(mixinElementInternals(MadeElement))));
class MadeSelect extends selectBaseClass {
    constructor() {
        super(...arguments);
        this.LABEL = 'label';
        this.hasSlotController = new HasSlotController(this, this.LABEL, 'supporting-text');
        this.hasFocus = false;
        this.hasNativeError = false;
        this.nativeErrorText = '';
        /** Gets or sets whether or not the select is in a visually invalid state. */
        this.error = false;
        /** The current value of the select, submitted as a name/value pair with form data. */
        this.value = '';
        /** The select's label. If you need to display HTML, use the label slot instead. */
        this.label = '';
        /** The select's size. */
        this.size = 'medium';
        /** Provides extra details below the select, including instructions on its usage. */
        this.supportingText = '';
        /** Placeholder text shown as the first option when no value is selected. */
        this.placeholder = '';
        /**
         * Makes the select a required field.
         *
         * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/select#required
         */
        this.required = false;
        /** Visual indicators that are appended to the label of a required field. */
        this.requiredIndicator = 'required';
        /** Add the optional visual indicator to a non-required field. */
        this.optional = false;
        /** Hides the required visual indicator on the label, when the select is required. */
        this.noRequiredIndicator = false;
    }
    get hasError() {
        return this.error || this.hasNativeError;
    }
    connectedCallback() {
        super.connectedCallback();
        this._observer = new MutationObserver(() => this._syncOptions());
        this._observer.observe(this, { childList: true, subtree: true });
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this._observer?.disconnect();
    }
    firstUpdated() {
        this._syncOptions();
    }
    updated(changedProperties) {
        if (changedProperties.has('value') && this.select) {
            this.select.value = this.value;
        }
        if (changedProperties.has('placeholder')) {
            this._syncOptions();
        }
    }
    _syncOptions() {
        if (!this.select)
            return;
        const currentValue = this.value || this.select.value;
        // Clear all existing content from shadow select (replaceChildren removes
        // both <option> and <optgroup> elements, unlike remove(index) which only
        // removes from the flat options collection and leaves empty <optgroup> shells)
        this.select.replaceChildren();
        // Add placeholder option if specified
        if (this.placeholder) {
            const placeholderOption = document.createElement('option');
            placeholderOption.value = '';
            placeholderOption.textContent = this.placeholder;
            placeholderOption.disabled = true;
            placeholderOption.hidden = true;
            this.select.appendChild(placeholderOption);
        }
        // Clone options, optgroups, and <hr> separators from light DOM children
        Array.from(this.children).forEach(child => {
            if (child.tagName === 'OPTION' ||
                child.tagName === 'OPTGROUP' ||
                child.tagName === 'HR') {
                this.select.appendChild(child.cloneNode(true));
            }
        });
        // Restore the selected value
        this.select.value = currentValue;
        this.value = this.select.value;
    }
    get supportingOrValidationText() {
        return (this.hasError && this.getErrorText()) || this.supportingText;
    }
    renderSupportingText() {
        if (!this.supportingOrValidationText) {
            return null;
        }
        return html `
      ${this.hasError
            ? html `<span class="made-c-form-control__error-icon">
            <made-icon
              name="exclamation-triangle-solid-12px"
              library="system"
            ></made-icon>
          </span>`
            : nothing}
      <span>${this.supportingOrValidationText}</span>
    `;
    }
    render() {
        const hasLabel = this.label
            ? true
            : !!this.hasSlotController.test(this.LABEL);
        const role = this.hasError ? 'alert' : null;
        const supportingText = this.renderSupportingText();
        return html `
      <div
        part="form-control"
        class=${classMap({
            'made-c-form-control': true,
            'made-c-form-control--medium': this.size === 'medium',
            'made-c-form-control--large': this.size === 'large',
            'made-c-form-control--has-label': hasLabel,
            'made-c-form-control--error': !this.disabled && this.hasError,
            'made-c-form-control--has-supporting-text': !!supportingText
        })}
      >
        <label
          part="label"
          class="made-c-form-control__label"
          for="select"
          aria-hidden=${hasLabel ? 'false' : 'true'}
        >
          <slot name=${this.LABEL}>${this.label}</slot>
        </label>
        <div
          part="select-container"
          class=${classMap({
            'made-c-select': true,
            'made-c-select--medium': this.size === 'medium',
            'made-c-select--large': this.size === 'large',
            'made-c-select--disabled': this.disabled,
            'made-c-select--focused': this.hasFocus
        })}
        >
          <select
            part="select"
            id="select"
            class="made-c-select__control"
            name=${ifDefined(this.name)}
            ?disabled=${this.disabled}
            ?required=${this.required}
            .value=${live(this.value)}
            autocomplete="off"
            ?autofocus=${this.autofocus}
            aria-describedby="supporting-text"
            @change=${this.handleChange}
            @focus=${this.handleFocus}
            @blur=${this.handleBlur}
          ></select>
          <span
            part="chevron"
            class="made-c-select__chevron"
            aria-hidden="true"
          >
            <made-icon name="chevron-down" library="system"></made-icon>
          </span>
        </div>
        <div
          part="supporting-text"
          id="supporting-text"
          class="made-c-form-control__supporting-text"
          role=${ifDefined(role ?? undefined)}
        >
          ${supportingText}
        </div>
      </div>
    `;
    }
    reset() {
        this.value = this.getAttribute('value') ?? '';
        this.hasNativeError = false;
        this.nativeErrorText = '';
    }
    getErrorText() {
        return this.error ? this.errorText : this.nativeErrorText;
    }
    handleBlur() {
        this.hasFocus = false;
    }
    handleChange() {
        this.value = this.select.value;
        if (this.hasError) {
            this.reportValidity();
        }
        this.dispatchEvent(new Event('change', { bubbles: true, composed: true }));
    }
    handleFocus() {
        this.hasFocus = true;
    }
    /** Sets focus on the select. */
    focus(options) {
        this.select.focus(options);
    }
    /** Removes focus from the select. */
    blur() {
        this.select.blur();
    }
    [getFormValue]() {
        return this.value;
    }
    formResetCallback() {
        this.reset();
    }
    // eslint-disable-next-line @typescript-eslint/no-shadow
    formStateRestoreCallback(state) {
        this.value = state;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [createValidator]() {
        return new SelectValidator(() => ({
            value: this.value,
            required: this.required
        }));
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [getValidityAnchor]() {
        return this.select;
    }
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-expect-error
    [onReportValidity](invalidEvent) {
        // Prevent default pop-up behavior.
        invalidEvent?.preventDefault();
        this.hasNativeError = !!invalidEvent;
        this.nativeErrorText = this.validationMessage;
    }
}
MadeSelect.styles = [componentStyles, formControlStyles, styles];
MadeSelect.dependencies = {
    'made-icon': MadeIcon
};
export default MadeSelect;
__decorate([
    query('select')
], MadeSelect.prototype, "select", void 0);
__decorate([
    state()
], MadeSelect.prototype, "hasFocus", void 0);
__decorate([
    state()
], MadeSelect.prototype, "hasNativeError", void 0);
__decorate([
    state()
], MadeSelect.prototype, "nativeErrorText", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeSelect.prototype, "error", void 0);
__decorate([
    property({ attribute: 'error-text' })
], MadeSelect.prototype, "errorText", void 0);
__decorate([
    property()
], MadeSelect.prototype, "value", void 0);
__decorate([
    property()
], MadeSelect.prototype, "label", void 0);
__decorate([
    property({ reflect: true })
], MadeSelect.prototype, "size", void 0);
__decorate([
    property({ attribute: 'supporting-text' })
], MadeSelect.prototype, "supportingText", void 0);
__decorate([
    property()
], MadeSelect.prototype, "placeholder", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeSelect.prototype, "required", void 0);
__decorate([
    property({ attribute: 'required-indicator', reflect: true })
], MadeSelect.prototype, "requiredIndicator", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeSelect.prototype, "optional", void 0);
__decorate([
    property({
        type: Boolean,
        reflect: true,
        attribute: 'no-required-indicator'
    })
], MadeSelect.prototype, "noRequiredIndicator", void 0);
__decorate([
    property({ type: Boolean })
], MadeSelect.prototype, "autofocus", void 0);
