// @mc-made/web-components/dist/components/sheet/src/sheet.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { classMap } from 'lit/directives/class-map.js';
import { html } from 'lit';
import { property, query } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIconButton from '../../icon-button/made-icon-button.js';
import styles from './sheet.styles.js';
/**
 * @summary Sheets slide in from the edge of the screen to present secondary
 * content or actions without navigating away from the current page.
 * @documentation '../stories/sheet.mdx'
 *
 * @fires mde-show - Emitted when the sheet opens (before the open animation completes).
 * @fires mde-hide - Emitted after the sheet finishes closing.
 * @fires mde-request-close - Emitted when the user attempts to close the sheet
 *   (backdrop click, Escape key, or close button). Call `event.preventDefault()` to cancel.
 *   The event detail is `{ source: 'backdrop' | 'escape' | 'close-button' }`.
 *
 * @slot - The sheet's main content.
 *
 * @csspart base - The `<dialog>` element.
 * @csspart panel - The inner panel that slides in from the edge.
 * @csspart body - The scrollable content region inside the panel.
 *
 * @cssproperty --size - Override the sheet's width (for start/end placements) or
 *   height (for top/bottom placements). Default: `24rem`.
 *
 * @dependency made-icon-button
 */
class MadeSheet extends MadeElement {
    constructor() {
        super(...arguments);
        /** Whether the sheet is open. */
        this.open = false;
        /**
         * The edge the sheet slides in from.
         * @type {'start' | 'end' | 'top' | 'bottom'}
         */
        this.placement = 'end';
        // ─── Event handlers ──────────────────────────────────────────────────────
        this._handleCancel = (event) => {
            event.preventDefault();
            this._requestClose('escape');
        };
    }
    // ─── Lifecycle ───────────────────────────────────────────────────────────
    firstUpdated() {
        if (this.open) {
            this.dialog.showModal();
            this._animateIn();
            document.body.style.overflow = 'hidden';
        }
    }
    updated(changedProperties) {
        if (changedProperties.has('open')) {
            if (this.open && !this.dialog.open) {
                this.dialog.showModal();
                document.body.style.overflow = 'hidden';
                this.emit('mde-show');
                // Double rAF: the first frame lets the browser paint the panel at its
                // initial off-screen transform; the second adds --open to start the
                // CSS transition. This is required when showing from display:none.
                requestAnimationFrame(() => {
                    requestAnimationFrame(() => this._animateIn());
                });
            }
            else if (!this.open && this.dialog.open) {
                this._animateOut(() => {
                    this.dialog.close();
                    document.body.style.overflow = '';
                    this.emit('mde-hide');
                });
            }
        }
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        if (this.open) {
            document.body.style.overflow = '';
        }
    }
    // ─── Animation helpers ───────────────────────────────────────────────────
    _animateIn() {
        this.panel.classList.add('made-c-sheet__panel--open');
        this.panel.classList.remove('made-c-sheet__panel--close');
    }
    _animateOut(onDone) {
        this.panel.classList.add('made-c-sheet__panel--close');
        this.panel.classList.remove('made-c-sheet__panel--open');
        let done = false;
        const finish = () => {
            if (done)
                return;
            done = true;
            this.panel.classList.remove('made-c-sheet__panel--close');
            onDone();
        };
        this.panel.addEventListener('transitionend', finish, { once: true });
        let transitionStarted = false;
        const onTransitionRun = () => {
            transitionStarted = true;
        };
        this.panel.addEventListener('transitionrun', onTransitionRun, {
            once: true
        });
        setTimeout(() => {
            this.panel.removeEventListener('transitionrun', onTransitionRun);
            if (!transitionStarted) {
                this.panel.removeEventListener('transitionend', finish);
                finish();
            }
        }, 50);
    }
    // ─── Close request ───────────────────────────────────────────────────────
    _requestClose(source) {
        const event = this.emit('mde-request-close', {
            cancelable: true,
            detail: { source }
        });
        if (!event.defaultPrevented) {
            this.open = false;
        }
    }
    _handlePointerDown(event) {
        if (event.target === this.dialog) {
            this._requestClose('backdrop');
        }
    }
    // ─── Render ──────────────────────────────────────────────────────────────
    render() {
        return html `
      <dialog
        part="base"
        class=${classMap({
            'made-c-sheet': true,
            'made-c-sheet--start': this.placement === 'start',
            'made-c-sheet--end': this.placement === 'end',
            'made-c-sheet--top': this.placement === 'top',
            'made-c-sheet--bottom': this.placement === 'bottom'
        })}
        aria-modal="true"
        @cancel=${this._handleCancel}
        @pointerdown=${this._handlePointerDown}
      >
        <div part="panel" class="made-c-sheet__panel">
          <made-icon-button
            class="made-c-sheet__close-button"
            variant="ghost"
            icon-name="x-mark-outline"
            library="system"
            size="small"
            aria-label=${this.localize.term('close')}
            @click=${this._requestClose.bind(this, 'close-button')}
          ></made-icon-button>
          <div part="body" class="made-c-sheet__body">
            <slot></slot>
          </div>
        </div>
      </dialog>
    `;
    }
}
MadeSheet.styles = [componentStyles, styles];
MadeSheet.dependencies = {
    'made-icon-button': MadeIconButton
};
export default MadeSheet;
__decorate([
    query('dialog')
], MadeSheet.prototype, "dialog", void 0);
__decorate([
    query('.made-c-sheet__panel')
], MadeSheet.prototype, "panel", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], MadeSheet.prototype, "open", void 0);
__decorate([
    property({ reflect: true })
], MadeSheet.prototype, "placement", void 0);
