// @mc-made/web-components/dist/components/sheet/src/sheet.styles.js
import { css } from 'lit';
export default css `
  :host {
    --size: 24rem;
    --transition-duration: var(--made-motion-duration-longest);
    display: contents;
  }

  /* ── Dialog (backdrop layer) ─────────────────────────────────────── */

  .made-c-sheet {
    position: fixed;
    inset: 0;
    width: 100%;
    height: 100%;
    max-width: none;
    max-height: none;
    margin: 0;
    padding: 0;
    background: transparent;
    border: none;
    overflow: visible;
  }

  .made-c-sheet:focus {
    outline: none;
  }

  .made-c-sheet::backdrop {
    background-color: color-mix(
      in srgb,
      var(--made-color-background-backdrop) 40%,
      transparent
    );
    opacity: 0;
    transition: opacity var(--transition-duration)
      var(--made-motion-easing-standard-ease);
  }

  /* Show backdrop when panel is open */
  .made-c-sheet__panel--open ~ dialog::backdrop,
  .made-c-sheet:has(.made-c-sheet__panel--open)::backdrop {
    opacity: 1;
  }

  /* ── Panel base ──────────────────────────────────────────────────── */

  .made-c-sheet__panel {
    position: fixed;
    display: flex;
    flex-direction: column;
    padding: var(--made-component-sheet-padding-y)
      var(--made-component-sheet-padding-x);
    background-color: var(--made-color-background-surface-default);
    font-family: var(--made-font-family-primary);
    color: var(--made-color-text-primary);
    overflow: hidden;
    border-radius: var(--made-component-sheet-border-radius);
    box-shadow: var(--made-component-sheet-box-shadow);
    transition: transform var(--transition-duration)
      var(--made-motion-easing-standard-ease);
    will-change: transform;
    z-index: 1;
  }

  .made-c-sheet__close-button {
    position: absolute;
    top: var(--made-component-sheet-padding-y);
    right: var(--made-component-sheet-padding-x);
  }

  .made-c-sheet__close-button::part(base) {
    border-radius: var(--made-border-radius-04);
    height: var(--made-size-space-6-x);
    width: var(--made-size-space-6-x);
  }

  /* ── Placement: end (default, right) ─────────────────────────────── */

  .made-c-sheet--end .made-c-sheet__panel {
    top: 0;
    inset-inline-end: 0;
    height: 100%;
    width: min(var(--size), 75dvw);
    border-inline-start: 1px solid var(--made-component-sheet-border-color);
    transform: translateX(100%);
  }

  :host([dir='rtl']) .made-c-sheet--end .made-c-sheet__panel {
    transform: translateX(-100%);
  }

  /* ── Placement: start (left) ──────────────────────────────────────── */

  .made-c-sheet--start .made-c-sheet__panel {
    top: 0;
    inset-inline-start: 0;
    height: 100%;
    width: min(var(--size), 75dvw);
    border-inline-end: 1px solid var(--made-component-sheet-border-color);
    transform: translateX(-100%);
  }

  :host([dir='rtl']) .made-c-sheet--start .made-c-sheet__panel {
    transform: translateX(100%);
  }

  /* ── Placement: top ──────────────────────────────────────────────── */

  .made-c-sheet--top .made-c-sheet__panel {
    top: 0;
    left: 0;
    right: 0;
    width: 100%;
    height: min(var(--size), 75dvh);
    border-block-end: 1px solid var(--made-component-sheet-border-color);
    transform: translateY(-100%);
  }

  /* ── Placement: bottom ───────────────────────────────────────────── */

  .made-c-sheet--bottom .made-c-sheet__panel {
    bottom: 0;
    left: 0;
    right: 0;
    width: 100%;
    height: min(var(--size), 75dvh);
    border-block-start: 1px solid var(--made-component-sheet-border-color);
    transform: translateY(100%);
  }

  /* ── Open / close animation states ──────────────────────────────── */

  .made-c-sheet__panel--open {
    transform: translate(0, 0) !important;
  }

  /* Close state: return to initial off-screen position, driven by placement class */
  .made-c-sheet--end .made-c-sheet__panel--close {
    transform: translateX(100%);
  }
  .made-c-sheet--start .made-c-sheet__panel--close {
    transform: translateX(-100%);
  }
  .made-c-sheet--top .made-c-sheet__panel--close {
    transform: translateY(-100%);
  }
  .made-c-sheet--bottom .made-c-sheet__panel--close {
    transform: translateY(100%);
  }

  /* ── Body ────────────────────────────────────────────────────────── */

  .made-c-sheet__body {
    flex: 1 1 auto;
    overflow: auto;
    -webkit-overflow-scrolling: touch;
  }
`;
