// @mc-made/web-components/dist/components/breadcrumb/src/breadcrumb.component.js
/* v8 ignore next 6 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html } from 'lit';
import { ifDefined } from 'lit/directives/if-defined.js';
import { property, query, state } from 'lit/decorators.js';
import componentStyles from '../../../styles/component.styles.js';
import MadeElement from '../../../internal/made-element.js';
import MadeIcon from '../../icon/made-icon.js';
import styles from './breadcrumb.styles.js';
/**
 * @summary Breadcrumb provides a navigation trail for users, showing their current page within a hierarchy of pages.
 *
 * @dependency made-icon
 *
 * @slot - One or more made-breadcrumb-item elements.
 *
 * @csspart base - The component's base wrapper.
 * @csspart list - The wrapper around the list of breadcrumb items.
 * @csspart item - The wrapper around an individual breadcrumb item.
 * @csspart separator - The separator between breadcrumb items.
 */
class MadeBreadcrumb extends MadeElement {
    constructor() {
        super(...arguments);
        /**
         * The separator to display between breadcrumb items.
         */
        this.separator = 'chevron';
        /**
         * Size of the breadcrumbs text.
         */
        this.size = 'medium';
        // In handleSlotChange, this will contain the slot names of the breadcrumb items (e.g., ['item-0', 'item-1', ...])
        // These slot names will be used to render the breadcrumb items in order with their separators and, as such, need to be reactive (e.g., a @state variable)
        this.slotNames = [];
        // All generated slot names will follow the template of "item-{id}" where id is a unique number generated for each breadcrumb item.
        this.slotId = 0;
        // Maps each breadcrumb item element to its assigned slot name. This way we can assign the same slot name to the same breadcrumb item element across renders/slot changes.
        this.slotElementToNameMapping = new WeakMap();
        // Keeps track of whether we're currently processing the slot names in response to a slot change.
        this.isProcessing = false;
    }
    /**
     * We want each breadcrumb item to have its own shadow DOM to allow so that we can provide consistent styling regardless of how the breadcrumb item
     * is rendered (e.g., via the prop-based API with made-breadcrumb-item or via the slotted content API with a custom link component). To do this, we need
     * to assign a unique slot name for each breadcrumb item (and its separators, if any). Since we also want to support dynamic addition of breadcrumb items,
     * we can't rely on a static set of slot names. Instead, we need to dynamically assign slot names to each breadcrumb item whenever the default slot changes
     * (e.g., when a breadcrumb items is added).
     *
     * (Note that currently, dynamically removal is not supported in this implementation, as that would likely require a more complex implementation that involves a
     * MutationObserver to detect removed breadcrumb items and re-assign slot names accordingly. However, we do support dynamic addition of breadcrumb items, which
     * is likely the more common use case.)
     */
    handleSlotChange() {
        // Skip if default slot is not ready/available - this is most likely because the `@query` decorator hasn't populated the `defaultSlot` property yet.
        // In this case, we can return early and attempt to sync again the next time the slot changes.
        if (!this.defaultSlot)
            return;
        // Because this method is changing the DOM by assigning slot names to breadcrumb items, it may trigger additional slotchange events. To avoid an infinite loop of
        // slotchange events triggering more slotchange events, we use a simple locking mechanism with the `isProcessing` flag to ensure that we only process one slot change at
        // a time. If a slot change occurs while we're already processing, we skip it since we know we'll be processing again once the current process is complete.
        if (this.isProcessing) {
            return;
        }
        // Set lock to prevent processing concurrent/additional slot changes until we're done syncing
        this.isProcessing = true;
        // Grab all the light DOM children that are breadcrumb items (e.g., made-breadcrumb-item) - we can identify them by their tag name since we know the tag name of
        // breadcrumb items will always be "made-breadcrumb-item" regardless of how they are rendered (e.g., via the prop-based API or the slotted content API).
        // Realistically, there shouldn't be any non-breadcrumb-item children, but we do this defensively to ensure we only assign slot names to breadcrumb items.
        // Note we pull from the children of the breadcrumb component itself, not the default slot, since some/all breadcrumb items may have assigned slot names and, as such,
        // won't be in the default slot (since elements cannot be in multiple slots simultaneously).
        const breadcrumbItemElements = Array.from(this.children).filter(child => child.tagName.toLowerCase() === 'made-breadcrumb-item');
        // If there are no breadcrumb items, we can clear the slot names and return early since there's nothing to render.
        if (breadcrumbItemElements.length === 0) {
            this.slotNames = [];
            this.isProcessing = false;
            this.requestUpdate();
            return;
        }
        // Local variable to keep track of slot names for each breadcrumb item
        const breadcrumbItemSlotNames = [];
        // Keep track of whether we need to update the slot names (e.g., if new breadcrumb items were added that need slot names)
        // This will help us avoid unnecessary updates/renders if nothing actually changed
        let needsUpdate = false;
        for (const breadcrumbItemElement of breadcrumbItemElements) {
            // If breadcrumb item already has a slot name (typically from a previous render), then keep it and skip to the next breadcrumb item
            let slotName = breadcrumbItemElement.getAttribute('slot');
            if (slotName && slotName !== '') {
                breadcrumbItemSlotNames.push(slotName);
                continue;
            }
            // For new breadcrumb items (e.g., ones that we are processing for the first time), generate a stable id for its slot name.
            slotName =
                this.slotElementToNameMapping.get(breadcrumbItemElement) || null;
            if (!slotName) {
                // We generate the id using a simple incrementing counter (`slotId`) that ensures uniqueness across breadcrumb items and renders.
                slotName = `item-${this.slotId++}`;
                // Store the mapping of breadcrumb item element to slot name to ensure that if the same breadcrumb item element is processed again (e.g., due to another slot change),
                // we can assign it the same slot name and avoid unnecessary changes/updates.
                this.slotElementToNameMapping.set(breadcrumbItemElement, slotName);
            }
            // Set slot attribute for breadcrumb item so that it now has a named slot
            breadcrumbItemElement.setAttribute('slot', slotName);
            // Add slot name to the list of slot names for rendering
            breadcrumbItemSlotNames.push(slotName);
            needsUpdate = true;
        }
        // Mark last breadcrumb item as current, which is useful for accessibility purposes
        if (breadcrumbItemElements.length > 0) {
            const lastBreadcrumbElement = breadcrumbItemElements[breadcrumbItemElements.length - 1];
            lastBreadcrumbElement.setAttribute('current', 'true');
        }
        // Only update slot names in component state if something has changed (e.g., new breadcrumb items were added)
        if (needsUpdate ||
            this.slotNames.length !== breadcrumbItemSlotNames.length) {
            this.slotNames = breadcrumbItemSlotNames;
            // Trigger an update to re-render breadcrumb items with new slot names
            this.requestUpdate();
        }
        // Release lock after a microtask to allow the DOM to settle
        queueMicrotask(() => {
            this.isProcessing = false;
        });
    }
    // Render the separator element based on the selected separator type
    // We treat the separator as purely presentational and hide it from assistive technologies since it doesn't convey meaningful information on its own
    // We allow the separator to be customized via CSS part for customization (e.g., custom spacing, custom separator)
    renderSeparator() {
        // The current set of system icons does not contain a slash icon, so we render the slash separator with a simple span element instead of using made-icon
        if (this.separator === 'slash') {
            return html `<span
        class="made-c-breadcrumb__separator"
        part="separator"
        aria-hidden="true"
        >&nbsp;/&nbsp;</span
      >`;
        }
        return html `<made-icon
      class="made-c-breadcrumb__separator"
      part="separator"
      library="system"
      name="chevron-right"
      size="small"
      aria-hidden="true"
    ></made-icon>`;
    }
    // Whenever the default slot changes (e.g., breadcrumb items are added), we re-sync the slot names for all breadcrumb items to ensure new items are rendered
    // Initially the default slot will contain all breadcrumb items. Then as we assign slot names to each breadcrumb item, they will be rendered with that assigned name.
    // The made-breadcrumb-item remain in the same DOM position - the slot attribute only affects how they are rendered in the shadow DOM, not where they actually are in the
    // light DOM.
    render() {
        return html `
      <nav
        part="base"
        class="made-c-breadcrumb"
        aria-label=${ifDefined(this.ariaLabel)}
      >
        <ol
          part="list"
          class="made-c-breadcrumb__list ${this.size === 'small'
            ? 'made-c-breadcrumb__list--small'
            : ''}"
        >
          ${this.slotNames.map((slotName, index) => {
            const isLast = index === this.slotNames.length - 1;
            return html `
              <li
                part="item"
                class="made-c-breadcrumb__item ${this.size === 'small'
                ? 'made-c-breadcrumb__item--small'
                : ''}"
              >
                <slot name=${slotName}></slot>
                ${!isLast ? this.renderSeparator() : ''}
              </li>
            `;
        })}
        </ol>
        <div style="display: none;">
          <slot @slotchange=${this.handleSlotChange}></slot>
        </div>
      </nav>
    `;
    }
}
MadeBreadcrumb.styles = [componentStyles, styles];
// we use made-icon to render the default chevron separator
MadeBreadcrumb.dependencies = {
    'made-icon': MadeIcon
};
export default MadeBreadcrumb;
__decorate([
    property()
], MadeBreadcrumb.prototype, "separator", void 0);
__decorate([
    property()
], MadeBreadcrumb.prototype, "size", void 0);
__decorate([
    property({ attribute: 'aria-label' })
], MadeBreadcrumb.prototype, "ariaLabel", void 0);
__decorate([
    query('slot:not([name])')
], MadeBreadcrumb.prototype, "defaultSlot", void 0);
__decorate([
    state()
], MadeBreadcrumb.prototype, "slotNames", void 0);
