// @mc-made/web-components/dist/components/breadcrumb/src/breadcrumb.styles.js
import { css } from 'lit';
export default css `
  :host {
    display: block;
    position: relative;
  }

  .made-c-breadcrumb {
    display: flex;
    width: 100%;
  }

  .made-c-breadcrumb__list {
    align-items: center;
    display: flex;
    flex-wrap: wrap;
    list-style: none;
    margin: 0;
    padding: 0;
  }

  .made-c-breadcrumb__item {
    align-items: center;
    display: inline-flex;
    font-size: var(--made-component-breadcrumb-medium-font-size);
    font-weight: var(--made-component-breadcrumb-font-weight);
    line-height: var(--made-component-breadcrumb-medium-line-height);
    color: var(--made-color-text-secondary);
    letter-spacing: var(--made-letter-spacing-default);
    text-decoration: var(--made-text-decoration-none);
  }

  .made-c-breadcrumb__item--small {
    font-size: var(--made-component-breadcrumb-small-font-size);
    line-height: var(--made-component-breadcrumb-small-line-height);
  }

  .made-c-breadcrumb__separator {
    display: inline-flex;
    padding-left: var(--made-component-breadcrumb-gap);
    padding-right: var(--made-component-breadcrumb-gap);
  }
  .made-c-breadcrumb__item--small .made-c-breadcrumb__separator {
    --width: 0.75rem;
    --height: 0.75rem;
  }
`;
